<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.450">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="dcterms.date" content="2024-03-11">

<title>Survive with Omics - Supplemental information for ‘Tutorial on survival modeling with applications to omics data’</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
/* CSS for citations */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging-indent div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>

  <script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

</head>

<body class="nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg navbar-dark ">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container">
    <a class="navbar-brand" href="./index.html">
    <span class="navbar-title">Survive with Omics</span>
    </a>
  </div>
        <div class="quarto-navbar-tools ms-auto">
    <div class="dropdown">
      <a href="" title="" id="quarto-navigation-tool-dropdown-0" class="quarto-navigation-tool dropdown-toggle px-1" data-bs-toggle="dropdown" aria-expanded="false" aria-label=""><i class="bi bi-github"></i></a>
      <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="quarto-navigation-tool-dropdown-0">
          <li>
            <a class="dropdown-item quarto-navbar-tools ms-auto-item" href="https://github.com/ocbe-uio/survomics">
            Source Code
            </a>
          </li>
          <li>
            <a class="dropdown-item quarto-navbar-tools ms-auto-item" href="https://github.com/ocbe-uio/survomics/issues/new">
            Report a Bug
            </a>
          </li>
      </ul>
    </div>
</div>
          <div id="quarto-search" class="" title="Search"></div>
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-full page-navbar">
<!-- sidebar -->
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title"><strong>Table of Contents</strong></h2>
   
  <ul>
  <li><a href="#introduction" id="toc-introduction" class="nav-link active" data-scroll-target="#introduction">Introduction</a></li>
  <li><a href="#tcga-survival-and-clinical-data" id="toc-tcga-survival-and-clinical-data" class="nav-link" data-scroll-target="#tcga-survival-and-clinical-data">TCGA survival and clinical data</a></li>
  <li><a href="#tcga-omics-data" id="toc-tcga-omics-data" class="nav-link" data-scroll-target="#tcga-omics-data">TCGA omics data</a></li>
  <li><a href="#survival-analysis-with-low-dimensional-input-data" id="toc-survival-analysis-with-low-dimensional-input-data" class="nav-link" data-scroll-target="#survival-analysis-with-low-dimensional-input-data">Survival analysis with low-dimensional input data</a>
  <ul>
  <li><a href="#nonparametric-survival-analysis" id="toc-nonparametric-survival-analysis" class="nav-link" data-scroll-target="#nonparametric-survival-analysis">Nonparametric survival analysis</a></li>
  <li><a href="#feature-preselectionfiltering" id="toc-feature-preselectionfiltering" class="nav-link" data-scroll-target="#feature-preselectionfiltering">Feature preselection/filtering</a>
  <ul class="collapse">
  <li><a href="#knowledge-filter" id="toc-knowledge-filter" class="nav-link" data-scroll-target="#knowledge-filter">Knowledge filter</a></li>
  <li><a href="#p-value-filter" id="toc-p-value-filter" class="nav-link" data-scroll-target="#p-value-filter">P-value filter</a></li>
  <li><a href="#variance-filter" id="toc-variance-filter" class="nav-link" data-scroll-target="#variance-filter">Variance filter</a></li>
  </ul></li>
  </ul></li>
  <li><a href="#survival-analysis-with-high-dimensional-input-data" id="toc-survival-analysis-with-high-dimensional-input-data" class="nav-link" data-scroll-target="#survival-analysis-with-high-dimensional-input-data">Survival analysis with high-dimensional input data</a>
  <ul>
  <li><a href="#unsupervised-learning-omics-data" id="toc-unsupervised-learning-omics-data" class="nav-link" data-scroll-target="#unsupervised-learning-omics-data">Unsupervised learning (omics data)</a></li>
  <li><a href="#supervised-learning-omics-and-survival-data" id="toc-supervised-learning-omics-and-survival-data" class="nav-link" data-scroll-target="#supervised-learning-omics-and-survival-data">Supervised learning (omics and survival data)</a>
  <ul class="collapse">
  <li><a href="#dimension-reduction-for-cox-models" id="toc-dimension-reduction-for-cox-models" class="nav-link" data-scroll-target="#dimension-reduction-for-cox-models">Dimension reduction for Cox models</a></li>
  <li><a href="#penalized-cox-models" id="toc-penalized-cox-models" class="nav-link" data-scroll-target="#penalized-cox-models">Penalized Cox models</a></li>
  <li><a href="#sparse-bayesian-cox-models" id="toc-sparse-bayesian-cox-models" class="nav-link" data-scroll-target="#sparse-bayesian-cox-models">Sparse Bayesian Cox models</a></li>
  </ul></li>
  </ul></li>
  <li><a href="#survival-model-validation" id="toc-survival-model-validation" class="nav-link" data-scroll-target="#survival-model-validation">Survival model validation</a>
  <ul>
  <li><a href="#model-evaluation-classic" id="toc-model-evaluation-classic" class="nav-link" data-scroll-target="#model-evaluation-classic">Model evaluation (classic)</a>
  <ul class="collapse">
  <li><a href="#discrimination-metrics" id="toc-discrimination-metrics" class="nav-link" data-scroll-target="#discrimination-metrics">Discrimination metrics</a></li>
  <li><a href="#calibration-metrics" id="toc-calibration-metrics" class="nav-link" data-scroll-target="#calibration-metrics">Calibration metrics</a></li>
  <li><a href="#overall-metrics" id="toc-overall-metrics" class="nav-link" data-scroll-target="#overall-metrics">Overall metrics</a></li>
  <li><a href="#uq1" id="toc-uq1" class="nav-link" data-scroll-target="#uq1">Uncertainty Quantification</a></li>
  <li><a href="#feature-stability-analysis" id="toc-feature-stability-analysis" class="nav-link" data-scroll-target="#feature-stability-analysis">Feature stability analysis</a></li>
  <li><a href="#graphComp" id="toc-graphComp" class="nav-link" data-scroll-target="#graphComp">Graphical representation</a></li>
  </ul></li>
  <li><a href="#mlr3" id="toc-mlr3" class="nav-link" data-scroll-target="#mlr3">Model evaluation (mlr3)</a>
  <ul class="collapse">
  <li><a href="#workflow" id="toc-workflow" class="nav-link" data-scroll-target="#workflow">Workflow</a></li>
  <li><a href="#discrimination-metrics-1" id="toc-discrimination-metrics-1" class="nav-link" data-scroll-target="#discrimination-metrics-1">Discrimination metrics</a></li>
  <li><a href="#calibration-metrics-1" id="toc-calibration-metrics-1" class="nav-link" data-scroll-target="#calibration-metrics-1">Calibration metrics</a></li>
  <li><a href="#overall-metrics-1" id="toc-overall-metrics-1" class="nav-link" data-scroll-target="#overall-metrics-1">Overall metrics</a></li>
  <li><a href="#uncertainty-quantification" id="toc-uncertainty-quantification" class="nav-link" data-scroll-target="#uncertainty-quantification">Uncertainty Quantification</a></li>
  <li><a href="#feature-stability-analysis-1" id="toc-feature-stability-analysis-1" class="nav-link" data-scroll-target="#feature-stability-analysis-1">Feature stability analysis</a></li>
  </ul></li>
  </ul></li>
  <li><a href="#multi-omics-integrative-modeling" id="toc-multi-omics-integrative-modeling" class="nav-link" data-scroll-target="#multi-omics-integrative-modeling">Multi-omics integrative modeling</a></li>
  <li><a href="#references" id="toc-references" class="nav-link" data-scroll-target="#references">References</a></li>
  </ul>
<div class="toc-actions"><div><i class="bi bi-github"></i></div><div class="action-links"><p><a href="https://github.com/ocbe-uio/survomics/blob/main/survomics.qmd" class="toc-action">View source</a></p><p><a href="https://github.com/ocbe-uio/survomics/issues/new" class="toc-action">Report an issue</a></p></div></div></nav>
    </div>
<!-- main -->
<main class="content column-page-left" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Supplemental information for ‘Tutorial on survival modeling with applications to omics data’</h1>
</div>



<div class="quarto-title-meta">

    
    <div>
    <div class="quarto-title-meta-heading">Published</div>
    <div class="quarto-title-meta-contents">
      <p class="date">March 11, 2024</p>
    </div>
  </div>
  
    
  </div>
  

</header>

<p><br> This is a step-by-step R tutorial using <a href="https://www.cancer.gov/about-nci/organization/ccg/research/structural-genomics/tcga">The Cancer Genome Atlas</a> (TCGA) survival and omics data for the article <a href="https://doi.org/10.1093/bioinformatics/btae132"><strong><em>Tutorial on survival modeling with applications to omics data</em></strong></a> <span class="citation" data-cites="Zhao2024">(<a href="#ref-Zhao2024" role="doc-biblioref">Z. Zhao et al. 2024</a>)</span>.</p>
<section id="introduction" class="level2 unnumbered">
<h2 class="unnumbered anchored" data-anchor-id="introduction">Introduction</h2>
<p>The TCGA database provides an enormous collection of cancer data sets, including survival, clinical and multi-omics data.</p>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
We will use TCGA data to demonstrate:
</div>
</div>
<div class="callout-body-container callout-body">
<ul>
<li>The different data types</li>
<li>Preprocessing of survival and omics data</li>
<li>Analysis of survival data by classical statistical methods</li>
<li>Unsupervised learning for omics data</li>
<li>Frequentist &amp; Bayesian supervised learning for survival and omics data</li>
</ul>
</div>
</div>
</section>
<section id="tcga-survival-and-clinical-data" class="level2 unnumbered">
<h2 class="unnumbered anchored" data-anchor-id="tcga-survival-and-clinical-data">TCGA survival and clinical data</h2>
<p>The R/Bioconductor package <a href="https://bioconductor.org/packages/TCGAbiolinks/"><strong>TCGAbiolinks</strong></a> <span class="citation" data-cites="Mounir2019">(<a href="#ref-Mounir2019" role="doc-biblioref">Mounir 2019</a>)</span> provides a few functions to download and preprocess clinical and multi-omics data from the <a href="https://gdc.cancer.gov/">Genomic Data Commons</a> (GDC) Data Portal for further analysis.</p>
<p>First we load all necessary libraries used in this tutorial except <a href="#mlr3"><strong>mlr3</strong> libraries</a> which will be introduced later. Then we use function <code>GDCquery_clinic()</code> from <strong>TCGAbiolinks</strong> package to query and download TCGA survival and clinical data from multiple cancer types:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="co"># load all libraries used in this tutorial except mlr3</span></span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">"TCGAbiolinks"</span>)</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">"SummarizedExperiment"</span>)</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">"DESeq2"</span>)</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">"dplyr"</span>)</span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">"ggplot2"</span>)</span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">"survival"</span>)</span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">"survminer"</span>)</span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">"M3C"</span>)</span>
<span id="cb1-10"><a href="#cb1-10" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">"glmnet"</span>)</span>
<span id="cb1-11"><a href="#cb1-11" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">"plotmo"</span>)</span>
<span id="cb1-12"><a href="#cb1-12" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">"grpreg"</span>)</span>
<span id="cb1-13"><a href="#cb1-13" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">"SGL"</span>)</span>
<span id="cb1-14"><a href="#cb1-14" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">"psbcGroup"</span>)</span>
<span id="cb1-15"><a href="#cb1-15" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">"psbcSpeedUp"</span>)</span>
<span id="cb1-16"><a href="#cb1-16" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">"BhGLM"</span>)</span>
<span id="cb1-17"><a href="#cb1-17" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">"risksetROC"</span>)</span>
<span id="cb1-18"><a href="#cb1-18" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">"riskRegression"</span>)</span>
<span id="cb1-19"><a href="#cb1-19" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">"peperr"</span>)</span>
<span id="cb1-20"><a href="#cb1-20" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">"c060"</span>)</span>
<span id="cb1-21"><a href="#cb1-21" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">"rms"</span>)</span>
<span id="cb1-22"><a href="#cb1-22" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">"survAUC"</span>)</span>
<span id="cb1-23"><a href="#cb1-23" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">"regplot"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="co"># download the clinical data and extract data for multiple cancers using GDC api method</span></span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a>cancer_types <span class="ot">&lt;-</span> <span class="fu">c</span>(</span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a>  <span class="st">"TCGA-BLCA"</span>, <span class="st">"TCGA-BRCA"</span>, <span class="st">"TCGA-COAD"</span>, <span class="st">"TCGA-LIHC"</span>,</span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a>  <span class="st">"TCGA-LUAD"</span>, <span class="st">"TCGA-PAAD"</span>, <span class="st">"TCGA-PRAD"</span>, <span class="st">"TCGA-THCA"</span></span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true" tabindex="-1"></a>clin <span class="ot">&lt;-</span> <span class="cn">NULL</span></span>
<span id="cb2-7"><a href="#cb2-7" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (i <span class="cf">in</span> <span class="fu">seq_along</span>(cancer_types)) {</span>
<span id="cb2-8"><a href="#cb2-8" aria-hidden="true" tabindex="-1"></a>  tmp <span class="ot">&lt;-</span> TCGAbiolinks<span class="sc">::</span><span class="fu">GDCquery_clinic</span>(<span class="at">project =</span> cancer_types[i], <span class="at">type =</span> <span class="st">"clinical"</span>)</span>
<span id="cb2-9"><a href="#cb2-9" aria-hidden="true" tabindex="-1"></a>  clin <span class="ot">&lt;-</span> <span class="fu">rbind</span>(clin, tmp[, <span class="fu">c</span>(</span>
<span id="cb2-10"><a href="#cb2-10" aria-hidden="true" tabindex="-1"></a>    <span class="st">"project"</span>, <span class="st">"submitter_id"</span>, <span class="st">"vital_status"</span>,</span>
<span id="cb2-11"><a href="#cb2-11" aria-hidden="true" tabindex="-1"></a>    <span class="st">"days_to_last_follow_up"</span>, <span class="st">"days_to_death"</span>,</span>
<span id="cb2-12"><a href="#cb2-12" aria-hidden="true" tabindex="-1"></a>    <span class="st">"age_at_diagnosis"</span>, <span class="st">"gender"</span>, <span class="st">"race"</span>,</span>
<span id="cb2-13"><a href="#cb2-13" aria-hidden="true" tabindex="-1"></a>    <span class="st">"ethnicity"</span>, <span class="st">"ajcc_pathologic_t"</span></span>
<span id="cb2-14"><a href="#cb2-14" aria-hidden="true" tabindex="-1"></a>  )])</span>
<span id="cb2-15"><a href="#cb2-15" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb2-16"><a href="#cb2-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-17"><a href="#cb2-17" aria-hidden="true" tabindex="-1"></a><span class="co"># extract the observed time for each patient and use years as unit</span></span>
<span id="cb2-18"><a href="#cb2-18" aria-hidden="true" tabindex="-1"></a>clin<span class="sc">$</span>time <span class="ot">&lt;-</span> <span class="fu">apply</span>(clin[, <span class="fu">c</span>(<span class="st">"days_to_death"</span>, <span class="st">"days_to_last_follow_up"</span>)], <span class="dv">1</span>, max, <span class="at">na.rm =</span> <span class="cn">TRUE</span>) <span class="sc">/</span> <span class="fl">365.25</span></span>
<span id="cb2-19"><a href="#cb2-19" aria-hidden="true" tabindex="-1"></a>clin<span class="sc">$</span>age <span class="ot">&lt;-</span> clin<span class="sc">$</span>age_at_diagnosis <span class="sc">/</span> <span class="fl">365.25</span></span>
<span id="cb2-20"><a href="#cb2-20" aria-hidden="true" tabindex="-1"></a>clin<span class="sc">$</span>status <span class="ot">&lt;-</span> clin<span class="sc">$</span>vital_status</span>
<span id="cb2-21"><a href="#cb2-21" aria-hidden="true" tabindex="-1"></a>clin <span class="ot">&lt;-</span> clin[, <span class="fu">c</span>(<span class="st">"project"</span>, <span class="st">"submitter_id"</span>, <span class="st">"status"</span>, <span class="st">"time"</span>, <span class="st">"gender"</span>, <span class="st">"age"</span>, <span class="st">"race"</span>, <span class="st">"ethnicity"</span>)]</span>
<span id="cb2-22"><a href="#cb2-22" aria-hidden="true" tabindex="-1"></a><span class="co"># extract patients with positive overall survival time</span></span>
<span id="cb2-23"><a href="#cb2-23" aria-hidden="true" tabindex="-1"></a>clin <span class="ot">&lt;-</span> clin[(clin<span class="sc">$</span>time <span class="sc">&gt;</span> <span class="dv">0</span>) <span class="sc">&amp;</span> (clin<span class="sc">$</span>status <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">"Alive"</span>, <span class="st">"Dead"</span>)), ]</span>
<span id="cb2-24"><a href="#cb2-24" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-25"><a href="#cb2-25" aria-hidden="true" tabindex="-1"></a><span class="co"># frequency table of the patients w.r.t. status, gender and ethnicity</span></span>
<span id="cb2-26"><a href="#cb2-26" aria-hidden="true" tabindex="-1"></a>clin <span class="sc">%&gt;%</span></span>
<span id="cb2-27"><a href="#cb2-27" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">count</span>(status, gender, ethnicity) <span class="sc">%&gt;%</span></span>
<span id="cb2-28"><a href="#cb2-28" aria-hidden="true" tabindex="-1"></a>  <span class="fu">group_by</span>(status) <span class="sc">%&gt;%</span></span>
<span id="cb2-29"><a href="#cb2-29" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">prop =</span> <span class="fu">prop.table</span>(n))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<pre class="border"><code># A tibble: 12 × 5
# Groups:   status [2]
   status gender ethnicity                  n    prop
   &lt;chr&gt;  &lt;chr&gt;  &lt;chr&gt;                  &lt;int&gt;   &lt;dbl&gt;
 1 Alive  female hispanic or latino        75 0.0240 
 2 Alive  female not hispanic or latino  1367 0.438  
 3 Alive  female not reported             328 0.105  
 4 Alive  male   hispanic or latino        34 0.0109 
 5 Alive  male   not hispanic or latino  1041 0.334  
 6 Alive  male   not reported             276 0.0884 
 7 Dead   female hispanic or latino         7 0.00809
 8 Dead   female not hispanic or latino   377 0.436  
 9 Dead   female not reported              64 0.0740 
10 Dead   male   hispanic or latino        10 0.0116 
11 Dead   male   not hispanic or latino   327 0.378  
12 Dead   male   not reported              80 0.0925 </code></pre>
<div class="cell">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="co"># censoring plot by cancer types</span></span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a>ID <span class="ot">&lt;-</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">nrow</span>(clin)</span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a>clin <span class="sc">%&gt;%</span></span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(</span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true" tabindex="-1"></a>    <span class="fu">aes</span>(<span class="at">y =</span> ID, <span class="at">x =</span> time, <span class="at">colour =</span> project, <span class="at">shape =</span> <span class="fu">factor</span>(status))</span>
<span id="cb4-6"><a href="#cb4-6" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">+</span></span>
<span id="cb4-7"><a href="#cb4-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_segment</span>(<span class="fu">aes</span>(<span class="at">x =</span> time, <span class="at">y =</span> ID, <span class="at">xend =</span> <span class="dv">0</span>, <span class="at">yend =</span> ID)) <span class="sc">+</span></span>
<span id="cb4-8"><a href="#cb4-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>() <span class="sc">+</span></span>
<span id="cb4-9"><a href="#cb4-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggtitle</span>(<span class="st">""</span>) <span class="sc">+</span></span>
<span id="cb4-10"><a href="#cb4-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x =</span> <span class="st">"Years"</span>, <span class="at">y =</span> <span class="st">"Patients"</span>) <span class="sc">+</span></span>
<span id="cb4-11"><a href="#cb4-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_shape_discrete</span>(<span class="at">name =</span> <span class="st">"Status"</span>, <span class="at">labels =</span> <span class="fu">c</span>(<span class="st">"Censored"</span>, <span class="st">"Dead"</span>)) <span class="sc">+</span></span>
<span id="cb4-12"><a href="#cb4-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_color_discrete</span>(</span>
<span id="cb4-13"><a href="#cb4-13" aria-hidden="true" tabindex="-1"></a>    <span class="at">name =</span> <span class="st">"Cancer"</span>,</span>
<span id="cb4-14"><a href="#cb4-14" aria-hidden="true" tabindex="-1"></a>    <span class="at">labels =</span> <span class="fu">c</span>(<span class="st">"Bladder"</span>, <span class="st">"Breast"</span>, <span class="st">"Colon"</span>, <span class="st">"Liver"</span>, <span class="st">"Lung adeno"</span>, <span class="st">"Pancreatic"</span>, <span class="st">"Prostate"</span>, <span class="st">"Thyroid"</span>)</span>
<span id="cb4-15"><a href="#cb4-15" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">+</span></span>
<span id="cb4-16"><a href="#cb4-16" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">legend.position =</span> <span class="st">"top"</span>, <span class="at">legend.direction =</span> <span class="st">"vertical"</span>) <span class="sc">+</span></span>
<span id="cb4-17"><a href="#cb4-17" aria-hidden="true" tabindex="-1"></a>  <span class="fu">guides</span>(<span class="at">color =</span> <span class="fu">guide_legend</span>(<span class="at">nrow =</span> <span class="dv">2</span>, <span class="at">byrow =</span> <span class="cn">TRUE</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><img src="fig/TCGA_survival.png" class="img-fluid figure-img" style="width:60.0%"></p>
<figcaption class="figure-caption"><em>Overall survival times and status of pan-cancer patients from TCGA.</em></figcaption>
</figure>
</div>
<p><br></p>
</section>
<section id="tcga-omics-data" class="level2 unnumbered">
<h2 class="unnumbered anchored" data-anchor-id="tcga-omics-data">TCGA omics data</h2>
<p>We use function <code>GDCquery()</code> to query and use <code>GDCdownload()</code> and <code>GDCprepare()</code> to download TCGA omics data from one cancer type (breast cancer). The argument <code>data.category</code> in function <code>GDCquery()</code> specifies the type of omics data, such as <code>"Copy Number Variation"</code>, <code>"DNA Methylation"</code>, <code>"Transcriptome Profiling"</code>, <code>"Simple Nucleotide Variation"</code>. Note that the downloaded omics data are accompanied by metadata including survival outcomes, clinical and demographic variables. The accompanied metadata are almost the same as the clinical data downloaded via <code>GDCquery_clinic()</code> in the previous section but here only corresponding to one cancer type.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="co"># download TCGA breast cancer (BRCA) mRNA-Seq data using GDC api method</span></span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a>query <span class="ot">&lt;-</span> TCGAbiolinks<span class="sc">::</span><span class="fu">GDCquery</span>(</span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">project =</span> <span class="st">"TCGA-BRCA"</span>,</span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">data.category =</span> <span class="st">"Transcriptome Profiling"</span>,</span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">data.type =</span> <span class="st">"Gene Expression Quantification"</span>,</span>
<span id="cb5-6"><a href="#cb5-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">workflow.type =</span> <span class="st">"STAR - Counts"</span>,</span>
<span id="cb5-7"><a href="#cb5-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">experimental.strategy =</span> <span class="st">"RNA-Seq"</span>,</span>
<span id="cb5-8"><a href="#cb5-8" aria-hidden="true" tabindex="-1"></a>  <span class="at">sample.type =</span> <span class="fu">c</span>(<span class="st">"Primary Tumor"</span>)</span>
<span id="cb5-9"><a href="#cb5-9" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb5-10"><a href="#cb5-10" aria-hidden="true" tabindex="-1"></a>TCGAbiolinks<span class="sc">::</span><span class="fu">GDCdownload</span>(<span class="at">query =</span> query, <span class="at">method =</span> <span class="st">"api"</span>)</span>
<span id="cb5-11"><a href="#cb5-11" aria-hidden="true" tabindex="-1"></a>dat <span class="ot">&lt;-</span> TCGAbiolinks<span class="sc">::</span><span class="fu">GDCprepare</span>(<span class="at">query =</span> query)</span>
<span id="cb5-12"><a href="#cb5-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-13"><a href="#cb5-13" aria-hidden="true" tabindex="-1"></a>SummarizedExperiment<span class="sc">::</span><span class="fu">assays</span>(dat)<span class="sc">$</span>unstranded[<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">2</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<pre class="border"><code>                   TCGA-A7-A26E-01B-06R-A277-07 TCGA-A2-A0CU-01A-12R-A034-07
ENSG00000000003.15                          691                         1429
ENSG00000000005.6                            20                           73
ENSG00000000419.13                          335                         1674
ENSG00000000457.14                         1292                         1018
ENSG00000000460.17                          536                          450</code></pre>
<p>It is recommended to use DESeq2 or TMM normalization method for RNA-seq data before further statistical analysis <span class="citation" data-cites="ZhaoY2021">(<a href="#ref-ZhaoY2021" role="doc-biblioref">Y. Zhao et al. 2021</a>)</span>. Here we demonstrate how to use the R/Bioconductor package <a href="https://bioconductor.org/packages/DESeq2/"><strong>DESeq2</strong></a> <span class="citation" data-cites="Love2014">(<a href="#ref-Love2014" role="doc-biblioref">Love, Huber, and Anders 2014</a>)</span> to normalize the RNA count data.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a>meta <span class="ot">&lt;-</span> <span class="fu">colData</span>(dat)[, <span class="fu">c</span>(<span class="st">"project_id"</span>, <span class="st">"submitter_id"</span>, <span class="st">"age_at_diagnosis"</span>, <span class="st">"ethnicity"</span>, <span class="st">"gender"</span>, <span class="st">"days_to_death"</span>, <span class="st">"days_to_last_follow_up"</span>, <span class="st">"vital_status"</span>, <span class="st">"paper_BRCA_Subtype_PAM50"</span>, <span class="st">"treatments"</span>)]</span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a>meta<span class="sc">$</span>treatments <span class="ot">&lt;-</span> <span class="fu">unlist</span>(<span class="fu">lapply</span>(meta<span class="sc">$</span>treatments, <span class="cf">function</span>(xx) {</span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">any</span>(xx<span class="sc">$</span>treatment_or_therapy <span class="sc">==</span> <span class="st">"yes"</span>)</span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true" tabindex="-1"></a>}))</span>
<span id="cb7-5"><a href="#cb7-5" aria-hidden="true" tabindex="-1"></a>dds <span class="ot">&lt;-</span> DESeq2<span class="sc">::</span><span class="fu">DESeqDataSetFromMatrix</span>(<span class="fu">assays</span>(dat)<span class="sc">$</span>unstranded, <span class="at">colData =</span> meta, <span class="at">design =</span> <span class="sc">~</span><span class="dv">1</span>)</span>
<span id="cb7-6"><a href="#cb7-6" aria-hidden="true" tabindex="-1"></a>dds2 <span class="ot">&lt;-</span> DESeq2<span class="sc">::</span><span class="fu">estimateSizeFactors</span>(dds)</span>
<span id="cb7-7"><a href="#cb7-7" aria-hidden="true" tabindex="-1"></a>RNA_count <span class="ot">&lt;-</span> DESeq2<span class="sc">::</span><span class="fu">counts</span>(dds2, <span class="at">normalized =</span> <span class="cn">TRUE</span>)</span>
<span id="cb7-8"><a href="#cb7-8" aria-hidden="true" tabindex="-1"></a>RNA_count[<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">2</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<pre class="border"><code>                   TCGA-A7-A26E-01B-06R-A277-07 TCGA-A2-A0CU-01A-12R-A034-07
ENSG00000000003.15                   1899.76848                   1419.51789
ENSG00000000005.6                      54.98606                     72.51561
ENSG00000000419.13                    921.01656                   1662.89219
ENSG00000000457.14                   3552.09968                   1011.24507
ENSG00000000460.17                   1473.62649                    447.01403</code></pre>
<p>To perform survival analysis with both clinical/demographic variables and omics data, in the following code we extract female breast cancer patients with their corresponding survival outcomes, clinical/demographic variables and RNA-seq features.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a>meta<span class="sc">$</span>time <span class="ot">&lt;-</span> <span class="fu">apply</span>(meta[, <span class="fu">c</span>(<span class="st">"days_to_death"</span>, <span class="st">"days_to_last_follow_up"</span>)], <span class="dv">1</span>, max, <span class="at">na.rm =</span> <span class="cn">TRUE</span>) <span class="sc">/</span> <span class="fl">365.25</span></span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a>meta<span class="sc">$</span>status <span class="ot">&lt;-</span> meta<span class="sc">$</span>vital_status</span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a>meta<span class="sc">$</span>age <span class="ot">&lt;-</span> meta<span class="sc">$</span>age_at_diagnosis <span class="sc">/</span> <span class="fl">365.25</span></span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true" tabindex="-1"></a>clin <span class="ot">&lt;-</span> <span class="fu">subset</span>(meta, gender <span class="sc">==</span> <span class="st">"female"</span> <span class="sc">&amp;</span> <span class="sc">!</span><span class="fu">duplicated</span>(submitter_id) <span class="sc">&amp;</span> time <span class="sc">&gt;</span> <span class="dv">0</span> <span class="sc">&amp;</span> <span class="sc">!</span><span class="fu">is.na</span>(age))</span>
<span id="cb9-5"><a href="#cb9-5" aria-hidden="true" tabindex="-1"></a>clin <span class="ot">&lt;-</span> clin[<span class="fu">order</span>(clin<span class="sc">$</span>submitter_id), ]</span>
<span id="cb9-6"><a href="#cb9-6" aria-hidden="true" tabindex="-1"></a>RNA_count <span class="ot">&lt;-</span> RNA_count[, <span class="fu">rownames</span>(clin)]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<ul>
<li><p><a href="https://bioconductor.org/packages/release/bioc/html/TCGAbiolinks.html">Bioconductor</a> might provide an outdated version of <strong>TCGAbiolinks</strong>. Here, we use the GitHub version TCGAbiolinks_2.29.6. If you encounter some issues when using this tutorial, please check your installed <strong>TCGAbiolinks</strong> version. If necessary, you can re-install the package from its <a href="https://github.com/BioinformaticsFMRP/TCGAbiolinks.git">GitHub repository</a>. Otherwise, download the data from <a href="https://doi.org/10.5281/zenodo.10044741"><img src="https://zenodo.org/badge/DOI/10.5281/zenodo.10044741.svg" class="img-fluid" alt="DOI"></a> and load the <code>dat</code> object with: <code>load("TCGA_data.rda")</code>.</p></li>
<li><p>The package <strong>TCGAbiolinks</strong> cannot retrieve any proteomics or metabolomics data. It is always useful to look at your data first, in particular the data type and dimensions (i.e.&nbsp;numbers of rows and columns for a data frame or matrix).</p></li>
</ul>
</div>
</div>
<p><br></p>
</section>
<section id="survival-analysis-with-low-dimensional-input-data" class="level2 unnumbered">
<h2 class="unnumbered anchored" data-anchor-id="survival-analysis-with-low-dimensional-input-data">Survival analysis with low-dimensional input data</h2>
<section id="nonparametric-survival-analysis" class="level3 unnumbered">
<h3 class="unnumbered anchored" data-anchor-id="nonparametric-survival-analysis">Nonparametric survival analysis</h3>
<p>For the data of TCGA breast cancer patients that we extracted in the previous section, Kaplan-Meier estimates of the survival probabilities can be obtained via function <code>survfit()</code> from <a href="https://CRAN.R-project.org/package=survival"><strong>survival</strong></a> package. The dashed lines in the following figure indicate the median survival time.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Kaplan-Meier (KM) estimation</span></span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a>clin<span class="sc">$</span>status[clin<span class="sc">$</span>status <span class="sc">==</span> <span class="st">"Dead"</span>] <span class="ot">&lt;-</span> <span class="dv">1</span></span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a>clin<span class="sc">$</span>status[clin<span class="sc">$</span>status <span class="sc">==</span> <span class="st">"Alive"</span>] <span class="ot">&lt;-</span> <span class="dv">0</span></span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a>clin<span class="sc">$</span>status <span class="ot">&lt;-</span> <span class="fu">as.numeric</span>(clin<span class="sc">$</span>status)</span>
<span id="cb10-5"><a href="#cb10-5" aria-hidden="true" tabindex="-1"></a>sfit <span class="ot">&lt;-</span> survival<span class="sc">::</span><span class="fu">survfit</span>(<span class="fu">Surv</span>(time, status) <span class="sc">~</span> <span class="dv">1</span>, <span class="at">data =</span> clin)</span>
<span id="cb10-6"><a href="#cb10-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-7"><a href="#cb10-7" aria-hidden="true" tabindex="-1"></a><span class="co"># calculate survival probability at 1-, 3- and 5-year time points</span></span>
<span id="cb10-8"><a href="#cb10-8" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(sfit, <span class="at">times =</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">3</span>, <span class="dv">5</span>))</span>
<span id="cb10-9"><a href="#cb10-9" aria-hidden="true" tabindex="-1"></a><span class="fu">theme_set</span>(<span class="fu">theme_bw</span>())</span>
<span id="cb10-10"><a href="#cb10-10" aria-hidden="true" tabindex="-1"></a>ggsurv <span class="ot">&lt;-</span> survminer<span class="sc">::</span><span class="fu">ggsurvplot</span>(sfit,</span>
<span id="cb10-11"><a href="#cb10-11" aria-hidden="true" tabindex="-1"></a>  <span class="at">conf.int =</span> <span class="cn">TRUE</span>, <span class="at">risk.table =</span> <span class="cn">TRUE</span>,</span>
<span id="cb10-12"><a href="#cb10-12" aria-hidden="true" tabindex="-1"></a>  <span class="at">xlab =</span> <span class="st">"Time since diagnosis (year)"</span>,</span>
<span id="cb10-13"><a href="#cb10-13" aria-hidden="true" tabindex="-1"></a>  <span class="at">legend =</span> <span class="st">"none"</span>, <span class="at">surv.median.line =</span> <span class="st">"hv"</span></span>
<span id="cb10-14"><a href="#cb10-14" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb10-15"><a href="#cb10-15" aria-hidden="true" tabindex="-1"></a>ggsurv<span class="sc">$</span>plot <span class="ot">&lt;-</span> ggsurv<span class="sc">$</span>plot <span class="sc">+</span> <span class="fu">annotate</span>(<span class="st">"text"</span>, <span class="at">x =</span> <span class="dv">20</span>, <span class="at">y =</span> <span class="fl">0.9</span>, <span class="at">label =</span> <span class="st">"+  Censor"</span>)</span>
<span id="cb10-16"><a href="#cb10-16" aria-hidden="true" tabindex="-1"></a>ggsurv</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><img src="fig/TCGA_surv_km1.png" class="img-fluid figure-img" style="width:60.0%"></p>
<figcaption class="figure-caption"><em>Kaplan-Meier curve for 1061 BRCA patients data from TCGA.</em></figcaption>
</figure>
</div>
<p><br></p>
<p>To compare the survival curves of two groups of patients, for example, treatment (i.e.&nbsp;pharmaceutical or radiation therapy) or nontreatment, the <code>R</code> function <code>survival::survdiff()</code> can perform the log-rank test to compare two survival curves. Alternatively, the <code>R</code> function <code>survival::survfit</code> with a formula including the treatment group as a covariate can return the (KM) survival probabilities for each groups. Then the <code>R</code> function <code>survminer::ggsurvplot()</code> with a <code>survfit</code> object will draw the two survival curves and perform the log-rank test as shown in the following figure.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a>survival<span class="sc">::</span><span class="fu">survdiff</span>(<span class="fu">Surv</span>(time, status) <span class="sc">~</span> treatments, <span class="at">data =</span> clin)</span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a>sfit2 <span class="ot">&lt;-</span> <span class="fu">survfit</span>(<span class="fu">Surv</span>(time, status) <span class="sc">~</span> treatments, <span class="at">data =</span> clin)</span>
<span id="cb11-4"><a href="#cb11-4" aria-hidden="true" tabindex="-1"></a>ggsurv <span class="ot">&lt;-</span> <span class="fu">ggsurvplot</span>(sfit2,</span>
<span id="cb11-5"><a href="#cb11-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">conf.int =</span> <span class="cn">TRUE</span>, <span class="at">risk.table =</span> <span class="cn">TRUE</span>,</span>
<span id="cb11-6"><a href="#cb11-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">xlab =</span> <span class="st">"Time since diagnosis (year)"</span>, <span class="at">legend =</span> <span class="fu">c</span>(.<span class="dv">6</span>, .<span class="dv">9</span>),</span>
<span id="cb11-7"><a href="#cb11-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">legend.labs =</span> <span class="fu">c</span>(<span class="st">"No"</span>, <span class="st">"Yes"</span>), <span class="at">legend.title =</span> <span class="st">"Treatment"</span>,</span>
<span id="cb11-8"><a href="#cb11-8" aria-hidden="true" tabindex="-1"></a>  <span class="at">risk.table.y.text.col =</span> <span class="cn">TRUE</span>, <span class="at">risk.table.y.text =</span> <span class="cn">FALSE</span></span>
<span id="cb11-9"><a href="#cb11-9" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb11-10"><a href="#cb11-10" aria-hidden="true" tabindex="-1"></a>ggsurv<span class="sc">$</span>plot <span class="ot">&lt;-</span> ggsurv<span class="sc">$</span>plot <span class="sc">+</span></span>
<span id="cb11-11"><a href="#cb11-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">annotate</span>(<span class="st">"text"</span>, <span class="at">x =</span> <span class="dv">21</span>, <span class="at">y =</span> <span class="dv">1</span>, <span class="at">label =</span> <span class="st">"+  Censor"</span>) <span class="sc">+</span></span>
<span id="cb11-12"><a href="#cb11-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">annotate</span>(<span class="st">"text"</span>, <span class="at">x =</span> <span class="dv">22</span>, <span class="at">y =</span> .<span class="dv">88</span>, <span class="at">label =</span> <span class="fu">paste0</span>(<span class="st">"Log-rank test:</span><span class="sc">\n</span><span class="st">"</span>, <span class="fu">surv_pvalue</span>(sfit2)<span class="sc">$</span>pval.txt))</span>
<span id="cb11-13"><a href="#cb11-13" aria-hidden="true" tabindex="-1"></a>ggsurv</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><img src="fig/TCGA_surv_km2.png" class="img-fluid figure-img" style="width:60.0%"></p>
<figcaption class="figure-caption"><em>Kaplan-Meier curves of the BRCA patients’ survival data from TCGA grouped by treatment (i.e.&nbsp;pharmaceutical or radiation therapy) or nontreatment. The log-rank test is to compare the two survival distributions corresponding to the two groups of patients.</em></figcaption>
</figure>
</div>
<p><br></p>
<p>To analyze if a continuous variable, e.g.&nbsp;age, is associated with the survival outcomes, we can use the <code>R</code> function <code>coxph()</code> for fitting a Cox model, which is similar to the function <code>lm()</code> for fitting linear models.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a>fit_cox <span class="ot">&lt;-</span> <span class="fu">coxph</span>(<span class="fu">Surv</span>(time, status) <span class="sc">~</span> age, <span class="at">data =</span> clin)</span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(fit_cox)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<pre class="border"><code>Call:
coxph(formula = Surv(time, status) ~ age, data = clin)

  n= 1047, number of events= 149 
   (14 observations deleted due to missingness)

        coef exp(coef) se(coef)     z Pr(&gt;|z|)    
age 0.034244  1.034837 0.006703 5.109 3.24e-07 ***
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

    exp(coef) exp(-coef) lower .95 upper .95
age     1.035     0.9663     1.021     1.049

Concordance= 0.639  (se = 0.029 )
Likelihood ratio test= 26.34  on 1 df,   p=3e-07
Wald test            = 26.1  on 1 df,   p=3e-07
Score (logrank) test = 26.63  on 1 df,   p=2e-07</code></pre>
<p>The Cox model assumes proportional hazards and log-linearity of the covariates. To check the log-linearity for a clinical or demographic variable, e.g.&nbsp;age, we can fit a penalized smoothing spline for age effect. The following code shows that the nonlinear part of the smoothing spline has a significant effect (<span class="math inline">\(p = 0.00013\)</span>). Thus, the assumption of log-linearity for age is not satisfied.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a>fit_cox_spline <span class="ot">&lt;-</span> <span class="fu">coxph</span>(<span class="fu">Surv</span>(time, status) <span class="sc">~</span> <span class="fu">pspline</span>(age), <span class="at">data =</span> clin)</span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a>fit_cox_spline</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<pre class="border"><code>Call:
coxph(formula = Surv(time, status) ~ pspline(age), data = clin)

                         coef se(coef)      se2    Chisq   DF       p
pspline(age), linear  0.03509  0.00577  0.00577 36.98323 1.00 1.2e-09
pspline(age), nonlin                            20.69146 3.03 0.00013

Iterations: 5 outer, 15 Newton-Raphson
     Theta= 0.828 
Degrees of freedom for terms= 4 
Likelihood ratio test=46.4  on 4.03 df, p=2e-09
n= 1047, number of events= 149 </code></pre>
<p>To check proportional hazards of age, we can add a time-dependent covariate <span class="math inline">\(age \times g(t)\)</span>, where <span class="math inline">\(g(t)\)</span> is a known function e.g.&nbsp;<span class="math inline">\(g(t) = \log t\)</span>. The following code shows that the time-dependent age is significant using a score test (<span class="math inline">\(p = 0.0087\)</span>). Thus, the assumption of proportional hazards for age is not satisfied. The above two tests indicate a non-loglinear or time-dependent association of age with the survival outcomes.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a>survival<span class="sc">::</span><span class="fu">cox.zph</span>(fit_cox, <span class="at">transform =</span> <span class="st">"log"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<pre class="border"><code>       chisq df    p
age     6.88  1 0.0087
GLOBAL  6.88  1 0.0087</code></pre>
<div class="info-box note">
<p>Here the approaches for checking log-linearity or proportional hazards can only be used in low-dimensional data settings. When including high-dimensional omics data, there are no standard approaches for checking log-linearity or proportional hazards currently.</p>
</div>
<p><br></p>
</section>
<section id="feature-preselectionfiltering" class="level3 unnumbered">
<h3 class="unnumbered anchored" data-anchor-id="feature-preselectionfiltering">Feature preselection/filtering</h3>
<p>From a practical point of view, since most omics profiles contain thousands of variables and most supervised statistical methods are not suited for high dimensional omics features, it is better to filter the omics features first. In addition, we perceive that not too many omics features are relevant to one medical problem. We will demonstrate <strong>three different filtering approaches for high-dimensional omics data</strong>:</p>
<ul>
<li>Knowledge-based filtering</li>
<li>P-value-based filtering</li>
<li>Variance-based filtering</li>
</ul>
<section id="knowledge-filter" class="level4 unnumbered">
<h4 class="unnumbered anchored" data-anchor-id="knowledge-filter">Knowledge filter</h4>
<p>One can be interested in only some biologically meaningful genes or only protein-coding genes in a specific study. For example, the code below filters protein-coding genes.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a>filtered_rna <span class="ot">&lt;-</span> RNA_count[<span class="fu">rowData</span>(dat)<span class="sc">$</span>gene_type <span class="sc">==</span> <span class="st">"protein_coding"</span>, ]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="p-value-filter" class="level4 unnumbered">
<h4 class="unnumbered anchored" data-anchor-id="p-value-filter">P-value filter</h4>
<p>Before joint analyzing the associations between the thousands of omics features and survival outcomes, one can analyze the association between each omics feature and the survival outcomes, and filter omics features at a statistical significance level <span class="math inline">\(0.1\)</span> or <span class="math inline">\(0.2\)</span> (larger than 0.05 to reduce false negative identification of omics features in multivariate analysis). For demonstration, based on the <span class="math inline">\(100\)</span> mRNA-Seq features from TCGA breast cancer patients preprocessed previously, the code below filters omics features at the statistical significance level <span class="math inline">\(0.2\)</span>, i.e.&nbsp;<span class="math inline">\(p &lt; 0.2\)</span>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a>RNA_log2count <span class="ot">&lt;-</span> <span class="fu">log2</span>(RNA_count[<span class="dv">1</span><span class="sc">:</span><span class="dv">100</span>, ] <span class="sc">+</span> <span class="dv">1</span>)</span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a>pvalues <span class="ot">&lt;-</span> <span class="fu">rep</span>(<span class="cn">NA</span>, <span class="fu">nrow</span>(RNA_log2count))</span>
<span id="cb19-3"><a href="#cb19-3" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (j <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">nrow</span>(RNA_log2count)) {</span>
<span id="cb19-4"><a href="#cb19-4" aria-hidden="true" tabindex="-1"></a>  fit_cox <span class="ot">&lt;-</span> <span class="fu">coxph</span>(<span class="fu">Surv</span>(clin<span class="sc">$</span>time, clin<span class="sc">$</span>status) <span class="sc">~</span> RNA_log2count[j, ], <span class="at">data =</span> clin)</span>
<span id="cb19-5"><a href="#cb19-5" aria-hidden="true" tabindex="-1"></a>  pvalues[j] <span class="ot">&lt;-</span> <span class="fu">summary</span>(fit_cox)<span class="sc">$</span>coefficients[, <span class="st">"Pr(&gt;|z|)"</span>]</span>
<span id="cb19-6"><a href="#cb19-6" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb19-7"><a href="#cb19-7" aria-hidden="true" tabindex="-1"></a>filtered_rna <span class="ot">&lt;-</span> RNA_log2count[<span class="fu">which</span>(pvalues <span class="sc">&lt;</span> <span class="fl">0.2</span>), ]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="variance-filter" class="level4 unnumbered">
<h4 class="unnumbered anchored" data-anchor-id="variance-filter">Variance filter</h4>
<p>The other common and easy way to decrease the number of omics features is to filter the most variable ones for further analysis. Note that the variance-based filtering step should be done before data standardization (i.e.&nbsp;calculating <span class="math inline">\(z\)</span>-score), but can be performed after count data normalization and log2-transformation for instance.</p>
<p>The <code>R</code> package <a href="https://bioconductor.org/packages/M3C/"><strong>M3C</strong></a> <span class="citation" data-cites="John2020">(<a href="#ref-John2020" role="doc-biblioref">John et al. 2020</a>)</span> provides a filter function <code>featurefilter()</code> by using different variance-type metrics, for example, variance, median absolute deviation (MAD), coefficient of variation (A) and its second order derivative (A2). The simple variance filter can be used if the variance does not change with the corresponding mean, otherwise the coefficient of variation can be used. If the omics data include outliers, MAD filter is more robust than the variance filter. Based on the <span class="math inline">\(60660\)</span> mRNA-Seq features from TCGA breast cancer patients preprocessed previously, the code below extracts the <span class="math inline">\(1\%\)</span> most variable features using variance as a filtering metric.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a>RNA_log2count <span class="ot">&lt;-</span> <span class="fu">log2</span>(RNA_count <span class="sc">+</span> <span class="dv">1</span>)</span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a>filtered <span class="ot">&lt;-</span> M3C<span class="sc">::</span><span class="fu">featurefilter</span>(RNA_log2count, <span class="at">percentile =</span> <span class="dv">1</span>, <span class="at">method =</span> <span class="st">"var"</span>, <span class="at">topN =</span> <span class="dv">5</span>)</span>
<span id="cb20-3"><a href="#cb20-3" aria-hidden="true" tabindex="-1"></a>filtered_rna1 <span class="ot">&lt;-</span> filtered<span class="sc">$</span>filtered_data</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<pre class="border"><code>***feature filter function***
extracting the most variable: 1 percent
features to start with: 60660
performing calculations for variance
printing topN most variable features with statistics...
                              feature      mean      var       sd
ENSG00000166509.12 ENSG00000166509.12  6.086125 31.60384 5.621729
ENSG00000110484.7   ENSG00000110484.7 11.005136 26.13755 5.112489
ENSG00000153002.12 ENSG00000153002.12  8.212895 25.89105 5.088325
ENSG00000134184.13 ENSG00000134184.13  5.371435 23.23511 4.820281
ENSG00000160182.3   ENSG00000160182.3  9.902195 21.41407 4.627534
features remaining: 607</code></pre>
<p>Another variance-type filter is to remain features with certain percentage of <strong>cumulative variances</strong>, which will usually filter fewer features than the approaches above. The code below extracts the most variable features explaining <span class="math inline">\(1\%\)</span> <strong>cumulative variances</strong>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a>cumsum_var <span class="ot">&lt;-</span> <span class="fu">cumsum</span>(filtered<span class="sc">$</span>statistics<span class="sc">$</span>var)</span>
<span id="cb22-2"><a href="#cb22-2" aria-hidden="true" tabindex="-1"></a>cumsum_cutoff <span class="ot">&lt;-</span> cumsum_var[<span class="fu">length</span>(cumsum_var)] <span class="sc">*</span> <span class="fl">0.01</span></span>
<span id="cb22-3"><a href="#cb22-3" aria-hidden="true" tabindex="-1"></a>filtered_names <span class="ot">&lt;-</span> filtered<span class="sc">$</span>statistics<span class="sc">$</span>feature[cumsum_var <span class="sc">&lt;</span> cumsum_cutoff]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><br></p>
</section>
</section>
</section>
<section id="survival-analysis-with-high-dimensional-input-data" class="level2 unnumbered">
<h2 class="unnumbered anchored" data-anchor-id="survival-analysis-with-high-dimensional-input-data">Survival analysis with high-dimensional input data</h2>
<section id="unsupervised-learning-omics-data" class="level3 unnumbered">
<h3 class="unnumbered anchored" data-anchor-id="unsupervised-learning-omics-data">Unsupervised learning (omics data)</h3>
<p>In this section we will use the mRNA-Seq data of breast cancer patients from TCGA. The following unsupervised methods can be applied to other omics data as well (the same applies to the supervised learning methods). One important thing is that the input omics data, especially the data type and dimensions, should be suited to the methods.</p>
<p>Unsupervised learning for omics data can be helpful to explore subpopulations of the data, for example, patients from one cancer type can be divided to several omics-related subtypes. We demonstrate three unsupervised learning methods, i.e.&nbsp;principal component analysis (PCA), <span class="math inline">\(t\)</span>-stochastic neighbour embedding (<span class="math inline">\(t\)</span>-SNE) and uniform manifold approximation and projection (UMAP), based on the PAM50 genes <span class="citation" data-cites="Parker2009">(<a href="#ref-Parker2009" role="doc-biblioref">Parker et al. 2009</a>)</span>. The <code>R</code> package <a href="https://bioconductor.org/packages/M3C/"><strong>M3C</strong></a> <span class="citation" data-cites="John2020">(<a href="#ref-John2020" role="doc-biblioref">John et al. 2020</a>)</span> provides the analyses and visualization of all the three methods.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a><span class="co"># identify indexes of the PAM50 genes in the TCGA-BRCA data</span></span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true" tabindex="-1"></a>idx <span class="ot">&lt;-</span> <span class="fu">which</span>(<span class="fu">rowData</span>(dat)<span class="sc">$</span>gene_name <span class="sc">%in%</span></span>
<span id="cb23-3"><a href="#cb23-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">c</span>(<span class="st">"UBE2T"</span>, <span class="st">"BIRC5"</span>, <span class="st">"NUF2"</span>, <span class="st">"CDC6"</span>, <span class="st">"CCNB1"</span>, <span class="st">"TYMS"</span>, <span class="st">"MYBL2"</span>, <span class="st">"CEP55"</span>, <span class="st">"MELK"</span>, <span class="st">"NDC80"</span>, <span class="st">"RRM2"</span>, <span class="st">"UBE2C"</span>, <span class="st">"CENPF"</span>, <span class="st">"PTTG1"</span>, <span class="st">"EXO1"</span>, <span class="st">"ORC6"</span>, <span class="st">"ANLN"</span>, <span class="st">"CCNE1"</span>, <span class="st">"CDC20"</span>, <span class="st">"MKI67"</span>, <span class="st">"KIF2C"</span>, <span class="st">"ACTR3B"</span>, <span class="st">"MYC"</span>, <span class="st">"EGFR"</span>, <span class="st">"KRT5"</span>, <span class="st">"PHGDH"</span>, <span class="st">"CDH3"</span>, <span class="st">"MIA"</span>, <span class="st">"KRT17"</span>, <span class="st">"FOXC1"</span>, <span class="st">"SFRP1"</span>, <span class="st">"KRT14"</span>, <span class="st">"ESR1"</span>, <span class="st">"SLC39A6"</span>, <span class="st">"BAG1"</span>, <span class="st">"MAPT"</span>, <span class="st">"PGR"</span>, <span class="st">"CXXC5"</span>, <span class="st">"MLPH"</span>, <span class="st">"BCL2"</span>, <span class="st">"MDM2"</span>, <span class="st">"NAT1"</span>, <span class="st">"FOXA1"</span>, <span class="st">"BLVRA"</span>, <span class="st">"MMP11"</span>, <span class="st">"GPR160"</span>, <span class="st">"FGFR4"</span>, <span class="st">"GRB7"</span>, <span class="st">"TMEM45B"</span>, <span class="st">"ERBB2"</span>))</span>
<span id="cb23-4"><a href="#cb23-4" aria-hidden="true" tabindex="-1"></a><span class="co"># extract the PAM50 genes of TCGA-BRCA patients</span></span>
<span id="cb23-5"><a href="#cb23-5" aria-hidden="true" tabindex="-1"></a>TCGA_PAM50 <span class="ot">&lt;-</span> RNA_count[idx, ]</span>
<span id="cb23-6"><a href="#cb23-6" aria-hidden="true" tabindex="-1"></a><span class="co"># use gene symbols instead of Ensembl IDs</span></span>
<span id="cb23-7"><a href="#cb23-7" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(TCGA_PAM50) <span class="ot">&lt;-</span> <span class="fu">rowData</span>(dat)<span class="sc">$</span>gene_name[idx]</span>
<span id="cb23-8"><a href="#cb23-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-9"><a href="#cb23-9" aria-hidden="true" tabindex="-1"></a><span class="co"># log2-transformation of the normalized count data</span></span>
<span id="cb23-10"><a href="#cb23-10" aria-hidden="true" tabindex="-1"></a>TCGA_PAM50 <span class="ot">&lt;-</span> <span class="fu">log2</span>(TCGA_PAM50 <span class="sc">+</span> <span class="dv">1</span>)</span>
<span id="cb23-11"><a href="#cb23-11" aria-hidden="true" tabindex="-1"></a>pam50 <span class="ot">&lt;-</span> <span class="fu">factor</span>(clin<span class="sc">$</span>paper_BRCA_Subtype_PAM50)</span>
<span id="cb23-12"><a href="#cb23-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-13"><a href="#cb23-13" aria-hidden="true" tabindex="-1"></a>M3C<span class="sc">::</span><span class="fu">pca</span>(TCGA_PAM50, <span class="at">labels =</span> pam50, <span class="at">dotsize =</span> <span class="dv">3</span>, <span class="at">legendtitle =</span> <span class="st">"Subtype"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><img src="fig/TCGA_pca.png" class="img-fluid figure-img" style="width:50.0%"></p>
<figcaption class="figure-caption"><em>Unsupervised clustering (principal component analysis, PCA) of transcriptomic data from TCGA breast cancer patients</em></figcaption>
</figure>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a>M3C<span class="sc">::</span><span class="fu">tsne</span>(TCGA_PAM50, <span class="at">labels =</span> pam50, <span class="at">dotsize =</span> <span class="dv">3</span>, <span class="at">legendtitle =</span> <span class="st">"Subtype"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><img src="fig/TCGA_tsne.png" class="img-fluid figure-img" style="width:50.0%"></p>
<figcaption class="figure-caption"><em>Unsupervised clustering (<span class="math inline">\(t\)</span>-stochastic neighbour embedding, <span class="math inline">\(t\)</span>-SNE) of transcriptomic data from TCGA breast cancer patients</em></figcaption>
</figure>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a>M3C<span class="sc">::</span><span class="fu">umap</span>(TCGA_PAM50, <span class="at">labels =</span> pam50, <span class="at">dotsize =</span> <span class="dv">3</span>, <span class="at">legendtitle =</span> <span class="st">"Subtype"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><img src="fig/TCGA_umap.png" class="img-fluid figure-img" style="width:50.0%"></p>
<figcaption class="figure-caption"><em>Unsupervised clustering (uniform manifold approximation and projection, UMAP) of transcriptomic data from TCGA breast cancer patients</em></figcaption>
</figure>
</div>
<p><br></p>
</section>
<section id="supervised-learning-omics-and-survival-data" class="level3 unnumbered">
<h3 class="unnumbered anchored" data-anchor-id="supervised-learning-omics-and-survival-data">Supervised learning (omics and survival data)</h3>
<p>To investigate the relationship between omics features and survival outcomes, regression methods (i.e.&nbsp;supervised learning) can be applied. Since omics data are high-dimensional, one can use unsupervised learning methods to summarize a few components (dimension reduction) and regress the survival outcomes on the low-dimensional components by some classical statistical methods, e.g.&nbsp;classical Cox model. There are also frequentist and Bayesian supervised learning methods suited to directly regress the survival outcomes on the high-dimensional omics features. Note that preselecting/filtering ultrahigh-dimensional omics features can be useful before running the frequentist and Bayesian supervised learning methods.</p>
<section id="dimension-reduction-for-cox-models" class="level4 unnumbered">
<h4 class="unnumbered anchored" data-anchor-id="dimension-reduction-for-cox-models">Dimension reduction for Cox models</h4>
<p>The following code demonstrates the use of the first two principal components of PCA as covariates for the <strong>purpose of survival prediction</strong>. Similarly, the first components from <span class="math inline">\(t\)</span>-SNE or UMAP can also be extracted as covariates.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a><span class="co"># principal component regression</span></span>
<span id="cb26-2"><a href="#cb26-2" aria-hidden="true" tabindex="-1"></a>x_tmp <span class="ot">&lt;-</span> <span class="fu">prcomp</span>(<span class="fu">t</span>(TCGA_PAM50))</span>
<span id="cb26-3"><a href="#cb26-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-4"><a href="#cb26-4" aria-hidden="true" tabindex="-1"></a><span class="co"># choose the top two components (subjective) as covariates</span></span>
<span id="cb26-5"><a href="#cb26-5" aria-hidden="true" tabindex="-1"></a>X_PC <span class="ot">&lt;-</span> x_tmp<span class="sc">$</span>x[, <span class="dv">1</span><span class="sc">:</span><span class="dv">2</span>]</span>
<span id="cb26-6"><a href="#cb26-6" aria-hidden="true" tabindex="-1"></a><span class="co"># build classical survival model (e.g. PH Cox model)</span></span>
<span id="cb26-7"><a href="#cb26-7" aria-hidden="true" tabindex="-1"></a>data_tmp <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">time =</span> clin<span class="sc">$</span>time, <span class="at">status =</span> clin<span class="sc">$</span>status, X_PC)</span>
<span id="cb26-8"><a href="#cb26-8" aria-hidden="true" tabindex="-1"></a>fit <span class="ot">&lt;-</span> <span class="fu">coxph</span>(<span class="fu">Surv</span>(time, status) <span class="sc">~</span> PC1 <span class="sc">+</span> PC2, <span class="at">data =</span> data_tmp)</span>
<span id="cb26-9"><a href="#cb26-9" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(fit)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<pre class="border"><code>Call:
coxph(formula = Surv(time, status) ~ PC1 + PC2, data = data_tmp)

  n= 1047, number of events= 149 

        coef exp(coef) se(coef)     z Pr(&gt;|z|)   
PC1 0.004679  1.004690 0.009675 0.484  0.62862   
PC2 0.038179  1.038918 0.013233 2.885  0.00391 **
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

    exp(coef) exp(-coef) lower .95 upper .95
PC1     1.005     0.9953    0.9858     1.024
PC2     1.039     0.9625    1.0123     1.066

Concordance= 0.58  (se = 0.028 )
Likelihood ratio test= 8.54  on 2 df,   p=0.01
Wald test            = 8.64  on 2 df,   p=0.01
Score (logrank) test = 8.66  on 2 df,   p=0.01</code></pre>
</section>
<section id="penalized-cox-models" class="level4 unnumbered">
<h4 class="unnumbered anchored" data-anchor-id="penalized-cox-models">Penalized Cox models</h4>
<p>For computational efficiency, we will use only the mRNA-Seq features corresponding to the PAM50 genes <span class="citation" data-cites="Parker2009">(<a href="#ref-Parker2009" role="doc-biblioref">Parker et al. 2009</a>)</span> instead of the variance filtered genes from the previous section. We perform an investigation of the relationships between the mRNA-Seq features, two clinical variables (i.e.&nbsp;the patients’ age at diagnosis and their ethnicity) and the survival outcomes.</p>
<p>The <code>R</code> package <a href="https://CRAN.R-project.org/package=glmnet"><strong>glmnet</strong></a> <span class="citation" data-cites="Friedman2010">(<a href="#ref-Friedman2010" role="doc-biblioref">Friedman, Hastie, and Tibshirani 2010</a>)</span> is very computationally efficient to run Lasso and Elastic Net Cox models. Lasso has a tuning parameter <span class="math inline">\(\lambda\)</span> to control the penalty strength of the coefficients which can be optimized by cross-validation (CV) via function <code>cv.glmnet()</code>. The <code>glmnet()</code> and <code>cv.glmnet()</code> functions provide the argument <code>penalty.factor</code> to allow different shrinkage for different features, which makes sense if one includes both clinical/demographic variables and omics features and does not want to perform feature selection for the clinical/demographic variables.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb28"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a><span class="do">## Lasso Cox model</span></span>
<span id="cb28-2"><a href="#cb28-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb28-3"><a href="#cb28-3" aria-hidden="true" tabindex="-1"></a><span class="do">## for demonstration simplicity, PAM50 genes are used here</span></span>
<span id="cb28-4"><a href="#cb28-4" aria-hidden="true" tabindex="-1"></a>x <span class="ot">&lt;-</span> <span class="fu">cbind</span>(<span class="at">age =</span> clin<span class="sc">$</span>age, <span class="at">ethnicity =</span> <span class="fu">factor</span>(clin<span class="sc">$</span>ethnicity), <span class="fu">t</span>(TCGA_PAM50))</span>
<span id="cb28-5"><a href="#cb28-5" aria-hidden="true" tabindex="-1"></a>y <span class="ot">&lt;-</span> <span class="fu">cbind</span>(<span class="at">time =</span> clin<span class="sc">$</span>time, <span class="at">status =</span> clin<span class="sc">$</span>status)</span>
<span id="cb28-6"><a href="#cb28-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb28-7"><a href="#cb28-7" aria-hidden="true" tabindex="-1"></a><span class="co"># set penalty factor without penalizing the two demographical variables</span></span>
<span id="cb28-8"><a href="#cb28-8" aria-hidden="true" tabindex="-1"></a>pf <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="fu">rep</span>(<span class="dv">0</span>, <span class="dv">2</span>), <span class="fu">rep</span>(<span class="dv">1</span>, <span class="fu">ncol</span>(x) <span class="sc">-</span> <span class="dv">2</span>))</span>
<span id="cb28-9"><a href="#cb28-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb28-10"><a href="#cb28-10" aria-hidden="true" tabindex="-1"></a><span class="co"># Lasso Cox by using cv.glmnet to obtain the 5-fold CV optimal lambda.min or lambda.1se</span></span>
<span id="cb28-11"><a href="#cb28-11" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">123</span>)</span>
<span id="cb28-12"><a href="#cb28-12" aria-hidden="true" tabindex="-1"></a>cvfit <span class="ot">&lt;-</span> glmnet<span class="sc">::</span><span class="fu">cv.glmnet</span>(x, y, <span class="at">family =</span> <span class="st">"cox"</span>, <span class="at">nfolds =</span> <span class="dv">5</span>, <span class="at">penalty.factor =</span> pf)</span>
<span id="cb28-13"><a href="#cb28-13" aria-hidden="true" tabindex="-1"></a>mod <span class="ot">&lt;-</span> cvfit<span class="sc">$</span>glmnet.fit</span>
<span id="cb28-14"><a href="#cb28-14" aria-hidden="true" tabindex="-1"></a>lambda_optimal <span class="ot">&lt;-</span> cvfit<span class="sc">$</span>lambda.min <span class="co"># optimal lambda</span></span>
<span id="cb28-15"><a href="#cb28-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb28-16"><a href="#cb28-16" aria-hidden="true" tabindex="-1"></a>betas <span class="ot">&lt;-</span> <span class="fu">as.vector</span>(<span class="fu">coef</span>(mod, <span class="at">s =</span> lambda_optimal))</span>
<span id="cb28-17"><a href="#cb28-17" aria-hidden="true" tabindex="-1"></a>beta.positive <span class="ot">&lt;-</span> <span class="fu">colnames</span>(x)[betas <span class="sc">&gt;</span> <span class="dv">0</span>]</span>
<span id="cb28-18"><a href="#cb28-18" aria-hidden="true" tabindex="-1"></a>beta.negative <span class="ot">&lt;-</span> <span class="fu">colnames</span>(x)[betas <span class="sc">&lt;</span> <span class="dv">0</span>]</span>
<span id="cb28-19"><a href="#cb28-19" aria-hidden="true" tabindex="-1"></a><span class="co"># get ordered list of variables as they appear at smallest lambda</span></span>
<span id="cb28-20"><a href="#cb28-20" aria-hidden="true" tabindex="-1"></a>allnames <span class="ot">&lt;-</span> <span class="fu">names</span>(<span class="fu">coef</span>(mod)[, <span class="fu">ncol</span>(<span class="fu">coef</span>(mod))]</span>
<span id="cb28-21"><a href="#cb28-21" aria-hidden="true" tabindex="-1"></a>[<span class="fu">order</span>(<span class="fu">coef</span>(mod)[, <span class="fu">ncol</span>(<span class="fu">coef</span>(mod))], <span class="at">decreasing =</span> <span class="cn">TRUE</span>)])</span>
<span id="cb28-22"><a href="#cb28-22" aria-hidden="true" tabindex="-1"></a><span class="co"># assign colors for positive (pink) and negative (green) coefficients</span></span>
<span id="cb28-23"><a href="#cb28-23" aria-hidden="true" tabindex="-1"></a>cols <span class="ot">&lt;-</span> <span class="fu">rep</span>(<span class="st">"gray80"</span>, <span class="fu">length</span>(allnames))</span>
<span id="cb28-24"><a href="#cb28-24" aria-hidden="true" tabindex="-1"></a>cols[allnames <span class="sc">%in%</span> beta.positive] <span class="ot">&lt;-</span> <span class="st">"seagreen3"</span></span>
<span id="cb28-25"><a href="#cb28-25" aria-hidden="true" tabindex="-1"></a>cols[allnames <span class="sc">%in%</span> beta.negative] <span class="ot">&lt;-</span> <span class="st">"hotpink"</span></span>
<span id="cb28-26"><a href="#cb28-26" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb28-27"><a href="#cb28-27" aria-hidden="true" tabindex="-1"></a><span class="co"># draw coefficient paths of a Lasso Cox model</span></span>
<span id="cb28-28"><a href="#cb28-28" aria-hidden="true" tabindex="-1"></a>plotmo<span class="sc">::</span><span class="fu">plot_glmnet</span>(mod,</span>
<span id="cb28-29"><a href="#cb28-29" aria-hidden="true" tabindex="-1"></a>  <span class="at">label =</span> <span class="cn">TRUE</span>, <span class="at">s =</span> lambda_optimal, <span class="at">col =</span> cols,</span>
<span id="cb28-30"><a href="#cb28-30" aria-hidden="true" tabindex="-1"></a>  <span class="at">xlab =</span> <span class="fu">expression</span>(log <span class="sc">~</span> <span class="er">~</span>lambda), <span class="at">ylab =</span> <span class="fu">expression</span>(beta)</span>
<span id="cb28-31"><a href="#cb28-31" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb28-32"><a href="#cb28-32" aria-hidden="true" tabindex="-1"></a><span class="fu">title</span>(<span class="st">"Lasso    </span><span class="sc">\n\n</span><span class="st">"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><img src="fig/TCGA_lasso.png" class="img-fluid figure-img" style="width:60.0%"></p>
<figcaption class="figure-caption"><em>Coefficient paths of a Lasso Cox model. The verticle gray line indicates the optimal <span class="math inline">\(\lambda\)</span> and its correspondingly selected features are marked as green (positive coefficient) and red (negative coefficient) colors. Note that the demographic variables age and ethnicity were not penalized, so that their coefficient paths did not start from zero in the figure.</em></figcaption>
</figure>
</div>
<p><br></p>
<p>Elastic Net Cox model includes the <span class="math inline">\(\lambda\)</span> and an additional penalty parameter <span class="math inline">\(\alpha \in [0,1]\)</span>. The parameter <span class="math inline">\(\alpha\)</span> can be fixed as <span class="math inline">\(0\)</span> (Ridge), <span class="math inline">\(1\)</span> (Lasso) or any value between <span class="math inline">\(0\)</span> and <span class="math inline">\(1\)</span> for making a compromise between Ridge and Lasso, which can also be optimized by cross-validation manually, see the example below.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb29"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a><span class="do">## Elastic Net Cox model</span></span>
<span id="cb29-2"><a href="#cb29-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb29-3"><a href="#cb29-3" aria-hidden="true" tabindex="-1"></a><span class="co"># set penalty parameter alpha which comprises between Lasso and ridge regressions</span></span>
<span id="cb29-4"><a href="#cb29-4" aria-hidden="true" tabindex="-1"></a>alpha <span class="ot">&lt;-</span> <span class="fu">seq</span>(<span class="fl">0.1</span>, <span class="dv">1</span>, <span class="at">length =</span> <span class="dv">10</span>)</span>
<span id="cb29-5"><a href="#cb29-5" aria-hidden="true" tabindex="-1"></a>fitEN <span class="ot">&lt;-</span> <span class="fu">list</span>()</span>
<span id="cb29-6"><a href="#cb29-6" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">123</span>)</span>
<span id="cb29-7"><a href="#cb29-7" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(alpha)) {</span>
<span id="cb29-8"><a href="#cb29-8" aria-hidden="true" tabindex="-1"></a>  fitEN[[i]] <span class="ot">&lt;-</span> <span class="fu">cv.glmnet</span>(x, y, <span class="at">family =</span> <span class="st">"cox"</span>, <span class="at">alpha =</span> alpha[i], <span class="at">nfolds =</span> <span class="dv">5</span>, <span class="at">penalty.factor =</span> pf)</span>
<span id="cb29-9"><a href="#cb29-9" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb29-10"><a href="#cb29-10" aria-hidden="true" tabindex="-1"></a>idx <span class="ot">&lt;-</span> <span class="fu">which.min</span>(<span class="fu">sapply</span>(fitEN, <span class="cf">function</span>(xx) {</span>
<span id="cb29-11"><a href="#cb29-11" aria-hidden="true" tabindex="-1"></a>  xx<span class="sc">$</span>cvm[xx<span class="sc">$</span>lambda <span class="sc">==</span> xx<span class="sc">$</span>lambda.min]</span>
<span id="cb29-12"><a href="#cb29-12" aria-hidden="true" tabindex="-1"></a>}))</span>
<span id="cb29-13"><a href="#cb29-13" aria-hidden="true" tabindex="-1"></a>cvfit <span class="ot">&lt;-</span> fitEN[[idx]]</span>
<span id="cb29-14"><a href="#cb29-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb29-15"><a href="#cb29-15" aria-hidden="true" tabindex="-1"></a><span class="co"># the following code is the same as Lasso previously</span></span>
<span id="cb29-16"><a href="#cb29-16" aria-hidden="true" tabindex="-1"></a>mod <span class="ot">&lt;-</span> cvfit<span class="sc">$</span>glmnet.fit</span>
<span id="cb29-17"><a href="#cb29-17" aria-hidden="true" tabindex="-1"></a>lambda_optimal <span class="ot">&lt;-</span> cvfit<span class="sc">$</span>lambda.min <span class="co"># optimal lambda</span></span>
<span id="cb29-18"><a href="#cb29-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb29-19"><a href="#cb29-19" aria-hidden="true" tabindex="-1"></a>betas <span class="ot">&lt;-</span> <span class="fu">as.vector</span>(<span class="fu">coef</span>(mod, <span class="at">s =</span> lambda_optimal))</span>
<span id="cb29-20"><a href="#cb29-20" aria-hidden="true" tabindex="-1"></a>beta.positive <span class="ot">&lt;-</span> <span class="fu">colnames</span>(x)[betas <span class="sc">&gt;</span> <span class="dv">0</span>]</span>
<span id="cb29-21"><a href="#cb29-21" aria-hidden="true" tabindex="-1"></a>beta.negative <span class="ot">&lt;-</span> <span class="fu">colnames</span>(x)[betas <span class="sc">&lt;</span> <span class="dv">0</span>]</span>
<span id="cb29-22"><a href="#cb29-22" aria-hidden="true" tabindex="-1"></a>allnames <span class="ot">&lt;-</span> <span class="fu">names</span>(<span class="fu">coef</span>(mod)[, <span class="fu">ncol</span>(<span class="fu">coef</span>(mod))]</span>
<span id="cb29-23"><a href="#cb29-23" aria-hidden="true" tabindex="-1"></a>[<span class="fu">order</span>(<span class="fu">coef</span>(mod)[, <span class="fu">ncol</span>(<span class="fu">coef</span>(mod))], <span class="at">decreasing =</span> <span class="cn">TRUE</span>)])</span>
<span id="cb29-24"><a href="#cb29-24" aria-hidden="true" tabindex="-1"></a>cols <span class="ot">&lt;-</span> <span class="fu">rep</span>(<span class="st">"gray80"</span>, <span class="fu">length</span>(allnames))</span>
<span id="cb29-25"><a href="#cb29-25" aria-hidden="true" tabindex="-1"></a>cols[allnames <span class="sc">%in%</span> beta.positive] <span class="ot">&lt;-</span> <span class="st">"seagreen3"</span></span>
<span id="cb29-26"><a href="#cb29-26" aria-hidden="true" tabindex="-1"></a>cols[allnames <span class="sc">%in%</span> beta.negative] <span class="ot">&lt;-</span> <span class="st">"hotpink"</span></span>
<span id="cb29-27"><a href="#cb29-27" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb29-28"><a href="#cb29-28" aria-hidden="true" tabindex="-1"></a>plotmo<span class="sc">::</span><span class="fu">plot_glmnet</span>(mod,</span>
<span id="cb29-29"><a href="#cb29-29" aria-hidden="true" tabindex="-1"></a>  <span class="at">label =</span> <span class="cn">TRUE</span>, <span class="at">s =</span> lambda_optimal, <span class="at">col =</span> cols,</span>
<span id="cb29-30"><a href="#cb29-30" aria-hidden="true" tabindex="-1"></a>  <span class="at">xlab =</span> <span class="fu">expression</span>(log <span class="sc">~</span> <span class="er">~</span>lambda), <span class="at">ylab =</span> <span class="fu">expression</span>(beta)</span>
<span id="cb29-31"><a href="#cb29-31" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb29-32"><a href="#cb29-32" aria-hidden="true" tabindex="-1"></a><span class="fu">title</span>(<span class="st">"Elastic Net     </span><span class="sc">\n\n</span><span class="st">"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><img src="fig/TCGA_elastic.png" class="img-fluid figure-img" style="width:60.0%"></p>
<figcaption class="figure-caption"><em>Coefficient paths of an Elastic Net Cox model. The verticle gray line indicates the optimal <span class="math inline">\(\lambda\)</span> and its correspondingly selected features are marked as green (positive coefficient) and red (negative coefficient) colors. Note that the demographic variables age and ethnicity were not penalized, so that their coefficient paths did not start from zero in the figure.</em></figcaption>
</figure>
</div>
<p><br></p>
<p>Adaptive Lasso Cox model needs to pre-estimate all coefficients which will be used as weights via the argument <code>penalty.factor</code> in the <code>glmnet()</code> and <code>cv.glmnet()</code> functions to fit a Lasso Cox model. The pre-estimation can be done by a Ridge Cox model, see an example below.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb30"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a><span class="do">## Adaptive Lasso Cox model</span></span>
<span id="cb30-2"><a href="#cb30-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-3"><a href="#cb30-3" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">123</span>)</span>
<span id="cb30-4"><a href="#cb30-4" aria-hidden="true" tabindex="-1"></a>fit <span class="ot">&lt;-</span> <span class="fu">cv.glmnet</span>(x, y, <span class="at">family =</span> <span class="st">"cox"</span>, <span class="at">alpha =</span> <span class="dv">0</span>, <span class="at">nfolds =</span> <span class="dv">5</span>)</span>
<span id="cb30-5"><a href="#cb30-5" aria-hidden="true" tabindex="-1"></a>weights <span class="ot">&lt;-</span> <span class="fu">abs</span>(<span class="dv">1</span> <span class="sc">/</span> <span class="fu">as.vector</span>(<span class="fu">coef</span>(fit, <span class="at">s =</span> <span class="st">"lambda.min"</span>)))</span>
<span id="cb30-6"><a href="#cb30-6" aria-hidden="true" tabindex="-1"></a>weights[<span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">2</span>)] <span class="ot">=</span> <span class="dv">0</span> <span class="co"># don't penalize age and ethnicity</span></span>
<span id="cb30-7"><a href="#cb30-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-8"><a href="#cb30-8" aria-hidden="true" tabindex="-1"></a><span class="co"># adaptive Lasso Cox by using cv.glmnet to obtain the 5-fold CV optimal lambda.min or lambda.1se</span></span>
<span id="cb30-9"><a href="#cb30-9" aria-hidden="true" tabindex="-1"></a>cvfit <span class="ot">&lt;-</span> <span class="fu">cv.glmnet</span>(x, y, <span class="at">family =</span> <span class="st">"cox"</span>, <span class="at">nfolds =</span> <span class="dv">5</span>, <span class="at">penalty.factor =</span> weights)</span>
<span id="cb30-10"><a href="#cb30-10" aria-hidden="true" tabindex="-1"></a>mod <span class="ot">&lt;-</span> cvfit<span class="sc">$</span>glmnet.fit</span>
<span id="cb30-11"><a href="#cb30-11" aria-hidden="true" tabindex="-1"></a>lambda_optimal <span class="ot">&lt;-</span> cvfit<span class="sc">$</span>lambda.min <span class="co"># optimal lambda</span></span>
<span id="cb30-12"><a href="#cb30-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-13"><a href="#cb30-13" aria-hidden="true" tabindex="-1"></a>betas <span class="ot">&lt;-</span> <span class="fu">as.vector</span>(<span class="fu">coef</span>(mod, <span class="at">s =</span> lambda_optimal))</span>
<span id="cb30-14"><a href="#cb30-14" aria-hidden="true" tabindex="-1"></a>beta.positive <span class="ot">&lt;-</span> <span class="fu">colnames</span>(x)[betas <span class="sc">&gt;</span> <span class="dv">0</span>]</span>
<span id="cb30-15"><a href="#cb30-15" aria-hidden="true" tabindex="-1"></a>beta.negative <span class="ot">&lt;-</span> <span class="fu">colnames</span>(x)[betas <span class="sc">&lt;</span> <span class="dv">0</span>]</span>
<span id="cb30-16"><a href="#cb30-16" aria-hidden="true" tabindex="-1"></a><span class="co"># get ordered list of variables as they appear at smallest lambda</span></span>
<span id="cb30-17"><a href="#cb30-17" aria-hidden="true" tabindex="-1"></a>allnames <span class="ot">&lt;-</span> <span class="fu">names</span>(<span class="fu">coef</span>(mod)[, <span class="fu">ncol</span>(<span class="fu">coef</span>(mod))]</span>
<span id="cb30-18"><a href="#cb30-18" aria-hidden="true" tabindex="-1"></a>[<span class="fu">order</span>(<span class="fu">coef</span>(mod)[, <span class="fu">ncol</span>(<span class="fu">coef</span>(mod))], <span class="at">decreasing =</span> <span class="cn">TRUE</span>)])</span>
<span id="cb30-19"><a href="#cb30-19" aria-hidden="true" tabindex="-1"></a><span class="co"># assign colors</span></span>
<span id="cb30-20"><a href="#cb30-20" aria-hidden="true" tabindex="-1"></a>cols <span class="ot">&lt;-</span> <span class="fu">rep</span>(<span class="st">"gray80"</span>, <span class="fu">length</span>(allnames))</span>
<span id="cb30-21"><a href="#cb30-21" aria-hidden="true" tabindex="-1"></a>cols[allnames <span class="sc">%in%</span> beta.positive] <span class="ot">&lt;-</span> <span class="st">"seagreen3"</span></span>
<span id="cb30-22"><a href="#cb30-22" aria-hidden="true" tabindex="-1"></a>cols[allnames <span class="sc">%in%</span> beta.negative] <span class="ot">&lt;-</span> <span class="st">"hotpink"</span></span>
<span id="cb30-23"><a href="#cb30-23" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-24"><a href="#cb30-24" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_glmnet</span>(mod,</span>
<span id="cb30-25"><a href="#cb30-25" aria-hidden="true" tabindex="-1"></a>  <span class="at">label =</span> <span class="cn">TRUE</span>, <span class="at">s =</span> lambda_optimal, <span class="at">col =</span> cols,</span>
<span id="cb30-26"><a href="#cb30-26" aria-hidden="true" tabindex="-1"></a>  <span class="at">xlab =</span> <span class="fu">expression</span>(log <span class="sc">~</span> lambda), <span class="at">ylab =</span> <span class="fu">expression</span>(beta)</span>
<span id="cb30-27"><a href="#cb30-27" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb30-28"><a href="#cb30-28" aria-hidden="true" tabindex="-1"></a><span class="fu">title</span>(<span class="st">"Adative Lasso    </span><span class="sc">\n\n</span><span class="st">"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><img src="fig/TCGA_adaptiveLasso.png" class="img-fluid figure-img" style="width:60.0%"></p>
<figcaption class="figure-caption"><em>Coefficient paths of an adaptive Lasso Cox model. The verticle gray line indicates the optimal <span class="math inline">\(\lambda\)</span> and its correspondingly selected features are marked as green (positive coefficient) and red (negative coefficient) colors. Note that the demographic variables age and ethnicity were not penalized, so that their coefficient paths did not start from zero in the figure.</em></figcaption>
</figure>
</div>
<p><br></p>
<p>Group Lasso Cox model can be implemented through the <code>R</code> package <a href="https://CRAN.R-project.org/package=grpreg"><strong>grpreg</strong></a> <span class="citation" data-cites="Breheny2015">(<a href="#ref-Breheny2015" role="doc-biblioref">Breheny and Huang 2015</a>)</span>. For an illustration, we specify the two demographic variables as the first group, the first <span class="math inline">\(10\)</span> PAM50 genes as the second group, the last <span class="math inline">\(40\)</span> PAM50 genes as the third group. A <span class="math inline">\(k\)</span>-fold cross-validation (CV) for the group Lasso Cox model is performed through function <code>cv.grpsurv()</code>. The returned object <code>cvfit$lambda.min</code> is the value of CV-optimized <span class="math inline">\(\lambda\)</span>. The following results show that</p>
<ul>
<li>when choosing the CV-optimized <span class="math inline">\(\lambda = 0.0143\)</span> (output matrix has lambda values as column names), the estimated coefficients of the first two groups are nonzero (i.e.&nbsp;selecting first and second groups);</li>
<li>when choosing the <span class="math inline">\(10\)</span>-th lambda <span class="math inline">\(\lambda = 0.0217\)</span>, only the first group of covariates has nonzero coefficients (i.e.&nbsp;selecting first group);</li>
<li>when choosing the <span class="math inline">\(15\)</span>-th lambda <span class="math inline">\(\lambda = 0.0108\)</span>, the estimated coefficients of all the three groups are nonzero (i.e.&nbsp;selecting all groups).</li>
</ul>
<p>Note that the <code>R</code> package <a href="https://CRAN.R-project.org/package=grpreg"><strong>grpreg</strong></a> <span class="citation" data-cites="Breheny2015">(<a href="#ref-Breheny2015" role="doc-biblioref">Breheny and Huang 2015</a>)</span> also implements group smoothly clipped absolute deviation (SCAD) model and some others, see <span class="citation" data-cites="Breheny2021">Breheny, Zeng, and Kurth (<a href="#ref-Breheny2021" role="doc-biblioref">2021</a>)</span> for details.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb31"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a><span class="co"># group Lasso Cox model</span></span>
<span id="cb31-2"><a href="#cb31-2" aria-hidden="true" tabindex="-1"></a>group <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="fu">rep</span>(<span class="st">"demographic"</span>, <span class="dv">2</span>), <span class="fu">rep</span>(<span class="st">"PAM50_1"</span>, <span class="dv">10</span>), <span class="fu">rep</span>(<span class="st">"PAM50_2"</span>, <span class="dv">40</span>))</span>
<span id="cb31-3"><a href="#cb31-3" aria-hidden="true" tabindex="-1"></a>group <span class="ot">&lt;-</span> <span class="fu">factor</span>(group)</span>
<span id="cb31-4"><a href="#cb31-4" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">123</span>)</span>
<span id="cb31-5"><a href="#cb31-5" aria-hidden="true" tabindex="-1"></a>cvfit <span class="ot">&lt;-</span> grpreg<span class="sc">::</span><span class="fu">cv.grpsurv</span>(<span class="at">X =</span> x, <span class="at">y =</span> y, <span class="at">group =</span> group, <span class="at">penalty =</span> <span class="st">"grLasso"</span>, <span class="at">returnY =</span> <span class="cn">TRUE</span>)</span>
<span id="cb31-6"><a href="#cb31-6" aria-hidden="true" tabindex="-1"></a><span class="fu">round</span>(cvfit<span class="sc">$</span>fit<span class="sc">$</span>beta[, <span class="fu">c</span>(<span class="fu">which.min</span>(cvfit<span class="sc">$</span>cve), <span class="dv">10</span>, <span class="dv">20</span>)], <span class="at">digits =</span> <span class="dv">4</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<pre class="border"><code>           0.0143  0.0217  0.0108
age        0.0218  0.0154  0.0247
ethnicity -0.0542 -0.0425 -0.0570
ANLN       0.0193  0.0000  0.0713
FOXC1     -0.0032  0.0000 -0.0104
CDH3      -0.0028  0.0000 -0.0090
UBE2T      0.0154  0.0000  0.0571
NDC80     -0.0239  0.0000 -0.0862
PGR       -0.0027  0.0000 -0.0086
BIRC5     -0.0133  0.0000 -0.0497
ORC6       0.0140  0.0000  0.0489
ESR1      -0.0002  0.0000 -0.0008
PHGDH      0.0008  0.0000  0.0024
CDC6       0.0000  0.0000 -0.0094
MMP11      0.0000  0.0000  0.0074
MYBL2      0.0000  0.0000  0.0018
SFRP1      0.0000  0.0000  0.0049
CCNE1      0.0000  0.0000  0.0000
BLVRA      0.0000  0.0000 -0.0436
BAG1       0.0000  0.0000 -0.0163
MLPH       0.0000  0.0000 -0.0155
CDC20      0.0000  0.0000 -0.0129
CENPF      0.0000  0.0000 -0.0245
KRT17      0.0000  0.0000 -0.0125
FOXA1      0.0000  0.0000  0.0040
ACTR3B     0.0000  0.0000 -0.0112
CCNB1      0.0000  0.0000  0.0302
MDM2       0.0000  0.0000 -0.0077
MYC        0.0000  0.0000  0.0002
CEP55      0.0000  0.0000 -0.0242
SLC39A6    0.0000  0.0000  0.0053
ERBB2      0.0000  0.0000 -0.0089
GRB7       0.0000  0.0000  0.0099
KIF2C      0.0000  0.0000  0.0219
NUF2       0.0000  0.0000  0.0210
EGFR       0.0000  0.0000 -0.0150
MKI67      0.0000  0.0000  0.0266
TMEM45B    0.0000  0.0000  0.0100
FGFR4      0.0000  0.0000  0.0023
PTTG1      0.0000  0.0000  0.0095
MELK       0.0000  0.0000 -0.0188
NAT1       0.0000  0.0000 -0.0052
CXXC5      0.0000  0.0000  0.0131
BCL2       0.0000  0.0000 -0.0082
RRM2       0.0000  0.0000 -0.0003
GPR160     0.0000  0.0000 -0.0043
EXO1       0.0000  0.0000  0.0041
UBE2C      0.0000  0.0000 -0.0052
TYMS       0.0000  0.0000 -0.0298
KRT5       0.0000  0.0000 -0.0025
KRT14      0.0000  0.0000  0.0085
MAPT       0.0000  0.0000 -0.0071
MIA        0.0000  0.0000 -0.0180</code></pre>
<p>Sparse group Lasso Cox model is implemented in the <code>R</code> package <a href="https://CRAN.R-project.org/package=SGL"><strong>SGL</strong></a> <span class="citation" data-cites="Simon2019">(<a href="#ref-Simon2019" role="doc-biblioref">N. Simon et al. 2019</a>)</span>. The function <code>cvSGL()</code> uses cross validation to optimize the penalty parameter <span class="math inline">\(\lambda\)</span>. The following example shows that it induces sparsity in each group of covariates.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb33"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a><span class="co"># sparse group Lasso Cox model</span></span>
<span id="cb33-2"><a href="#cb33-2" aria-hidden="true" tabindex="-1"></a>group <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="fu">rep</span>(<span class="st">"demographic"</span>, <span class="dv">2</span>), <span class="fu">rep</span>(<span class="st">"PAM50_1"</span>, <span class="dv">10</span>), <span class="fu">rep</span>(<span class="st">"PAM50_2"</span>, <span class="dv">40</span>))</span>
<span id="cb33-3"><a href="#cb33-3" aria-hidden="true" tabindex="-1"></a>group <span class="ot">&lt;-</span> <span class="fu">factor</span>(group)</span>
<span id="cb33-4"><a href="#cb33-4" aria-hidden="true" tabindex="-1"></a>dat_tmp <span class="ot">&lt;-</span> <span class="fu">list</span>(<span class="at">x =</span> x, <span class="at">time =</span> clin<span class="sc">$</span>time, <span class="at">status =</span> clin<span class="sc">$</span>status)</span>
<span id="cb33-5"><a href="#cb33-5" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">123</span>)</span>
<span id="cb33-6"><a href="#cb33-6" aria-hidden="true" tabindex="-1"></a>cvfit <span class="ot">&lt;-</span> SGL<span class="sc">::</span><span class="fu">cvSGL</span>(dat_tmp, <span class="at">index =</span> group, <span class="at">type =</span> <span class="st">"cox"</span>, <span class="at">nfold =</span> <span class="dv">5</span>)</span>
<span id="cb33-7"><a href="#cb33-7" aria-hidden="true" tabindex="-1"></a>beta.hat <span class="ot">&lt;-</span> cvfit<span class="sc">$</span>fit<span class="sc">$</span>beta[, <span class="fu">which.min</span>(cvfit<span class="sc">$</span>lldiff)]</span>
<span id="cb33-8"><a href="#cb33-8" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(beta.hat) <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="st">"group"</span>, <span class="fu">as.numeric</span>(group), <span class="st">"."</span>, <span class="fu">c</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">2</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">40</span>))</span>
<span id="cb33-9"><a href="#cb33-9" aria-hidden="true" tabindex="-1"></a>beta.hat</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<pre class="border"><code>     group1.1      group1.2      group2.1      group2.2      group2.3      group2.4 
 5.6584838488  0.0000000000  0.4812006103  0.0000000000  0.0000000000  0.2481830177 
     group2.5      group2.6      group2.7      group2.8      group2.9     group2.10 
 0.0000000000 -0.0003042126  0.0000000000  0.3317385412  0.0000000000  0.0000000000 
     group3.1      group3.2      group3.3      group3.4      group3.5      group3.6 
 0.0000000000  0.3037631224  0.0000000000 -0.3782338997  0.0000000000 -2.6805881347 
     group3.7      group3.8      group3.9     group3.10     group3.11     group3.12 
-1.8418523757  0.0000000000  0.0000000000  0.0000000000 -1.7849923007  0.0000000000 
    group3.13     group3.14     group3.15     group3.16     group3.17     group3.18 
 0.0000000000  1.0290918041  0.0000000000  0.0000000000  0.0000000000  0.0000000000 
    group3.19     group3.20     group3.21     group3.22     group3.23     group3.24 
 0.0000000000  0.0000000000  0.0000000000  0.0000000000 -0.3679980817  0.0000000000 
    group3.25     group3.26     group3.27     group3.28     group3.29     group3.30 
 0.9925901529  0.0088469957  0.0000000000  0.0000000000  0.0000000000  0.0000000000 
    group3.31     group3.32     group3.33     group3.34     group3.35     group3.36 
-2.1975942364  0.0000000000  0.0000000000  0.0000000000  0.0000000000 -0.8407228093 
    group3.37     group3.38     group3.39     group3.40 
-1.8217490477  0.0000000000 -0.7323739107 -2.0111900380 </code></pre>
</section>
<section id="sparse-bayesian-cox-models" class="level4">
<h4 class="anchored" data-anchor-id="sparse-bayesian-cox-models">Sparse Bayesian Cox models</h4>
<p>The <code>R</code> package <a href="https://CRAN.R-project.org/package=psbcGroup"><strong>psbcGroup</strong></a> <span class="citation" data-cites="Lee2021">(<a href="#ref-Lee2021" role="doc-biblioref">Lee, Chakraborty, and Sun 2021</a>)</span> integrates a large set of sparse Bayesian Cox models. The function <code>psbcGL()</code> implements Bayesian Cox models with Lasso and group Lasso priors for feature selection and group selection respectively. For the Lasso prior, set the hyperparameter <code>priorPara$groupInd = 1:p</code> where <span class="math inline">\(p\)</span> is the total number of covariates. For the group Lasso prior, set the hyperparameter <code>priorPara$groupInd</code> as a vector of size <span class="math inline">\(p\)</span>, where each element denotes which group each covariate corresponds to.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb35"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Bayesian Cox model with Lasso prior</span></span>
<span id="cb35-2"><a href="#cb35-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-3"><a href="#cb35-3" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">123</span>)</span>
<span id="cb35-4"><a href="#cb35-4" aria-hidden="true" tabindex="-1"></a>survObj <span class="ot">&lt;-</span> <span class="fu">list</span>(<span class="at">t =</span> clin<span class="sc">$</span>time, <span class="at">di =</span> clin<span class="sc">$</span>status, <span class="at">x =</span> x)</span>
<span id="cb35-5"><a href="#cb35-5" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span> <span class="fu">ncol</span>(x)</span>
<span id="cb35-6"><a href="#cb35-6" aria-hidden="true" tabindex="-1"></a><span class="co"># set hyperparameters.</span></span>
<span id="cb35-7"><a href="#cb35-7" aria-hidden="true" tabindex="-1"></a><span class="co"># For Lasso prior (i.e. 'groupInd'= 1:p), larger ratio r/delta tends to force the posterior betas to be more concentrated at 0</span></span>
<span id="cb35-8"><a href="#cb35-8" aria-hidden="true" tabindex="-1"></a><span class="co"># For group Lasso prior (i.e. 'groupInd' as group indicator for covariates), larger ratio r/delta tends to force stronger grouping effect of covariates</span></span>
<span id="cb35-9"><a href="#cb35-9" aria-hidden="true" tabindex="-1"></a>s <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="fu">sort</span>(survObj<span class="sc">$</span>t[survObj<span class="sc">$</span>di <span class="sc">==</span> <span class="dv">1</span>]), <span class="dv">2</span> <span class="sc">*</span> <span class="fu">max</span>(survObj<span class="sc">$</span>t) <span class="sc">-</span> <span class="fu">max</span>(survObj<span class="sc">$</span>t[<span class="sc">-</span><span class="fu">which</span>(survObj<span class="sc">$</span>t <span class="sc">==</span> <span class="fu">max</span>(survObj<span class="sc">$</span>t))]))</span>
<span id="cb35-10"><a href="#cb35-10" aria-hidden="true" tabindex="-1"></a>priorPara <span class="ot">&lt;-</span> <span class="fu">list</span>(</span>
<span id="cb35-11"><a href="#cb35-11" aria-hidden="true" tabindex="-1"></a>  <span class="st">"eta0"</span> <span class="ot">=</span> <span class="dv">1</span>, <span class="st">"kappa0"</span> <span class="ot">=</span> <span class="dv">1</span>, <span class="st">"c0"</span> <span class="ot">=</span> <span class="dv">2</span>, <span class="st">"r"</span> <span class="ot">=</span> <span class="fl">0.5</span>,</span>
<span id="cb35-12"><a href="#cb35-12" aria-hidden="true" tabindex="-1"></a>  <span class="st">"delta"</span> <span class="ot">=</span> <span class="fl">0.0001</span>, <span class="st">"s"</span> <span class="ot">=</span> s, <span class="st">"J"</span> <span class="ot">=</span> <span class="fu">length</span>(s), <span class="st">"groupInd"</span> <span class="ot">=</span> <span class="dv">1</span><span class="sc">:</span>p</span>
<span id="cb35-13"><a href="#cb35-13" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb35-14"><a href="#cb35-14" aria-hidden="true" tabindex="-1"></a><span class="co"># set MCMC parameters</span></span>
<span id="cb35-15"><a href="#cb35-15" aria-hidden="true" tabindex="-1"></a>mcmcPara <span class="ot">&lt;-</span> <span class="fu">list</span>(<span class="st">"numBeta"</span> <span class="ot">=</span> p, <span class="st">"beta.prop.var"</span> <span class="ot">=</span> <span class="dv">1</span>)</span>
<span id="cb35-16"><a href="#cb35-16" aria-hidden="true" tabindex="-1"></a><span class="co"># set initial values of hyperparameters</span></span>
<span id="cb35-17"><a href="#cb35-17" aria-hidden="true" tabindex="-1"></a>lambdaSq <span class="ot">&lt;-</span> <span class="dv">1</span></span>
<span id="cb35-18"><a href="#cb35-18" aria-hidden="true" tabindex="-1"></a>initial <span class="ot">&lt;-</span> <span class="fu">list</span>(</span>
<span id="cb35-19"><a href="#cb35-19" aria-hidden="true" tabindex="-1"></a>  <span class="st">"beta.ini"</span> <span class="ot">=</span> <span class="fu">rep</span>(<span class="dv">0</span>, p), <span class="st">"lambdaSq"</span> <span class="ot">=</span> <span class="dv">1</span>, <span class="st">"sigmaSq"</span> <span class="ot">=</span> <span class="fu">runif</span>(<span class="dv">1</span>, <span class="fl">0.1</span>, <span class="dv">10</span>),</span>
<span id="cb35-20"><a href="#cb35-20" aria-hidden="true" tabindex="-1"></a>  <span class="st">"tauSq"</span> <span class="ot">=</span> <span class="fu">rexp</span>(<span class="fu">length</span>(<span class="fu">unique</span>(priorPara<span class="sc">$</span>groupInd)), <span class="st">"rate"</span> <span class="ot">=</span> lambdaSq <span class="sc">/</span> <span class="dv">2</span>),</span>
<span id="cb35-21"><a href="#cb35-21" aria-hidden="true" tabindex="-1"></a>  <span class="st">"h"</span> <span class="ot">=</span> <span class="fu">rgamma</span>(priorPara<span class="sc">$</span>J, <span class="dv">1</span>, <span class="dv">1</span>)</span>
<span id="cb35-22"><a href="#cb35-22" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb35-23"><a href="#cb35-23" aria-hidden="true" tabindex="-1"></a><span class="co"># in real applications, 'num.reps' should be large enough (e.g. 20000, 40000) and 'chain' to be &gt; 1</span></span>
<span id="cb35-24"><a href="#cb35-24" aria-hidden="true" tabindex="-1"></a><span class="co"># argument 'rw' should be FALSE for high-dimensional covariates</span></span>
<span id="cb35-25"><a href="#cb35-25" aria-hidden="true" tabindex="-1"></a>BayesLassofit <span class="ot">&lt;-</span> psbcGroup<span class="sc">::</span><span class="fu">psbcGL</span>(survObj, priorPara, initial, <span class="at">rw =</span> <span class="cn">TRUE</span>, mcmcPara, <span class="at">num.reps =</span> <span class="dv">100</span>, <span class="at">thin =</span> <span class="dv">1</span>, <span class="at">chain =</span> <span class="dv">1</span>)</span>
<span id="cb35-26"><a href="#cb35-26" aria-hidden="true" tabindex="-1"></a><span class="co"># burn-in the first half MCMC iterations</span></span>
<span id="cb35-27"><a href="#cb35-27" aria-hidden="true" tabindex="-1"></a>beta_p <span class="ot">&lt;-</span> BayesLassofit<span class="sc">$</span>beta.p[<span class="sc">-</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">51</span>), ]</span>
<span id="cb35-28"><a href="#cb35-28" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(beta_p) <span class="ot">&lt;-</span> <span class="fu">colnames</span>(x)</span>
<span id="cb35-29"><a href="#cb35-29" aria-hidden="true" tabindex="-1"></a>psbcSpeedUp<span class="sc">:::</span><span class="fu">plot.psbcSpeedUp</span>(beta_p)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><img src="fig/TCGA_bayeslasso.png" class="img-fluid figure-img" style="width:50.0%"></p>
<figcaption class="figure-caption"><em>Estimates of regression coefficients by a penalized semiparametric Bayesian Cox model with Lasso prior. Solid dots indicate the posterior mean over MCMC iterations (excluding burn-in period), and horizontal lines show the corresponding 95% credibility intervals.</em></figcaption>
</figure>
</div>
<p><br></p>
<p>Note that <strong>psbcGroup</strong> cannot distinguish mandatory (unpenalized) covariates with omics features, see <span class="citation" data-cites="Zucknick2015">Zucknick, Saadati, and Benner (<a href="#ref-Zucknick2015" role="doc-biblioref">2015</a>)</span> for an extended Bayesian Lasso Cox model. <span id="psbcSpeedUp"></span> The following code implements the Bayesian Lasso Cox model with mandatory covariates through the <code>R</code> package <a href="https://CRAN.R-project.org/package=psbcSpeedUp"><strong>psbcSpeedUp</strong></a> <span class="citation" data-cites="Zhao2023b">(<a href="#ref-Zhao2023b" role="doc-biblioref">Z. Zhao et al. 2023</a>)</span>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb36"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Bayesian Cox model with Lasso prior and mandatory covariates</span></span>
<span id="cb36-2"><a href="#cb36-2" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">123</span>)</span>
<span id="cb36-3"><a href="#cb36-3" aria-hidden="true" tabindex="-1"></a>survObjM <span class="ot">&lt;-</span> <span class="fu">list</span>(<span class="at">t =</span> clin<span class="sc">$</span>time, <span class="at">di =</span> clin<span class="sc">$</span>status, <span class="at">x =</span> x[, <span class="fu">c</span>(<span class="dv">3</span><span class="sc">:</span><span class="dv">52</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">2</span>)])</span>
<span id="cb36-4"><a href="#cb36-4" aria-hidden="true" tabindex="-1"></a>priorPara <span class="ot">&lt;-</span> <span class="fu">list</span>(<span class="st">"eta0"</span> <span class="ot">=</span> <span class="dv">1</span>, <span class="st">"kappa0"</span> <span class="ot">=</span> <span class="dv">1</span>, <span class="st">"c0"</span> <span class="ot">=</span> <span class="dv">2</span>, <span class="st">"r"</span> <span class="ot">=</span> <span class="fl">0.5</span>, <span class="st">"delta"</span> <span class="ot">=</span> <span class="fl">0.0001</span>)</span>
<span id="cb36-5"><a href="#cb36-5" aria-hidden="true" tabindex="-1"></a>BayesLassoMfit <span class="ot">&lt;-</span> psbcSpeedUp<span class="sc">::</span><span class="fu">psbcSpeedUp</span>(survObjM,</span>
<span id="cb36-6"><a href="#cb36-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">p =</span> <span class="dv">50</span>, <span class="at">q =</span> <span class="dv">2</span>, <span class="at">hyperpar =</span> priorPara,</span>
<span id="cb36-7"><a href="#cb36-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">nIter =</span> <span class="dv">100</span>, <span class="at">burnin =</span> <span class="dv">50</span>, <span class="at">thin =</span> <span class="dv">1</span>, <span class="at">rw =</span> <span class="cn">FALSE</span>, <span class="at">outFilePath =</span> <span class="st">"tmp"</span></span>
<span id="cb36-8"><a href="#cb36-8" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb36-9"><a href="#cb36-9" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(BayesLassoMfit)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<pre class="border"><code>Running MCMC iterations ...
[##################################################] 100%
DONE, exiting! </code></pre>
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><img src="fig/TCGA_bayesLassoM.png" class="img-fluid figure-img" style="width:50.0%"></p>
<figcaption class="figure-caption"><em>Estimates of regression coefficients by a penalized semiparametric Bayesian Cox model with Lasso prior and unpenalized covariates. Solid dots indicate the posterior mean over MCMC iterations (excluding burn-in period), and horizontal lines show the corresponding 95% credibility intervals.</em></figcaption>
</figure>
</div>
<p><br></p>
<p>In the <code>R</code> package <a href="https://CRAN.R-project.org/package=psbcGroup"><strong>psbcGroup</strong></a> <span class="citation" data-cites="Lee2021">(<a href="#ref-Lee2021" role="doc-biblioref">Lee, Chakraborty, and Sun 2021</a>)</span>, function <code>psbcEN()</code> implements Bayesian Cox models with Elastic Net prior for feature selection with grouping effect of correlated features. Function <code>psbcFL()</code> implements Bayesian Cox models with fused Lasso prior.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb38"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb38-1"><a href="#cb38-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Bayesian Cox model with Elastic Net prior</span></span>
<span id="cb38-2"><a href="#cb38-2" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">123</span>)</span>
<span id="cb38-3"><a href="#cb38-3" aria-hidden="true" tabindex="-1"></a><span class="co"># set hyperparameters</span></span>
<span id="cb38-4"><a href="#cb38-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Larger ratio r1/delta1 forces the posterior betas to be more concentrated at 0</span></span>
<span id="cb38-5"><a href="#cb38-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Larger ratio r2/delta2 forces stronger grouping effect of covariates</span></span>
<span id="cb38-6"><a href="#cb38-6" aria-hidden="true" tabindex="-1"></a>priorPara <span class="ot">&lt;-</span> <span class="fu">list</span>(</span>
<span id="cb38-7"><a href="#cb38-7" aria-hidden="true" tabindex="-1"></a>  <span class="st">"eta0"</span> <span class="ot">=</span> <span class="dv">1</span>, <span class="st">"kappa0"</span> <span class="ot">=</span> <span class="dv">1</span>, <span class="st">"c0"</span> <span class="ot">=</span> <span class="dv">2</span>, <span class="st">"r1"</span> <span class="ot">=</span> <span class="fl">0.1</span>, <span class="st">"r2"</span> <span class="ot">=</span> <span class="dv">1</span>,</span>
<span id="cb38-8"><a href="#cb38-8" aria-hidden="true" tabindex="-1"></a>  <span class="st">"delta1"</span> <span class="ot">=</span> <span class="fl">0.1</span>, <span class="st">"delta2"</span> <span class="ot">=</span> <span class="dv">1</span>, <span class="st">"s"</span> <span class="ot">=</span> s, <span class="st">"J"</span> <span class="ot">=</span> <span class="fu">length</span>(s)</span>
<span id="cb38-9"><a href="#cb38-9" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb38-10"><a href="#cb38-10" aria-hidden="true" tabindex="-1"></a><span class="co"># set MCMC parameters</span></span>
<span id="cb38-11"><a href="#cb38-11" aria-hidden="true" tabindex="-1"></a>mcmcPara <span class="ot">&lt;-</span> <span class="fu">list</span>(<span class="st">"numBeta"</span> <span class="ot">=</span> p, <span class="st">"beta.prop.var"</span> <span class="ot">=</span> <span class="dv">1</span>)</span>
<span id="cb38-12"><a href="#cb38-12" aria-hidden="true" tabindex="-1"></a><span class="co"># set initial values of hyperparameters</span></span>
<span id="cb38-13"><a href="#cb38-13" aria-hidden="true" tabindex="-1"></a>initial <span class="ot">&lt;-</span> <span class="fu">list</span>(</span>
<span id="cb38-14"><a href="#cb38-14" aria-hidden="true" tabindex="-1"></a>  <span class="st">"beta.ini"</span> <span class="ot">=</span> <span class="fu">rep</span>(<span class="dv">0</span>, p), <span class="st">"lambda1Sq"</span> <span class="ot">=</span> <span class="dv">1</span>, <span class="st">"lambda2"</span> <span class="ot">=</span> <span class="dv">1</span>, <span class="st">"sigmaSq"</span> <span class="ot">=</span> <span class="fu">runif</span>(<span class="dv">1</span>, <span class="fl">0.1</span>, <span class="dv">10</span>),</span>
<span id="cb38-15"><a href="#cb38-15" aria-hidden="true" tabindex="-1"></a>  <span class="st">"tauSq"</span> <span class="ot">=</span> <span class="fu">rexp</span>(p, <span class="at">rate =</span> <span class="dv">1</span> <span class="sc">/</span> <span class="dv">2</span>), <span class="st">"h"</span> <span class="ot">=</span> <span class="fu">rgamma</span>(priorPara<span class="sc">$</span>J, <span class="dv">1</span>, <span class="dv">1</span>)</span>
<span id="cb38-16"><a href="#cb38-16" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb38-17"><a href="#cb38-17" aria-hidden="true" tabindex="-1"></a><span class="co"># in real application, 'num.reps' should be large enough (e.g. 20000, 40000) and 'chain' to be &gt; 1</span></span>
<span id="cb38-18"><a href="#cb38-18" aria-hidden="true" tabindex="-1"></a>BayesENfit <span class="ot">&lt;-</span> <span class="fu">psbcEN</span>(survObj, priorPara, initial, <span class="at">rw =</span> <span class="cn">FALSE</span>, mcmcPara, </span>
<span id="cb38-19"><a href="#cb38-19" aria-hidden="true" tabindex="-1"></a>                     <span class="at">num.reps =</span> <span class="dv">100</span>, <span class="at">thin =</span> <span class="dv">1</span>, <span class="at">chain =</span> <span class="dv">1</span>)</span>
<span id="cb38-20"><a href="#cb38-20" aria-hidden="true" tabindex="-1"></a><span class="co"># burn-in the first half MCMC iterations</span></span>
<span id="cb38-21"><a href="#cb38-21" aria-hidden="true" tabindex="-1"></a>EN_beta_p <span class="ot">&lt;-</span> BayesENfit<span class="sc">$</span>beta.p[<span class="dv">52</span><span class="sc">:</span><span class="dv">101</span>, ]</span>
<span id="cb38-22"><a href="#cb38-22" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(EN_beta_p) <span class="ot">&lt;-</span> <span class="fu">colnames</span>(x)</span>
<span id="cb38-23"><a href="#cb38-23" aria-hidden="true" tabindex="-1"></a>psbcSpeedUp<span class="sc">:::</span><span class="fu">plot.psbcSpeedUp</span>(EN_beta_p)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><img src="fig/TCGA_bayesEN.png" class="img-fluid figure-img" style="width:50.0%"></p>
<figcaption class="figure-caption"><em>Estimates of regression coefficients by a penalized semiparametric Bayesian Cox model with Elastic Net prior. Solid dots indicate the posterior mean over MCMC iterations (excluding burn-in period), and horizontal lines show the corresponding 95% credibility intervals.</em></figcaption>
</figure>
</div>
<p><br></p>
<p>A penalized semiparametric Bayesian Cox model with double exponential spike-and-slab prior is implemented in the <code>R</code> package <a href="https://github.com/nyiuab/BhGLM.git"><strong>BhGLM</strong></a> <span class="citation" data-cites="Yi2019">(<a href="#ref-Yi2019" role="doc-biblioref">Yi et al. 2019</a>)</span>. Note that <strong>BhGLM</strong> provides frequentist confidence intervals of the posterior mode of the coefficients.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb39"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb39-1"><a href="#cb39-1" aria-hidden="true" tabindex="-1"></a><span class="co"># penalized semiparametric Bayesian Cox model with (double exponential) spike-and-slab prior</span></span>
<span id="cb39-2"><a href="#cb39-2" aria-hidden="true" tabindex="-1"></a>y_surv <span class="ot">&lt;-</span> <span class="fu">Surv</span>(clin<span class="sc">$</span>time, clin<span class="sc">$</span>status)</span>
<span id="cb39-3"><a href="#cb39-3" aria-hidden="true" tabindex="-1"></a>x_dataframe <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(x)</span>
<span id="cb39-4"><a href="#cb39-4" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">123</span>)</span>
<span id="cb39-5"><a href="#cb39-5" aria-hidden="true" tabindex="-1"></a>Bayesfit <span class="ot">&lt;-</span> BhGLM<span class="sc">::</span><span class="fu">bcoxph</span>(y_surv <span class="sc">~</span> ., x_dataframe, <span class="at">prior =</span> <span class="fu">mde</span>(<span class="dv">0</span>, <span class="fl">0.01</span>, <span class="fl">0.8</span>), <span class="at">control =</span> <span class="fu">coxph.control</span>(<span class="at">iter.max =</span> <span class="dv">200</span>))</span>
<span id="cb39-6"><a href="#cb39-6" aria-hidden="true" tabindex="-1"></a>BhGLM<span class="sc">::</span><span class="fu">plot.bh</span>(Bayesfit, <span class="at">col.pts =</span> <span class="fu">c</span>(<span class="st">"red"</span>, <span class="st">"blue"</span>), <span class="at">main =</span> <span class="st">"Cox with mixture double exponential</span><span class="sc">\n</span><span class="st">"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><img src="fig/TCGA_bayesSpikeSlab.png" class="img-fluid figure-img" style="width:60.0%"></p>
<figcaption class="figure-caption"><em>Coefficient estimates of a penalized semiparametric Bayesian Cox model with (double exponential) spike-and-slab prior. Solid dots denote the posterior mode of the coefficients and lines denote the 95% confidence intervals. Red colored text on the right side mark the significant features with <span class="math inline">\(p &lt; 0.05\)</span>.</em></figcaption>
</figure>
</div>
</section>
</section>
</section>
<section id="survival-model-validation" class="level2 unnumbered">
<h2 class="unnumbered anchored" data-anchor-id="survival-model-validation">Survival model validation</h2>
<p>The ideal evaluation of a prognostic model is based on completely independent validation data, since high-dimensional survival models built on the training data can be overfitted. If there are no independent validation data, it is recommended to use resampling-based methods for estimating the <strong>uncertainty</strong> of the model’s prediction performance. This can be done for example by repeatedly splitting the dataset to training/validation sets and evaluating a model’s performance on the different validation sets using various evaluation metrics.</p>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Model validation
</div>
</div>
<div class="callout-body-container callout-body">
<p>To validate a prediction model systematically, the predictive performance of the model is commonly addressed by</p>
<ul>
<li><strong>Discrimination</strong>: the ability of the model to distinguish between low and high risk patients</li>
<li><strong>Calibration</strong>: the agreement between the observed and predicted survival probabilities</li>
<li><strong>Overall performance</strong>: the distance between the observed and predicted survival probabilities</li>
</ul>
</div>
</div>
<p>The performance metrics can be <em>time-dependent</em> or <em>time-independent</em>, with the time-dependent metrics being more informative in general compared to integrated measures (i.e.&nbsp;evaluated across many time points). For survival data, we can assess the <strong>discriminatory power</strong> of a model (i.e.&nbsp;how well does it ranks patients) or how well a model is <strong>calibrated</strong> (i.e.&nbsp;how closely the predicted survival probabilities agree numerically with the actual survival outcomes). For example, measures such as the receiver operating characteristic (ROC) curve, the (integrated) area under time-specific ROC curves (<strong>AUC</strong>, <span class="citation" data-cites="Heagerty2005">Heagerty and Zheng (<a href="#ref-Heagerty2005" role="doc-biblioref">2005</a>)</span>) and the concordance index (<strong>C-index</strong>, <span class="citation" data-cites="Harrell1982">Harrell et al. (<a href="#ref-Harrell1982" role="doc-biblioref">1982</a>)</span>) are measures of discrimination, while the right-censored logarithmic loss (<strong>RCLL</strong>, <span class="citation" data-cites="Avati2020">Avati et al. (<a href="#ref-Avati2020" role="doc-biblioref">2020</a>)</span>) and the well-known <strong>Brier score</strong> <span class="citation" data-cites="Graf1999">(<a href="#ref-Graf1999" role="doc-biblioref">Graf et al. 1999</a>)</span> are used to evaluate both discrimination and calibration performance.</p>
<section id="model-evaluation-classic" class="level3 unnumbered">
<h3 class="unnumbered anchored" data-anchor-id="model-evaluation-classic">Model evaluation (classic)</h3>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p>‘Classic’ here refers to the use of manual <code>R</code> code in combination with many separate <code>R</code> packages which have been routinely used in academia the latest 10+ years for evaluating survival models.</p>
</div>
</div>
<p>To evaluate the performance of a statistical model, we first split the data into training and validation data sets. For example, we can randomly split the 1047 BRCA patients from TCGA into <span class="math inline">\(80\%\)</span> as training set and <span class="math inline">\(20\%\)</span> as validation set.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb40"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb40-1"><a href="#cb40-1" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">123</span>)</span>
<span id="cb40-2"><a href="#cb40-2" aria-hidden="true" tabindex="-1"></a>n <span class="ot">&lt;-</span> <span class="fu">nrow</span>(x)</span>
<span id="cb40-3"><a href="#cb40-3" aria-hidden="true" tabindex="-1"></a>idx <span class="ot">&lt;-</span> <span class="fu">sample</span>(<span class="dv">1</span><span class="sc">:</span>n, n <span class="sc">*</span> <span class="fl">0.8</span>, <span class="at">replace =</span> <span class="cn">FALSE</span>)</span>
<span id="cb40-4"><a href="#cb40-4" aria-hidden="true" tabindex="-1"></a>x_train <span class="ot">&lt;-</span> x[idx, ]</span>
<span id="cb40-5"><a href="#cb40-5" aria-hidden="true" tabindex="-1"></a>y_train <span class="ot">&lt;-</span> y[idx, ]</span>
<span id="cb40-6"><a href="#cb40-6" aria-hidden="true" tabindex="-1"></a>x_validate <span class="ot">&lt;-</span> x[<span class="sc">-</span>idx, ]</span>
<span id="cb40-7"><a href="#cb40-7" aria-hidden="true" tabindex="-1"></a>y_validate <span class="ot">&lt;-</span> y[<span class="sc">-</span>idx, ]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="callout callout-style-default callout-important callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Important
</div>
</div>
<div class="callout-body-container callout-body">
<p>The <span class="math inline">\(20\%\)</span> split of a dataset is often not considered an <strong>independent</strong> dataset and <strong>resampling-based methods</strong> should be used in such cases to provide an unbiased estimate of the predictive accuracy of a prognostic model.</p>
</div>
</div>
<section id="discrimination-metrics" class="level4 unnumbered">
<h4 class="unnumbered anchored" data-anchor-id="discrimination-metrics">Discrimination metrics</h4>
<p><font size="4"> <strong>Goodness-of-fit</strong> </font></p>
<p>The simplest way to demonstrate the prognostic power of a survival model is to dichotomize the prognostic scores (i.e., linear predictor <span class="math inline">\(lp\)</span> in the Cox model) by median value, and then to use a log-rank test to compare the survival curves of the patients in the two groups. We use the built model to predict the prognostic scores based on the <span class="math inline">\(20\%\)</span> validation data. The following code shows the <strong>goodness-of-fit</strong> of a Lasso Cox model with the BRCA patients survival and PAM50 mRNA-Seq data from TCGA.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb41"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb41-1"><a href="#cb41-1" aria-hidden="true" tabindex="-1"></a><span class="co"># train a Lasso Cox model, similarly for other Cox-type models</span></span>
<span id="cb41-2"><a href="#cb41-2" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">123</span>)</span>
<span id="cb41-3"><a href="#cb41-3" aria-hidden="true" tabindex="-1"></a>cvfit <span class="ot">&lt;-</span> <span class="fu">cv.glmnet</span>(x_train, y_train, <span class="at">family =</span> <span class="st">"cox"</span>, <span class="at">nfolds =</span> <span class="dv">5</span>, <span class="at">penalty.factor =</span> pf)</span>
<span id="cb41-4"><a href="#cb41-4" aria-hidden="true" tabindex="-1"></a>pred_lp <span class="ot">&lt;-</span> <span class="fu">predict</span>(cvfit, <span class="at">newx =</span> x_validate, <span class="at">s =</span> cvfit<span class="sc">$</span>lambda.min, <span class="at">type =</span> <span class="st">"link"</span>)</span>
<span id="cb41-5"><a href="#cb41-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-6"><a href="#cb41-6" aria-hidden="true" tabindex="-1"></a><span class="co"># dichotomize by prognostic scores (linear predictor)  by median to divide the validation patients into two groups</span></span>
<span id="cb41-7"><a href="#cb41-7" aria-hidden="true" tabindex="-1"></a>group_dichotomize <span class="ot">&lt;-</span> <span class="fu">as.numeric</span>(pred_lp <span class="sc">&gt;</span> <span class="fu">median</span>(pred_lp))</span>
<span id="cb41-8"><a href="#cb41-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-9"><a href="#cb41-9" aria-hidden="true" tabindex="-1"></a><span class="co"># draw two survival curves based on KM estimation and compare them by a log-rank test</span></span>
<span id="cb41-10"><a href="#cb41-10" aria-hidden="true" tabindex="-1"></a>dat_tmp <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">time =</span> y_validate[, <span class="dv">1</span>], <span class="at">status =</span> y_validate[, <span class="dv">2</span>], <span class="at">group =</span> group_dichotomize)</span>
<span id="cb41-11"><a href="#cb41-11" aria-hidden="true" tabindex="-1"></a>sfit <span class="ot">&lt;-</span> <span class="fu">survfit</span>(<span class="fu">Surv</span>(time, status) <span class="sc">~</span> group, <span class="at">data =</span> dat_tmp)</span>
<span id="cb41-12"><a href="#cb41-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-13"><a href="#cb41-13" aria-hidden="true" tabindex="-1"></a>ggsurv <span class="ot">&lt;-</span> <span class="fu">ggsurvplot</span>(sfit,</span>
<span id="cb41-14"><a href="#cb41-14" aria-hidden="true" tabindex="-1"></a>  <span class="at">conf.int =</span> <span class="cn">TRUE</span>, <span class="at">risk.table =</span> <span class="cn">TRUE</span>,</span>
<span id="cb41-15"><a href="#cb41-15" aria-hidden="true" tabindex="-1"></a>  <span class="at">xlab =</span> <span class="st">"Time since diagnosis (year)"</span>, <span class="at">legend =</span> <span class="fu">c</span>(.<span class="dv">2</span>, .<span class="dv">3</span>),</span>
<span id="cb41-16"><a href="#cb41-16" aria-hidden="true" tabindex="-1"></a>  <span class="at">legend.labs =</span> <span class="fu">c</span>(<span class="st">"Low risk"</span>, <span class="st">"High risk"</span>), <span class="at">legend.title =</span> <span class="st">"Dichotomized groups"</span>,</span>
<span id="cb41-17"><a href="#cb41-17" aria-hidden="true" tabindex="-1"></a>  <span class="at">risk.table.y.text.col =</span> <span class="cn">TRUE</span>, <span class="at">risk.table.y.text =</span> <span class="cn">FALSE</span></span>
<span id="cb41-18"><a href="#cb41-18" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb41-19"><a href="#cb41-19" aria-hidden="true" tabindex="-1"></a>ggsurv<span class="sc">$</span>plot <span class="ot">&lt;-</span> ggsurv<span class="sc">$</span>plot <span class="sc">+</span></span>
<span id="cb41-20"><a href="#cb41-20" aria-hidden="true" tabindex="-1"></a>  <span class="fu">annotate</span>(<span class="st">"text"</span>, <span class="at">x =</span> <span class="fl">2.6</span>, <span class="at">y =</span> .<span class="dv">03</span>, <span class="at">label =</span> <span class="fu">paste0</span>(<span class="st">"Log-rank test:</span><span class="sc">\n</span><span class="st">"</span>, <span class="fu">surv_pvalue</span>(sfit)<span class="sc">$</span>pval.txt))</span>
<span id="cb41-21"><a href="#cb41-21" aria-hidden="true" tabindex="-1"></a>ggsurv<span class="sc">$</span>table <span class="ot">&lt;-</span> ggsurv<span class="sc">$</span>table <span class="sc">+</span> <span class="fu">labs</span>(<span class="at">y =</span> <span class="st">"Dichotomized</span><span class="sc">\n</span><span class="st"> groups"</span>)</span>
<span id="cb41-22"><a href="#cb41-22" aria-hidden="true" tabindex="-1"></a>ggsurv</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><img src="fig/TCGA_surv_km_lasso.png" class="img-fluid figure-img" style="width:50.0%"></p>
<figcaption class="figure-caption"><em>Kaplan-Meier curves of the BRCA patients data dichotomized by the median of prognostic scores (calculated from the Lasso Cox model with patients’ survival and mRNA-Seq data) into two groups. The log-rank test is to compare the two survival distributions corresponding to the two groups of patients.</em></figcaption>
</figure>
</div>
<p><br></p>
<p>The prognostic scores can also be divided into three or more groups based on quantiles and the log-rank test can be used to compare the difference of multiple survival curves.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb42"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb42-1"><a href="#cb42-1" aria-hidden="true" tabindex="-1"></a>group <span class="ot">&lt;-</span> pred_lp</span>
<span id="cb42-2"><a href="#cb42-2" aria-hidden="true" tabindex="-1"></a>group[pred_lp <span class="sc">&gt;=</span> <span class="fu">quantile</span>(pred_lp, <span class="dv">2</span> <span class="sc">/</span> <span class="dv">3</span>)] <span class="ot">&lt;-</span> <span class="dv">3</span></span>
<span id="cb42-3"><a href="#cb42-3" aria-hidden="true" tabindex="-1"></a>group[pred_lp <span class="sc">&gt;=</span> <span class="fu">quantile</span>(pred_lp, <span class="dv">1</span> <span class="sc">/</span> <span class="dv">3</span>) <span class="sc">&amp;</span> pred_lp <span class="sc">&lt;</span> <span class="fu">quantile</span>(pred_lp, <span class="dv">2</span> <span class="sc">/</span> <span class="dv">3</span>)] <span class="ot">&lt;-</span> <span class="dv">2</span></span>
<span id="cb42-4"><a href="#cb42-4" aria-hidden="true" tabindex="-1"></a>group[pred_lp <span class="sc">&lt;</span> <span class="fu">quantile</span>(pred_lp, <span class="dv">1</span> <span class="sc">/</span> <span class="dv">3</span>)] <span class="ot">&lt;-</span> <span class="dv">1</span></span>
<span id="cb42-5"><a href="#cb42-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-6"><a href="#cb42-6" aria-hidden="true" tabindex="-1"></a><span class="co"># draw two survival curves based on KM estimation and compare them by a log-rank test</span></span>
<span id="cb42-7"><a href="#cb42-7" aria-hidden="true" tabindex="-1"></a>dat_tmp <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">time =</span> y_validate[, <span class="dv">1</span>], <span class="at">status =</span> y_validate[, <span class="dv">2</span>], <span class="at">group =</span> group)</span>
<span id="cb42-8"><a href="#cb42-8" aria-hidden="true" tabindex="-1"></a>sfit <span class="ot">&lt;-</span> <span class="fu">survfit</span>(<span class="fu">Surv</span>(time, status) <span class="sc">~</span> group, <span class="at">data =</span> dat_tmp)</span>
<span id="cb42-9"><a href="#cb42-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-10"><a href="#cb42-10" aria-hidden="true" tabindex="-1"></a>ggsurv <span class="ot">&lt;-</span> <span class="fu">ggsurvplot</span>(sfit,</span>
<span id="cb42-11"><a href="#cb42-11" aria-hidden="true" tabindex="-1"></a>  <span class="at">conf.int =</span> <span class="cn">TRUE</span>, <span class="at">risk.table =</span> <span class="cn">TRUE</span>,</span>
<span id="cb42-12"><a href="#cb42-12" aria-hidden="true" tabindex="-1"></a>  <span class="at">xlab =</span> <span class="st">"Time since diagnosis (year)"</span>, <span class="at">legend =</span> <span class="fu">c</span>(.<span class="dv">2</span>, .<span class="dv">3</span>),</span>
<span id="cb42-13"><a href="#cb42-13" aria-hidden="true" tabindex="-1"></a>  <span class="at">legend.labs =</span> <span class="fu">c</span>(<span class="st">"Low risk"</span>, <span class="st">"Middle risk"</span>, <span class="st">"High risk"</span>), <span class="at">legend.title =</span> <span class="st">"Groups"</span>,</span>
<span id="cb42-14"><a href="#cb42-14" aria-hidden="true" tabindex="-1"></a>  <span class="at">risk.table.y.text.col =</span> <span class="cn">TRUE</span>, <span class="at">risk.table.y.text =</span> <span class="cn">FALSE</span></span>
<span id="cb42-15"><a href="#cb42-15" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb42-16"><a href="#cb42-16" aria-hidden="true" tabindex="-1"></a>ggsurv<span class="sc">$</span>plot <span class="ot">&lt;-</span> ggsurv<span class="sc">$</span>plot <span class="sc">+</span></span>
<span id="cb42-17"><a href="#cb42-17" aria-hidden="true" tabindex="-1"></a>  <span class="fu">annotate</span>(<span class="st">"text"</span>, <span class="at">x =</span> <span class="fl">3.5</span>, <span class="at">y =</span> .<span class="dv">05</span>, <span class="at">label =</span> <span class="fu">paste0</span>(<span class="st">"Log-rank test:</span><span class="sc">\n</span><span class="st">"</span>, <span class="fu">surv_pvalue</span>(sfit)<span class="sc">$</span>pval.txt))</span>
<span id="cb42-18"><a href="#cb42-18" aria-hidden="true" tabindex="-1"></a>ggsurv</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><img src="fig/TCGA_surv_km_lasso2.png" class="img-fluid figure-img" style="width:50.0%"></p>
<figcaption class="figure-caption"><em>Kaplan-Meier curves of the BRCA patients data divided by 33% and 67% quantiles of prognostic scores (calculated from the Lasso Cox model with patients’ survival and mRNA-Seq data) into three groups. The log-rank test is to compare the two survival distributions corresponding to the three groups of patients.</em></figcaption>
</figure>
</div>
<p><br></p>
<p><font size="4"> <strong>ROC curve</strong> </font></p>
<p>The <code>R</code> package <a href="https://CRAN.R-project.org/package=risksetROC"><strong>risksetROC</strong></a> <span class="citation" data-cites="Heagerty2005">(<a href="#ref-Heagerty2005" role="doc-biblioref">Heagerty and Zheng 2005</a>)</span> can estimate a ROC curve at an evaluation time point. The following code draws a ROC curve at 5-years survival evaluation time point for the 20% TCGA validation data and based on a Lasso Cox model learned from the 80% training data.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb43"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb43-1"><a href="#cb43-1" aria-hidden="true" tabindex="-1"></a>ROC <span class="ot">&lt;-</span> <span class="fu">risksetROC</span>(</span>
<span id="cb43-2"><a href="#cb43-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">Stime =</span> y_validate[, <span class="dv">1</span>], <span class="at">status =</span> y_validate[, <span class="dv">2</span>],</span>
<span id="cb43-3"><a href="#cb43-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">marker =</span> pred_lp, <span class="at">predict.time =</span> <span class="dv">5</span>, <span class="at">method =</span> <span class="st">"Cox"</span>,</span>
<span id="cb43-4"><a href="#cb43-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">main =</span> <span class="st">"ROC Curve"</span>, <span class="at">col =</span> <span class="st">"seagreen3"</span>, <span class="at">type =</span> <span class="st">"s"</span>,</span>
<span id="cb43-5"><a href="#cb43-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">lwd =</span> <span class="dv">2</span>, <span class="at">xlab =</span> <span class="st">"1 - Specificity"</span>, <span class="at">ylab =</span> <span class="st">"Sensitivity"</span></span>
<span id="cb43-6"><a href="#cb43-6" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb43-7"><a href="#cb43-7" aria-hidden="true" tabindex="-1"></a><span class="fu">text</span>(<span class="fl">0.7</span>, <span class="fl">0.2</span>, <span class="fu">paste</span>(<span class="st">"AUC ="</span>, <span class="fu">round</span>(ROC<span class="sc">$</span>AUC, <span class="dv">3</span>)))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><img src="fig/TCGA_surv_roc.png" class="img-fluid figure-img" style="width:40.0%"></p>
<figcaption class="figure-caption"><em>ROC curve estimated at 5-years survival evaluation time point for the 20% TCGA validation data and based on a Lasso Cox model learned from the 80% training data. The AUC value is the area under the ROC curve. The diagonal line represents the performance of a random prediction of the outcome event with AUC = 0.5.</em></figcaption>
</figure>
</div>
<p><br></p>
<p><font size="4"> <strong>Time-dependent AUC</strong> </font></p>
<p>Both time-dependent and integrated AUCs can be estimated by the <code>R</code> package <a href="https://CRAN.R-project.org/package=risksetROC"><strong>risksetROC</strong></a>. We demonstrate the calculation based on both training and validation data.</p>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Survival prediction in Cox models
</div>
</div>
<div class="callout-body-container callout-body">
<p>A Cox proportional hazards model (and Lasso Cox as a consequence) is a semi-parametric model, which means that it does not produce survival distribution predictions by default. However, using the function <code>risksetROC::CoxWeights()</code> you can transform the <code>cv.glmnet</code>’s output linear predictors (<code>lp</code>) to survival distribution predictions. This transformation internally uses the <strong>Breslow estimator</strong> for the cumulative baseline hazard.</p>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb44"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb44-1"><a href="#cb44-1" aria-hidden="true" tabindex="-1"></a><span class="co"># unique event times for patients in the training and validation data sets</span></span>
<span id="cb44-2"><a href="#cb44-2" aria-hidden="true" tabindex="-1"></a>utimes_train <span class="ot">&lt;-</span> <span class="fu">sort</span>(<span class="fu">unique</span>(y_train[y_train[, <span class="dv">2</span>] <span class="sc">==</span> <span class="dv">1</span>, <span class="dv">1</span>]))</span>
<span id="cb44-3"><a href="#cb44-3" aria-hidden="true" tabindex="-1"></a>utimes_validate <span class="ot">&lt;-</span> <span class="fu">sort</span>(<span class="fu">unique</span>(y_validate[y_validate[, <span class="dv">2</span>] <span class="sc">==</span> <span class="dv">1</span>, <span class="dv">1</span>]))</span>
<span id="cb44-4"><a href="#cb44-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-5"><a href="#cb44-5" aria-hidden="true" tabindex="-1"></a><span class="co"># markers from the estimated linear predictors of a Lasso Cox model</span></span>
<span id="cb44-6"><a href="#cb44-6" aria-hidden="true" tabindex="-1"></a>pred_lp_train <span class="ot">&lt;-</span> <span class="fu">predict</span>(cvfit, <span class="at">newx =</span> x_train, <span class="at">s =</span> cvfit<span class="sc">$</span>lambda.min, <span class="at">type =</span> <span class="st">"link"</span>)</span>
<span id="cb44-7"><a href="#cb44-7" aria-hidden="true" tabindex="-1"></a>pred_lp_validate <span class="ot">&lt;-</span> <span class="fu">predict</span>(cvfit, <span class="at">newx =</span> x_validate, <span class="at">s =</span> cvfit<span class="sc">$</span>lambda.min, <span class="at">type =</span> <span class="st">"link"</span>)</span>
<span id="cb44-8"><a href="#cb44-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-9"><a href="#cb44-9" aria-hidden="true" tabindex="-1"></a><span class="do">## compute time-dependent AUC</span></span>
<span id="cb44-10"><a href="#cb44-10" aria-hidden="true" tabindex="-1"></a>AUC_train <span class="ot">&lt;-</span> <span class="fu">rep</span>(<span class="cn">NA</span>, <span class="fu">length</span>(utimes_train))</span>
<span id="cb44-11"><a href="#cb44-11" aria-hidden="true" tabindex="-1"></a>AUC_validate <span class="ot">&lt;-</span> <span class="fu">rep</span>(<span class="cn">NA</span>, <span class="fu">length</span>(utimes_validate))</span>
<span id="cb44-12"><a href="#cb44-12" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (j <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(utimes_train)) {</span>
<span id="cb44-13"><a href="#cb44-13" aria-hidden="true" tabindex="-1"></a>  out <span class="ot">&lt;-</span> risksetROC<span class="sc">::</span><span class="fu">CoxWeights</span>(</span>
<span id="cb44-14"><a href="#cb44-14" aria-hidden="true" tabindex="-1"></a>    <span class="at">marker =</span> pred_lp_train, <span class="at">Stime =</span> y_train[, <span class="dv">1</span>],</span>
<span id="cb44-15"><a href="#cb44-15" aria-hidden="true" tabindex="-1"></a>    <span class="at">status =</span> y_train[, <span class="dv">2</span>], <span class="at">predict.time =</span> utimes_train[j]</span>
<span id="cb44-16"><a href="#cb44-16" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb44-17"><a href="#cb44-17" aria-hidden="true" tabindex="-1"></a>  AUC_train[j] <span class="ot">&lt;-</span> out<span class="sc">$</span>AUC</span>
<span id="cb44-18"><a href="#cb44-18" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb44-19"><a href="#cb44-19" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (j <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(utimes_validate)) {</span>
<span id="cb44-20"><a href="#cb44-20" aria-hidden="true" tabindex="-1"></a>  out <span class="ot">&lt;-</span> risksetROC<span class="sc">::</span><span class="fu">CoxWeights</span>(</span>
<span id="cb44-21"><a href="#cb44-21" aria-hidden="true" tabindex="-1"></a>    <span class="at">marker =</span> pred_lp_validate, <span class="at">Stime =</span> y_validate[, <span class="dv">1</span>],</span>
<span id="cb44-22"><a href="#cb44-22" aria-hidden="true" tabindex="-1"></a>    <span class="at">status =</span> y_validate[, <span class="dv">2</span>], <span class="at">predict.time =</span> utimes_validate[j]</span>
<span id="cb44-23"><a href="#cb44-23" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb44-24"><a href="#cb44-24" aria-hidden="true" tabindex="-1"></a>  AUC_validate[j] <span class="ot">&lt;-</span> out<span class="sc">$</span>AUC</span>
<span id="cb44-25"><a href="#cb44-25" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb44-26"><a href="#cb44-26" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-27"><a href="#cb44-27" aria-hidden="true" tabindex="-1"></a><span class="co"># draw the time-dependent AUC from the training and validation data sets</span></span>
<span id="cb44-28"><a href="#cb44-28" aria-hidden="true" tabindex="-1"></a>dat_AUC <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(</span>
<span id="cb44-29"><a href="#cb44-29" aria-hidden="true" tabindex="-1"></a>  <span class="at">tAUC =</span> <span class="fu">c</span>(AUC_train, AUC_validate),</span>
<span id="cb44-30"><a href="#cb44-30" aria-hidden="true" tabindex="-1"></a>  <span class="at">times =</span> <span class="fu">c</span>(utimes_train, utimes_validate),</span>
<span id="cb44-31"><a href="#cb44-31" aria-hidden="true" tabindex="-1"></a>  <span class="at">group =</span> <span class="fu">c</span>(<span class="fu">rep</span>(<span class="st">"AUC_train"</span>, <span class="fu">length</span>(AUC_train)), <span class="fu">rep</span>(<span class="st">"AUC_validate"</span>, <span class="fu">length</span>(AUC_validate)))</span>
<span id="cb44-32"><a href="#cb44-32" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb44-33"><a href="#cb44-33" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(dat_AUC, <span class="fu">aes</span>(times, tAUC, <span class="at">group =</span> group, <span class="at">color =</span> group)) <span class="sc">+</span></span>
<span id="cb44-34"><a href="#cb44-34" aria-hidden="true" tabindex="-1"></a>  <span class="fu">xlab</span>(<span class="st">"Evaluation time points (year)"</span>) <span class="sc">+</span></span>
<span id="cb44-35"><a href="#cb44-35" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ylab</span>(<span class="st">"AUC"</span>) <span class="sc">+</span></span>
<span id="cb44-36"><a href="#cb44-36" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ylim</span>(<span class="fl">0.5</span>, <span class="dv">1</span>) <span class="sc">+</span></span>
<span id="cb44-37"><a href="#cb44-37" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_step</span>(<span class="at">direction =</span> <span class="st">"vh"</span>) <span class="sc">+</span></span>
<span id="cb44-38"><a href="#cb44-38" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">legend.position =</span> <span class="fu">c</span>(<span class="fl">0.7</span>, <span class="fl">0.8</span>), <span class="at">legend.title =</span> <span class="fu">element_blank</span>())</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><img src="fig/TCGA_surv_auc_lasso.png" class="img-fluid figure-img" style="width:40.0%"></p>
<figcaption class="figure-caption"><em>Time-dependent AUC based on a Lasso Cox model applied to the BRCA patients data from TCGA. The red line shows the Time-dependent AUC calculated from the 80% training data, and the green line shows the Time-dependent AUC calculated from the 20% validation data.</em></figcaption>
</figure>
</div>
<p><br></p>
<p><font size="4"> <strong>Integrated AUC</strong> </font></p>
<p>The <code>R</code> package <a href="https://CRAN.R-project.org/package=risksetROC"><strong>risksetROC</strong></a> <span class="citation" data-cites="Heagerty2005">(<a href="#ref-Heagerty2005" role="doc-biblioref">Heagerty and Zheng 2005</a>)</span> provides function <code>IntegrateAUC()</code> to estimate integrated AUC.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb45"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb45-1"><a href="#cb45-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Before computing integrated AUC, first estimate survival probabilities at unique survival times</span></span>
<span id="cb45-2"><a href="#cb45-2" aria-hidden="true" tabindex="-1"></a>surv_prob_train <span class="ot">&lt;-</span> <span class="fu">unique</span>(<span class="fu">survfit</span>(<span class="fu">Surv</span>(y_train[, <span class="dv">1</span>], y_train[, <span class="dv">2</span>]) <span class="sc">~</span> <span class="dv">1</span>)<span class="sc">$</span>surv)</span>
<span id="cb45-3"><a href="#cb45-3" aria-hidden="true" tabindex="-1"></a>surv_prob_validate <span class="ot">&lt;-</span> <span class="fu">unique</span>(<span class="fu">survfit</span>(<span class="fu">Surv</span>(y_validate[, <span class="dv">1</span>], y_validate[, <span class="dv">2</span>]) <span class="sc">~</span> <span class="dv">1</span>)<span class="sc">$</span>surv)</span>
<span id="cb45-4"><a href="#cb45-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-5"><a href="#cb45-5" aria-hidden="true" tabindex="-1"></a><span class="do">## integrated AUC (e.g. over tmax=10 years) to get concordance measure based on training data</span></span>
<span id="cb45-6"><a href="#cb45-6" aria-hidden="true" tabindex="-1"></a>(iAUC_train <span class="ot">&lt;-</span> risksetROC<span class="sc">::</span><span class="fu">IntegrateAUC</span>(AUC_train, utimes_train, surv_prob_train, <span class="at">tmax =</span> <span class="dv">10</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<pre class="border"><code>[1] 0.6279646</code></pre>
<div class="cell">
<div class="sourceCode cell-code" id="cb47"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb47-1"><a href="#cb47-1" aria-hidden="true" tabindex="-1"></a><span class="do">## integrated AUC (e.g. over tmax=10 years) to get concordance measure based on validation data</span></span>
<span id="cb47-2"><a href="#cb47-2" aria-hidden="true" tabindex="-1"></a>(iAUC_validate <span class="ot">&lt;-</span> risksetROC<span class="sc">::</span><span class="fu">IntegrateAUC</span>(AUC_validate, utimes_validate, surv_prob_validate, <span class="at">tmax =</span> <span class="dv">10</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<pre class="border"><code>[1] 0.6318253</code></pre>
<p><font size="4"> <strong>Time-dependent C-index</strong> </font></p>
<p>The C-index is not proper for <span class="math inline">\(t\)</span>-year predictions, see <span class="citation" data-cites="Blanche2019">Blanche, Kattan, and Gerds (<a href="#ref-Blanche2019" role="doc-biblioref">2019</a>)</span>. Consider using time-dependent AUC or time-dependent Brier score instead. For a time-dependent discrimination index for survival data, see <span class="citation" data-cites="Antolini2005">Antolini et al. (<a href="#ref-Antolini2005" role="doc-biblioref">2005</a>)</span>.</p>
<p><font size="4"> <strong>C-index</strong> </font></p>
<p>The <code>R</code> package <a href="https://CRAN.R-project.org/package=glmnet"><strong>glmnet</strong></a> provides the function <code>glmnet::Cindex()</code> to estimate Harrell’s C-index from a “coxnet” object. The <code>R</code> package <a href="https://CRAN.R-project.org/package=survAUC"><strong>survAUC</strong></a> provides the function <code>survAUC::UnoC()</code> to estimated Uno’s C-index. See an example calculation for both C-indexes using a Lasso Cox model below.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb49"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb49-1"><a href="#cb49-1" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">123</span>)</span>
<span id="cb49-2"><a href="#cb49-2" aria-hidden="true" tabindex="-1"></a>cvfit <span class="ot">&lt;-</span> <span class="fu">cv.glmnet</span>(x_train, y_train, <span class="at">family =</span> <span class="st">"cox"</span>, <span class="at">nfolds =</span> <span class="dv">5</span>, <span class="at">penalty.factor =</span> pf)</span>
<span id="cb49-3"><a href="#cb49-3" aria-hidden="true" tabindex="-1"></a>pred <span class="ot">&lt;-</span> <span class="fu">predict</span>(cvfit, <span class="at">newx =</span> x_validate, <span class="at">type =</span> <span class="st">"link"</span>, <span class="at">s =</span> cvfit<span class="sc">$</span>lambda.min)</span>
<span id="cb49-4"><a href="#cb49-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Harrell's C-index</span></span>
<span id="cb49-5"><a href="#cb49-5" aria-hidden="true" tabindex="-1"></a>(Cindex_Harrell <span class="ot">&lt;-</span> <span class="fu">Cindex</span>(<span class="at">pred =</span> pred[, <span class="dv">1</span>], <span class="at">y =</span> y_validate))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<pre class="border"><code>[1] 0.7246466</code></pre>
<div class="cell">
<div class="sourceCode cell-code" id="cb51"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb51-1"><a href="#cb51-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Uno's C-index</span></span>
<span id="cb51-2"><a href="#cb51-2" aria-hidden="true" tabindex="-1"></a>(Cindex_Uno <span class="ot">&lt;-</span> survAUC<span class="sc">::</span><span class="fu">UnoC</span>(y_train, y_validate, pred))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<pre class="border"><code>[1] 0.5772041</code></pre>
<p><br></p>
</section>
<section id="calibration-metrics" class="level4 unnumbered">
<h4 class="unnumbered anchored" data-anchor-id="calibration-metrics">Calibration metrics</h4>
<p>See a <a href="#slopeCali">calibration plot</a> in the following section <a href="#graphComp">Graphical computation</a>.</p>
<p><br></p>
</section>
<section id="overall-metrics" class="level4 unnumbered">
<h4 class="unnumbered anchored" data-anchor-id="overall-metrics">Overall metrics</h4>
<p><font size="4"> <strong>Time-dependent Brier score</strong> </font></p>
<p>The <code>R</code> package <a href="https://CRAN.R-project.org/package=riskRegression"><strong>riskRegression</strong></a> can assess the prediction error curves of survival models based on the time-dependent Brier score. Similar to the time-dependent AUC, one needs to first calculate the linear predictors (<span class="math inline">\(lp\)</span>) from a frequentist or Bayesian Cox model, and then use <code>survival::coxph()</code> to regress the survival outcomes on the linear predictor, which is prepared as input of <code>riskRegression::Score()</code> to estimate the (time-dependent) Brier score. If the survival model was fitted via the <code>R</code> package <a href="#psbcSpeedUp"><strong>psbcSpeedUp</strong></a>, the curve of the time-dependent Brier score can be drawn by using the function <code>psbcSpeedUp::plotBrier()</code>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb53"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb53-1"><a href="#cb53-1" aria-hidden="true" tabindex="-1"></a><span class="do">## time-dependent Brier score</span></span>
<span id="cb53-2"><a href="#cb53-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-3"><a href="#cb53-3" aria-hidden="true" tabindex="-1"></a><span class="co"># use the (x_train, y_train) 80% samples for training</span></span>
<span id="cb53-4"><a href="#cb53-4" aria-hidden="true" tabindex="-1"></a><span class="co"># and the (x_validate, y_validate) 20% samples for testing</span></span>
<span id="cb53-5"><a href="#cb53-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-6"><a href="#cb53-6" aria-hidden="true" tabindex="-1"></a>y_train_surv <span class="ot">&lt;-</span> <span class="fu">Surv</span>(y_train[, <span class="st">"time"</span>], y_train[, <span class="st">"status"</span>])</span>
<span id="cb53-7"><a href="#cb53-7" aria-hidden="true" tabindex="-1"></a>y_validate_surv <span class="ot">&lt;-</span> <span class="fu">Surv</span>(y_validate[, <span class="st">"time"</span>], y_validate[, <span class="st">"status"</span>])</span>
<span id="cb53-8"><a href="#cb53-8" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">123</span>)</span>
<span id="cb53-9"><a href="#cb53-9" aria-hidden="true" tabindex="-1"></a>cvfit <span class="ot">&lt;-</span> <span class="fu">cv.glmnet</span>(x_train, y_train_surv, <span class="at">family =</span> <span class="st">"cox"</span>, <span class="at">nfolds =</span> <span class="dv">5</span>, <span class="at">penalty.factor =</span> pf)</span>
<span id="cb53-10"><a href="#cb53-10" aria-hidden="true" tabindex="-1"></a>lp_train <span class="ot">&lt;-</span> <span class="fu">predict</span>(cvfit, <span class="at">newx =</span> x_train, <span class="at">s =</span> cvfit<span class="sc">$</span>lambda.min, <span class="at">type =</span> <span class="st">"link"</span>)</span>
<span id="cb53-11"><a href="#cb53-11" aria-hidden="true" tabindex="-1"></a>lp_validate <span class="ot">&lt;-</span> <span class="fu">predict</span>(cvfit, <span class="at">newx =</span> x_validate, <span class="at">s =</span> cvfit<span class="sc">$</span>lambda.min, <span class="at">type =</span> <span class="st">"link"</span>)</span>
<span id="cb53-12"><a href="#cb53-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-13"><a href="#cb53-13" aria-hidden="true" tabindex="-1"></a><span class="co"># prepare data format suited for function Score() from the riskRegression package</span></span>
<span id="cb53-14"><a href="#cb53-14" aria-hidden="true" tabindex="-1"></a>data_train <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">time =</span> y_train[, <span class="st">"time"</span>], <span class="at">status =</span> y_train[, <span class="st">"status"</span>], <span class="at">lp =</span> <span class="fu">as.vector</span>(lp_train))</span>
<span id="cb53-15"><a href="#cb53-15" aria-hidden="true" tabindex="-1"></a>data_validate <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">time =</span> y_validate[, <span class="st">"time"</span>], <span class="at">status =</span> y_validate[, <span class="st">"status"</span>], <span class="at">lp =</span> <span class="fu">as.vector</span>(lp_validate))</span>
<span id="cb53-16"><a href="#cb53-16" aria-hidden="true" tabindex="-1"></a>lasso_train <span class="ot">&lt;-</span> <span class="fu">coxph</span>(<span class="fu">Surv</span>(time, status) <span class="sc">~</span> lp, <span class="at">data =</span> data_train, <span class="at">y =</span> <span class="cn">TRUE</span>, <span class="at">x =</span> <span class="cn">TRUE</span>)</span>
<span id="cb53-17"><a href="#cb53-17" aria-hidden="true" tabindex="-1"></a>lasso_validate <span class="ot">&lt;-</span> <span class="fu">coxph</span>(<span class="fu">Surv</span>(time, status) <span class="sc">~</span> lp, <span class="at">data =</span> data_validate, <span class="at">y =</span> <span class="cn">TRUE</span>, <span class="at">x =</span> <span class="cn">TRUE</span>)</span>
<span id="cb53-18"><a href="#cb53-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-19"><a href="#cb53-19" aria-hidden="true" tabindex="-1"></a><span class="co"># calculate Brier scores based on both training and validation data</span></span>
<span id="cb53-20"><a href="#cb53-20" aria-hidden="true" tabindex="-1"></a>Brier_train <span class="ot">&lt;-</span> riskRegression<span class="sc">::</span><span class="fu">Score</span>(<span class="fu">list</span>(<span class="st">"Brier_train"</span> <span class="ot">=</span> lasso_train), <span class="at">formula =</span> <span class="fu">Surv</span>(time, status) <span class="sc">~</span> <span class="dv">1</span>, <span class="at">data =</span> data_train, <span class="at">conf.int =</span> <span class="cn">FALSE</span>, <span class="at">metrics =</span> <span class="st">"brier"</span>, <span class="at">summary =</span> <span class="st">"ibs"</span>, <span class="at">times =</span> <span class="fu">sort</span>(<span class="fu">unique</span>(data_train<span class="sc">$</span>time)))<span class="sc">$</span>Brier<span class="sc">$</span>score</span>
<span id="cb53-21"><a href="#cb53-21" aria-hidden="true" tabindex="-1"></a>Brier_validate <span class="ot">&lt;-</span> riskRegression<span class="sc">::</span><span class="fu">Score</span>(<span class="fu">list</span>(<span class="st">"Brier_validate"</span> <span class="ot">=</span> lasso_validate), <span class="at">formula =</span> <span class="fu">Surv</span>(time, status) <span class="sc">~</span> <span class="dv">1</span>, <span class="at">data =</span> data_validate, <span class="at">conf.int =</span> <span class="cn">FALSE</span>, <span class="at">metrics =</span> <span class="st">"brier"</span>, <span class="at">summary =</span> <span class="st">"ibs"</span>, <span class="at">times =</span> <span class="fu">sort</span>(<span class="fu">unique</span>(data_validate<span class="sc">$</span>time)))<span class="sc">$</span>Brier<span class="sc">$</span>score</span>
<span id="cb53-22"><a href="#cb53-22" aria-hidden="true" tabindex="-1"></a>Brier_score <span class="ot">&lt;-</span> <span class="fu">rbind</span>(Brier_train, Brier_validate)</span>
<span id="cb53-23"><a href="#cb53-23" aria-hidden="true" tabindex="-1"></a>Brier_score <span class="ot">&lt;-</span> Brier_score[Brier_score<span class="sc">$</span>model <span class="sc">!=</span> <span class="st">"Null model"</span>, ]</span>
<span id="cb53-24"><a href="#cb53-24" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-25"><a href="#cb53-25" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(Brier_score, <span class="fu">aes</span>(times, Brier, <span class="at">group =</span> model, <span class="at">color =</span> model)) <span class="sc">+</span></span>
<span id="cb53-26"><a href="#cb53-26" aria-hidden="true" tabindex="-1"></a>  <span class="fu">xlab</span>(<span class="st">"Evaluation time points (year)"</span>) <span class="sc">+</span></span>
<span id="cb53-27"><a href="#cb53-27" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ylab</span>(<span class="st">"Brier score"</span>) <span class="sc">+</span></span>
<span id="cb53-28"><a href="#cb53-28" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_step</span>(<span class="at">direction =</span> <span class="st">"vh"</span>) <span class="sc">+</span></span>
<span id="cb53-29"><a href="#cb53-29" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">legend.position =</span> <span class="fu">c</span>(<span class="fl">0.15</span>, <span class="fl">0.88</span>), <span class="at">legend.title =</span> <span class="fu">element_blank</span>())</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><img src="fig/TCGA_surv_brier_t_lasso.png" class="img-fluid figure-img" style="width:60.0%"></p>
<figcaption class="figure-caption"><em>Time-dependent Brier score based on a Lasso Cox model applied to the BRCA patients data from TCGA. The red line shows the Time-dependent Brier score calculated from the 80% training data, and the green line shows the Time-dependent Brier score calculated from the 20% validation data.</em></figcaption>
</figure>
</div>
<p><br></p>
<p><font size="4"> <strong>Integrated Brier score (IBS)</strong> </font></p>
<p>The function <code>riskRegression::Score()</code> also summarizes IBS when specifying argument <code>summary = "ibs"</code>. We can extract the IBS corresponding to the largest evaluation time point.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb54"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb54-1"><a href="#cb54-1" aria-hidden="true" tabindex="-1"></a>Brier_validate_ibs <span class="ot">&lt;-</span> Brier_validate[Brier_validate<span class="sc">$</span>model <span class="sc">==</span> <span class="st">"Brier_validate"</span>, ]</span>
<span id="cb54-2"><a href="#cb54-2" aria-hidden="true" tabindex="-1"></a>Brier_validate_ibs<span class="sc">$</span>IBS[<span class="fu">which.max</span>(Brier_validate_ibs<span class="sc">$</span>times)]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<pre class="border"><code>[1] 0.1721158</code></pre>
<p><br></p>
</section>
<section id="uq1" class="level4 unnumbered">
<h4 class="unnumbered anchored" data-anchor-id="uq1">Uncertainty Quantification</h4>
<div class="callout callout-style-default callout-important callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Important
</div>
</div>
<div class="callout-body-container callout-body">
<p><strong>It is recommended to use resampling-based methods</strong> for estimating the uncertainty of the model’s performance, if there are no <strong>independent</strong> validation data for model evaluation. This can be done for example by repeatedly splitting the dataset to training/validation sets and evaluating a model’s performance on the different validation sets using various discrimination or calibration metrics.</p>
</div>
</div>
<p>We demonstrate how to randomly split the data, e.g.&nbsp;<span class="math inline">\(100\)</span> times, train a Lasso Cox model and estimate the integrated AUC based on the validation data in each replication. For other Cox-type models, we can just replace the model fitting part <code>cv.glmnet()</code> (and <code>predict()</code>) in the <code>for</code> loop below. However, most of the Bayesian Cox models introduced previously are computationally time-consuming when randomly splitting the data many times.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb56"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb56-1"><a href="#cb56-1" aria-hidden="true" tabindex="-1"></a><span class="co"># split the data 100 times</span></span>
<span id="cb56-2"><a href="#cb56-2" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">123</span>)</span>
<span id="cb56-3"><a href="#cb56-3" aria-hidden="true" tabindex="-1"></a>k <span class="ot">&lt;-</span> <span class="dv">100</span></span>
<span id="cb56-4"><a href="#cb56-4" aria-hidden="true" tabindex="-1"></a>iAUC <span class="ot">&lt;-</span> <span class="fu">rep</span>(<span class="cn">NA</span>, k)</span>
<span id="cb56-5"><a href="#cb56-5" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span>k) {</span>
<span id="cb56-6"><a href="#cb56-6" aria-hidden="true" tabindex="-1"></a>  idx <span class="ot">&lt;-</span> <span class="fu">sample</span>(<span class="dv">1</span><span class="sc">:</span>n, n <span class="sc">*</span> <span class="fl">0.8</span>, <span class="at">replace =</span> <span class="cn">FALSE</span>)</span>
<span id="cb56-7"><a href="#cb56-7" aria-hidden="true" tabindex="-1"></a>  x_train <span class="ot">&lt;-</span> x[idx, ]</span>
<span id="cb56-8"><a href="#cb56-8" aria-hidden="true" tabindex="-1"></a>  y_train <span class="ot">&lt;-</span> y[idx, ]</span>
<span id="cb56-9"><a href="#cb56-9" aria-hidden="true" tabindex="-1"></a>  x_validate <span class="ot">&lt;-</span> x[<span class="sc">-</span>idx, ]</span>
<span id="cb56-10"><a href="#cb56-10" aria-hidden="true" tabindex="-1"></a>  y_validate <span class="ot">&lt;-</span> y[<span class="sc">-</span>idx, ]</span>
<span id="cb56-11"><a href="#cb56-11" aria-hidden="true" tabindex="-1"></a>  cvfit <span class="ot">&lt;-</span> <span class="fu">cv.glmnet</span>(x_train, y_train, <span class="at">family =</span> <span class="st">"cox"</span>, <span class="at">nfolds =</span> <span class="dv">5</span>, <span class="at">penalty.factor =</span> pf)</span>
<span id="cb56-12"><a href="#cb56-12" aria-hidden="true" tabindex="-1"></a>  pred_lp <span class="ot">&lt;-</span> <span class="fu">predict</span>(cvfit, <span class="at">newx =</span> x_validate, <span class="at">s =</span> cvfit<span class="sc">$</span>lambda.min, <span class="at">type =</span> <span class="st">"link"</span>)</span>
<span id="cb56-13"><a href="#cb56-13" aria-hidden="true" tabindex="-1"></a>  utimes <span class="ot">&lt;-</span> <span class="fu">sort</span>(<span class="fu">unique</span>(y_validate[y_validate[, <span class="dv">2</span>] <span class="sc">==</span> <span class="dv">1</span>, <span class="dv">1</span>]))</span>
<span id="cb56-14"><a href="#cb56-14" aria-hidden="true" tabindex="-1"></a>  AUC <span class="ot">&lt;-</span> <span class="fu">rep</span>(<span class="cn">NA</span>, <span class="fu">length</span>(utimes))</span>
<span id="cb56-15"><a href="#cb56-15" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span> (j <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(utimes)) {</span>
<span id="cb56-16"><a href="#cb56-16" aria-hidden="true" tabindex="-1"></a>    out <span class="ot">&lt;-</span> <span class="fu">CoxWeights</span>(<span class="at">marker =</span> pred_lp, <span class="at">Stime =</span> y_validate[, <span class="dv">1</span>], <span class="at">status =</span> y_validate[, <span class="dv">2</span>], <span class="at">predict.time =</span> utimes[j])</span>
<span id="cb56-17"><a href="#cb56-17" aria-hidden="true" tabindex="-1"></a>    AUC[j] <span class="ot">&lt;-</span> out<span class="sc">$</span>AUC</span>
<span id="cb56-18"><a href="#cb56-18" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb56-19"><a href="#cb56-19" aria-hidden="true" tabindex="-1"></a>  surv_prob <span class="ot">&lt;-</span> <span class="fu">unique</span>(<span class="fu">survfit</span>(<span class="fu">Surv</span>(y_validate[, <span class="dv">1</span>], y_validate[, <span class="dv">2</span>]) <span class="sc">~</span> <span class="dv">1</span>)<span class="sc">$</span>surv)</span>
<span id="cb56-20"><a href="#cb56-20" aria-hidden="true" tabindex="-1"></a>  iAUC[i] <span class="ot">&lt;-</span> <span class="fu">IntegrateAUC</span>(AUC, utimes, surv_prob, <span class="at">tmax =</span> <span class="dv">10</span>)</span>
<span id="cb56-21"><a href="#cb56-21" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb56-22"><a href="#cb56-22" aria-hidden="true" tabindex="-1"></a>dat_tmp <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">x =</span> <span class="st">"Lasso Cox"</span>, <span class="at">y =</span> iAUC)</span>
<span id="cb56-23"><a href="#cb56-23" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb56-24"><a href="#cb56-24" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">123</span>)</span>
<span id="cb56-25"><a href="#cb56-25" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(dat_tmp, <span class="fu">aes</span>(x, y)) <span class="sc">+</span></span>
<span id="cb56-26"><a href="#cb56-26" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_boxplot</span>() <span class="sc">+</span></span>
<span id="cb56-27"><a href="#cb56-27" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ylim</span>(<span class="fl">0.5</span>, <span class="dv">1</span>) <span class="sc">+</span></span>
<span id="cb56-28"><a href="#cb56-28" aria-hidden="true" tabindex="-1"></a>  <span class="fu">xlab</span>(<span class="st">""</span>) <span class="sc">+</span></span>
<span id="cb56-29"><a href="#cb56-29" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ylab</span>(<span class="st">"Integrated AUC"</span>) <span class="sc">+</span></span>
<span id="cb56-30"><a href="#cb56-30" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_jitter</span>(<span class="at">color =</span> <span class="st">"blue"</span>, <span class="at">size =</span> <span class="fl">0.5</span>, <span class="at">alpha =</span> <span class="fl">0.5</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><img src="fig/TCGA_surv_iauc_lasso.png" class="img-fluid figure-img" style="width:30.0%"></p>
<figcaption class="figure-caption"><em>Integrated AUC based on randomly split validation data 100 times. The blue dots are the 100 values of integrated AUC.</em></figcaption>
</figure>
</div>
<p><br></p>
<p>Similar to obtaining uncertainty of the integrated AUC, we can also estimate the uncertainty of the C-index for evaluating the global performance of our model’s discrimination.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb57"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb57-1"><a href="#cb57-1" aria-hidden="true" tabindex="-1"></a><span class="co"># split the data 100 times</span></span>
<span id="cb57-2"><a href="#cb57-2" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">123</span>)</span>
<span id="cb57-3"><a href="#cb57-3" aria-hidden="true" tabindex="-1"></a>k <span class="ot">&lt;-</span> <span class="dv">100</span></span>
<span id="cb57-4"><a href="#cb57-4" aria-hidden="true" tabindex="-1"></a>Cindex_all <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">Harrell =</span> <span class="fu">rep</span>(<span class="cn">NA</span>, k), <span class="at">Uno =</span> <span class="fu">rep</span>(<span class="cn">NA</span>, k))</span>
<span id="cb57-5"><a href="#cb57-5" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span>k) {</span>
<span id="cb57-6"><a href="#cb57-6" aria-hidden="true" tabindex="-1"></a>  idx <span class="ot">&lt;-</span> <span class="fu">sample</span>(<span class="dv">1</span><span class="sc">:</span>n, n <span class="sc">*</span> <span class="fl">0.8</span>, <span class="at">replace =</span> <span class="cn">FALSE</span>)</span>
<span id="cb57-7"><a href="#cb57-7" aria-hidden="true" tabindex="-1"></a>  x_train <span class="ot">&lt;-</span> x[idx, ]</span>
<span id="cb57-8"><a href="#cb57-8" aria-hidden="true" tabindex="-1"></a>  y_train <span class="ot">&lt;-</span> y[idx, ]</span>
<span id="cb57-9"><a href="#cb57-9" aria-hidden="true" tabindex="-1"></a>  x_validate <span class="ot">&lt;-</span> x[<span class="sc">-</span>idx, ]</span>
<span id="cb57-10"><a href="#cb57-10" aria-hidden="true" tabindex="-1"></a>  y_validate <span class="ot">&lt;-</span> y[<span class="sc">-</span>idx, ]</span>
<span id="cb57-11"><a href="#cb57-11" aria-hidden="true" tabindex="-1"></a>  cvfit <span class="ot">&lt;-</span> <span class="fu">cv.glmnet</span>(x_train, y_train, <span class="at">family =</span> <span class="st">"cox"</span>, <span class="at">nfolds =</span> <span class="dv">5</span>, <span class="at">penalty.factor =</span> pf)</span>
<span id="cb57-12"><a href="#cb57-12" aria-hidden="true" tabindex="-1"></a>  pred <span class="ot">&lt;-</span> <span class="fu">predict</span>(cvfit, <span class="at">newx =</span> x_validate, <span class="at">type =</span> <span class="st">"response"</span>, <span class="at">s =</span> cvfit<span class="sc">$</span>lambda.min)</span>
<span id="cb57-13"><a href="#cb57-13" aria-hidden="true" tabindex="-1"></a>  Cindex_all<span class="sc">$</span>Harrell[i] <span class="ot">&lt;-</span> <span class="fu">mean</span>(<span class="fu">apply</span>(pred, <span class="dv">2</span>, Cindex, <span class="at">y =</span> y_validate))</span>
<span id="cb57-14"><a href="#cb57-14" aria-hidden="true" tabindex="-1"></a>  Cindex_all<span class="sc">$</span>Uno[i] <span class="ot">&lt;-</span> <span class="fu">UnoC</span>(y_train, y_validate, pred)</span>
<span id="cb57-15"><a href="#cb57-15" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb57-16"><a href="#cb57-16" aria-hidden="true" tabindex="-1"></a>dat_tmp <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">x =</span> <span class="fu">rep</span>(<span class="fu">c</span>(<span class="st">"Harrell"</span>, <span class="st">"Uno"</span>), <span class="at">each =</span> k), <span class="at">y =</span> <span class="fu">unlist</span>(Cindex_all))</span>
<span id="cb57-17"><a href="#cb57-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-18"><a href="#cb57-18" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">123</span>)</span>
<span id="cb57-19"><a href="#cb57-19" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(dat_tmp, <span class="fu">aes</span>(x, y, <span class="at">col =</span> x)) <span class="sc">+</span></span>
<span id="cb57-20"><a href="#cb57-20" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_boxplot</span>() <span class="sc">+</span></span>
<span id="cb57-21"><a href="#cb57-21" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_jitter</span>(<span class="at">size =</span> <span class="fl">0.5</span>, <span class="at">alpha =</span> <span class="fl">0.5</span>) <span class="sc">+</span></span>
<span id="cb57-22"><a href="#cb57-22" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ylim</span>(<span class="dv">0</span>, <span class="dv">1</span>) <span class="sc">+</span></span>
<span id="cb57-23"><a href="#cb57-23" aria-hidden="true" tabindex="-1"></a>  <span class="fu">xlab</span>(<span class="st">""</span>) <span class="sc">+</span></span>
<span id="cb57-24"><a href="#cb57-24" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ylab</span>(<span class="st">"C-index"</span>) <span class="sc">+</span></span>
<span id="cb57-25"><a href="#cb57-25" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">legend.position =</span> <span class="st">"none"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><img src="fig/TCGA_surv_cindex_lasso.png" class="img-fluid figure-img" style="width:40.0%"></p>
<figcaption class="figure-caption"><em>C-index (Harrell’s and Uno’s) based on randomly split validation data 100 times.</em></figcaption>
</figure>
</div>
<p><br></p>
<p>The <code>R</code> package <a href="https://CRAN.R-project.org/package=c060"><strong>c060</strong></a> <span class="citation" data-cites="Sill2014">(<a href="#ref-Sill2014" role="doc-biblioref">Sill et al. 2014</a>)</span> includes wrapper functions for the <a href="https://CRAN.R-project.org/package=glmnet"><strong>glmnet</strong></a> algorithm and implements resampling-based methods (e.g.&nbsp;cross-validation and bootstrap - with and without replacement) based on the <a href="https://CRAN.R-project.org/package=peperr"><strong>peperr</strong></a> package to calculate the time-dependent Brier score. <a href="https://CRAN.R-project.org/package=c060"><strong>c060</strong></a> extends <a href="https://CRAN.R-project.org/package=peperr"><strong>peperr</strong></a> package to allow mandatory features without penalization. <span class="citation" data-cites="Binder2008">Binder and Schumacher (<a href="#ref-Binder2008" role="doc-biblioref">2008</a>)</span> recommends to draw bootstrap samples without replacement (i.e.&nbsp;subsampling), because bootstrap samples with replacement often result in too complex models in high-dimensional settings. To use resampling by CV properly for survival data, see <span class="citation" data-cites="Simon2011">R. M. Simon et al. (<a href="#ref-Simon2011" role="doc-biblioref">2011</a>)</span>. Note that resampling-based methods here are similar to splitting <span class="math inline">\(80\%/20\%\)</span> the data many times which allows us to quantify the uncertainty of the time-dependent Brier score.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb58"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb58-1"><a href="#cb58-1" aria-hidden="true" tabindex="-1"></a><span class="do">## time-dependent Brier score by subsampling from the whole data</span></span>
<span id="cb58-2"><a href="#cb58-2" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">123</span>)</span>
<span id="cb58-3"><a href="#cb58-3" aria-hidden="true" tabindex="-1"></a>peperr_object <span class="ot">&lt;-</span> peperr<span class="sc">::</span><span class="fu">peperr</span>(</span>
<span id="cb58-4"><a href="#cb58-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">response =</span> y_surv, <span class="at">x =</span> x, <span class="at">fit.fun =</span> fit.glmnet,</span>
<span id="cb58-5"><a href="#cb58-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">args.fit =</span> <span class="fu">list</span>(<span class="at">family =</span> <span class="st">"cox"</span>, <span class="at">penalty.factor =</span> pf),</span>
<span id="cb58-6"><a href="#cb58-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">complexity =</span> complexity.glmnet,</span>
<span id="cb58-7"><a href="#cb58-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">args.complexity =</span> <span class="fu">list</span>(<span class="at">family =</span> <span class="st">"cox"</span>, <span class="at">nfolds =</span> <span class="dv">5</span>, <span class="at">penalty.factor =</span> pf),</span>
<span id="cb58-8"><a href="#cb58-8" aria-hidden="true" tabindex="-1"></a>  <span class="at">indices =</span> <span class="fu">resample.indices</span>(<span class="at">n =</span> n, <span class="at">method =</span> <span class="st">"sub632"</span>, <span class="at">sample.n =</span> <span class="dv">100</span>)</span>
<span id="cb58-9"><a href="#cb58-9" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb58-10"><a href="#cb58-10" aria-hidden="true" tabindex="-1"></a>c060<span class="sc">::</span><span class="fu">Plot.peperr.curves</span>(peperr_object)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><img src="fig/TCGA_surv_brier_lasso.png" class="img-fluid figure-img" style="width:50.0%"></p>
<figcaption class="figure-caption"><em>Resampling-based prediction error curves (time-dependent Brier score) a the Lasso Cox model applied to the BRCA data set from TCGA. The gray area indicates the pointwise 2.5% and 97.5% quantiles of the 100 out-of-bag bootstrap samples. The other lines show the prediction error curves of the null model (estimated by the Kaplan-Meier estimator without covariate information), the full apparent error estimates (i.e., the errors as estimated when applying the model to the entire training data set), and the .632+ bootstrap error estimates.</em></figcaption>
</figure>
</div>
<p><br></p>
</section>
<section id="feature-stability-analysis" class="level4 unnumbered">
<h4 class="unnumbered anchored" data-anchor-id="feature-stability-analysis">Feature stability analysis</h4>
<p>To identify stable omics features, a straightforward way is to find the overlapped omics features with nonzero coefficients among different data subsets (e.g.&nbsp;CV folds or resamples). The following code summarizes the Lasso Cox selected omics features which were identified at least <span class="math inline">\(2\)</span> or <span class="math inline">\(5\)</span> out of <span class="math inline">\(10\)</span> resamples. Similarly, this approach can be applied to other Lasso-type or Bayesian Cox models that perform feature selection for identifying stable selected features.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb59"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb59-1"><a href="#cb59-1" aria-hidden="true" tabindex="-1"></a><span class="co"># specify the number of resamples k</span></span>
<span id="cb59-2"><a href="#cb59-2" aria-hidden="true" tabindex="-1"></a>k <span class="ot">&lt;-</span> <span class="dv">10</span></span>
<span id="cb59-3"><a href="#cb59-3" aria-hidden="true" tabindex="-1"></a>beta_all <span class="ot">&lt;-</span> <span class="fu">matrix</span>(<span class="at">nrow =</span> <span class="fu">ncol</span>(x), <span class="at">ncol =</span> k)</span>
<span id="cb59-4"><a href="#cb59-4" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">123</span>)</span>
<span id="cb59-5"><a href="#cb59-5" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (j <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span>k) {</span>
<span id="cb59-6"><a href="#cb59-6" aria-hidden="true" tabindex="-1"></a>  resample_id <span class="ot">&lt;-</span> <span class="fu">sample</span>(<span class="dv">1</span><span class="sc">:</span><span class="fu">nrow</span>(y), <span class="fu">nrow</span>(y), <span class="at">replace =</span> <span class="cn">TRUE</span>)</span>
<span id="cb59-7"><a href="#cb59-7" aria-hidden="true" tabindex="-1"></a>  resample_x <span class="ot">&lt;-</span> x[resample_id, ]</span>
<span id="cb59-8"><a href="#cb59-8" aria-hidden="true" tabindex="-1"></a>  resample_y <span class="ot">&lt;-</span> y[resample_id, ]</span>
<span id="cb59-9"><a href="#cb59-9" aria-hidden="true" tabindex="-1"></a>  cvfit <span class="ot">&lt;-</span> <span class="fu">cv.glmnet</span>(resample_x, resample_y, <span class="at">family =</span> <span class="st">"cox"</span>, <span class="at">nfolds =</span> <span class="dv">5</span>, <span class="at">penalty.factor =</span> pf)</span>
<span id="cb59-10"><a href="#cb59-10" aria-hidden="true" tabindex="-1"></a>  beta_all[, j] <span class="ot">&lt;-</span> <span class="fu">as.vector</span>(<span class="fu">coef</span>(cvfit, <span class="at">s =</span> cvfit<span class="sc">$</span>lambda.min))</span>
<span id="cb59-11"><a href="#cb59-11" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb59-12"><a href="#cb59-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb59-13"><a href="#cb59-13" aria-hidden="true" tabindex="-1"></a>(stable_features <span class="ot">&lt;-</span> <span class="fu">colnames</span>(x)[<span class="fu">rowSums</span>(beta_all <span class="sc">!=</span> <span class="dv">0</span>) <span class="sc">&gt;=</span> <span class="dv">2</span>])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<pre class="border"><code> [1] "age"       "ethnicity" "ANLN"      "UBE2T"     "NDC80"     "PGR"       "ORC6"     
 [8] "ESR1"      "PHGDH"     "MMP11"     "SFRP1"     "CCNE1"     "BLVRA"     "BAG1"     
[15] "MLPH"      "CENPF"     "KRT17"     "FOXA1"     "ACTR3B"    "CCNB1"     "MDM2"     
[22] "MYC"       "CEP55"     "SLC39A6"   "GRB7"      "NUF2"      "EGFR"      "MKI67"    
[29] "TMEM45B"   "FGFR4"     "MELK"      "NAT1"      "CXXC5"     "BCL2"      "GPR160"   
[36] "TYMS"      "KRT5"      "MAPT"      "MIA"</code></pre>
<div class="cell">
<div class="sourceCode cell-code" id="cb61"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb61-1"><a href="#cb61-1" aria-hidden="true" tabindex="-1"></a>(stable_features <span class="ot">&lt;-</span> <span class="fu">colnames</span>(x)[<span class="fu">rowSums</span>(beta_all <span class="sc">!=</span> <span class="dv">0</span>) <span class="sc">&gt;=</span> <span class="dv">5</span>])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<pre class="border"><code> [1] "age"       "ethnicity" "ANLN"      "ORC6"      "MMP11"     "BLVRA"     "BAG1"     
 [8] "CCNB1"     "EGFR"      "TMEM45B"   "BCL2"      "TYMS"      "KRT5"      "MIA"</code></pre>
<p>Alternatively for a Bayesian Cox model, its median probability model (MPM) can be obtained based on the coefficient estimates over MCMC iterations. The following code shows how to obtain the MPM’s coefficients of the penalized semiparametric Bayesian Cox model with Elastic Net prior run previously.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb63"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb63-1"><a href="#cb63-1" aria-hidden="true" tabindex="-1"></a>gammas <span class="ot">&lt;-</span> <span class="fu">colMeans</span>(<span class="fu">matrix</span>(<span class="fu">as.numeric</span>(EN_beta_p <span class="sc">!=</span> <span class="dv">0</span>), <span class="at">ncol =</span> <span class="fu">ncol</span>(EN_beta_p)))</span>
<span id="cb63-2"><a href="#cb63-2" aria-hidden="true" tabindex="-1"></a>beta_MPM <span class="ot">&lt;-</span> (gammas <span class="sc">&gt;=</span> <span class="fl">0.5</span>) <span class="sc">*</span> <span class="fu">colMeans</span>(EN_beta_p) <span class="sc">/</span> gammas</span>
<span id="cb63-3"><a href="#cb63-3" aria-hidden="true" tabindex="-1"></a>beta_MPM[<span class="fu">is.na</span>(beta_MPM)] <span class="ot">&lt;-</span> <span class="dv">0</span></span>
<span id="cb63-4"><a href="#cb63-4" aria-hidden="true" tabindex="-1"></a>beta_MPM</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<pre class="border"><code>          age     ethnicity          ANLN         FOXC1          CDH3         UBE2T 
 1.305162e-02  5.348458e-03 -1.299443e-03 -1.857811e-02 -6.123574e-03 -5.467111e-03 
        NDC80           PGR         BIRC5          ORC6          ESR1         PHGDH 
-6.652927e-03 -2.101243e-06 -1.640386e-02 -1.237153e-02 -1.077863e-02  2.483990e-02 
         CDC6         MMP11         MYBL2         SFRP1         CCNE1         BLVRA 
-9.079708e-03 -1.588726e-02  5.225344e-03 -1.383981e-02 -3.181265e-03 -2.632373e-02 
         BAG1          MLPH         CDC20         CENPF         KRT17         FOXA1 
-3.913529e-02 -1.435805e-02 -2.027232e-02 -2.476495e-02 -2.871143e-02 -3.017213e-03 
       ACTR3B         CCNB1          MDM2           MYC         CEP55       SLC39A6 
-2.504869e-03 -1.346817e-03 -2.156041e-02  1.431062e-02  1.421036e-02 -1.150196e-02 
        ERBB2          GRB7         KIF2C          NUF2          EGFR         MKI67 
-6.347367e-03 -1.008689e-02  6.033792e-03 -2.405689e-03 -1.964927e-02  1.956661e-02 
      TMEM45B         FGFR4         PTTG1          MELK          NAT1         CXXC5 
 2.736216e-02  1.842323e-03 -5.651905e-03  2.894074e-02 -2.126163e-02  2.571472e-02 
         BCL2          RRM2        GPR160          EXO1         UBE2C          TYMS 
-5.140894e-03  2.881004e-02 -3.927705e-02 -1.710419e-02 -1.343832e-02 -1.884342e-02 
         KRT5         KRT14          MAPT           MIA 
-2.180294e-02 -1.386489e-03 -2.587557e-02 -1.033317e-02</code></pre>
<p><br></p>
</section>
<section id="graphComp" class="level4 unnumbered">
<h4 class="unnumbered anchored" data-anchor-id="graphComp">Graphical representation</h4>
<p>After identifying stable omics features predictive of survival outcomes, we can draw a <strong>nomogram</strong> to allows the graphical calculation of survival probabilities and report a <strong>calibration plot</strong> for practitioners.</p>
<p><font size="4"> <strong>Nomogram</strong> </font></p>
<p>We demonstrate a nomogram using the stable selected features from TCGA breast cancer data preprocessed previously. The <code>R</code> package <strong>regplot</strong> draws an enhanced regression nomogram based on the <strong>rms</strong> package.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb65"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb65-1"><a href="#cb65-1" aria-hidden="true" tabindex="-1"></a><span class="co"># remove patients without reporting ethnicity</span></span>
<span id="cb65-2"><a href="#cb65-2" aria-hidden="true" tabindex="-1"></a>yy <span class="ot">&lt;-</span> y[x[, <span class="dv">2</span>] <span class="sc">!=</span> <span class="dv">3</span>, ]</span>
<span id="cb65-3"><a href="#cb65-3" aria-hidden="true" tabindex="-1"></a>xx <span class="ot">&lt;-</span> x[x[, <span class="dv">2</span>] <span class="sc">!=</span> <span class="dv">3</span>, ]</span>
<span id="cb65-4"><a href="#cb65-4" aria-hidden="true" tabindex="-1"></a><span class="co"># specify the number of resamples k</span></span>
<span id="cb65-5"><a href="#cb65-5" aria-hidden="true" tabindex="-1"></a>k <span class="ot">&lt;-</span> <span class="dv">10</span></span>
<span id="cb65-6"><a href="#cb65-6" aria-hidden="true" tabindex="-1"></a>beta_all <span class="ot">&lt;-</span> <span class="fu">matrix</span>(<span class="at">nrow =</span> <span class="fu">ncol</span>(xx), <span class="at">ncol =</span> k)</span>
<span id="cb65-7"><a href="#cb65-7" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">123</span>)</span>
<span id="cb65-8"><a href="#cb65-8" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (j <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span>k) {</span>
<span id="cb65-9"><a href="#cb65-9" aria-hidden="true" tabindex="-1"></a>  resample_id <span class="ot">&lt;-</span> <span class="fu">sample</span>(<span class="dv">1</span><span class="sc">:</span><span class="fu">nrow</span>(yy), <span class="fu">nrow</span>(yy), <span class="at">replace =</span> <span class="cn">TRUE</span>)</span>
<span id="cb65-10"><a href="#cb65-10" aria-hidden="true" tabindex="-1"></a>  resample_x <span class="ot">&lt;-</span> xx[resample_id, ]</span>
<span id="cb65-11"><a href="#cb65-11" aria-hidden="true" tabindex="-1"></a>  resample_y <span class="ot">&lt;-</span> yy[resample_id, ]</span>
<span id="cb65-12"><a href="#cb65-12" aria-hidden="true" tabindex="-1"></a>  cvfit <span class="ot">&lt;-</span> <span class="fu">cv.glmnet</span>(resample_x, resample_y, <span class="at">family =</span> <span class="st">"cox"</span>, <span class="at">nfolds =</span> <span class="dv">5</span>, <span class="at">penalty.factor =</span> pf)</span>
<span id="cb65-13"><a href="#cb65-13" aria-hidden="true" tabindex="-1"></a>  beta_all[, j] <span class="ot">&lt;-</span> <span class="fu">as.vector</span>(<span class="fu">coef</span>(cvfit, <span class="at">s =</span> cvfit<span class="sc">$</span>lambda.min))</span>
<span id="cb65-14"><a href="#cb65-14" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb65-15"><a href="#cb65-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb65-16"><a href="#cb65-16" aria-hidden="true" tabindex="-1"></a><span class="co"># identify features at least 80% frequently selected</span></span>
<span id="cb65-17"><a href="#cb65-17" aria-hidden="true" tabindex="-1"></a>x_stable <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(xx[, <span class="fu">rowSums</span>(beta_all <span class="sc">!=</span> <span class="dv">0</span>) <span class="sc">&gt;=</span> k <span class="sc">*</span> <span class="fl">0.8</span>])</span>
<span id="cb65-18"><a href="#cb65-18" aria-hidden="true" tabindex="-1"></a>x_stable<span class="sc">$</span>ethnicity <span class="ot">&lt;-</span> <span class="fu">factor</span>(x_stable<span class="sc">$</span>ethnicity)</span>
<span id="cb65-19"><a href="#cb65-19" aria-hidden="true" tabindex="-1"></a><span class="fu">levels</span>(x_stable<span class="sc">$</span>ethnicity) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"Hispanic/latino"</span>, <span class="st">"Not hispanic/latino"</span>)</span>
<span id="cb65-20"><a href="#cb65-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb65-21"><a href="#cb65-21" aria-hidden="true" tabindex="-1"></a>data_tmp <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">times =</span> yy[, <span class="st">"time"</span>], <span class="at">status =</span> yy[, <span class="st">"status"</span>], x_stable)</span>
<span id="cb65-22"><a href="#cb65-22" aria-hidden="true" tabindex="-1"></a>f <span class="ot">&lt;-</span> <span class="fu">cph</span>(</span>
<span id="cb65-23"><a href="#cb65-23" aria-hidden="true" tabindex="-1"></a>  <span class="at">formula =</span> <span class="fu">Surv</span>(times, status) <span class="sc">~</span> age <span class="sc">+</span> ethnicity <span class="sc">+</span> ANLN <span class="sc">+</span> BLVRA <span class="sc">+</span> EGFR,</span>
<span id="cb65-24"><a href="#cb65-24" aria-hidden="true" tabindex="-1"></a>  <span class="at">data =</span> data_tmp, <span class="at">x =</span> <span class="cn">TRUE</span>, <span class="at">y =</span> <span class="cn">TRUE</span>, <span class="at">surv =</span> <span class="cn">TRUE</span></span>
<span id="cb65-25"><a href="#cb65-25" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb65-26"><a href="#cb65-26" aria-hidden="true" tabindex="-1"></a>ddist <span class="ot">&lt;-</span> <span class="fu">datadist</span>(data_tmp)</span>
<span id="cb65-27"><a href="#cb65-27" aria-hidden="true" tabindex="-1"></a>oldoption <span class="ot">&lt;-</span> <span class="fu">options</span>(<span class="at">datadist =</span> <span class="st">"ddist"</span>)</span>
<span id="cb65-28"><a href="#cb65-28" aria-hidden="true" tabindex="-1"></a>surv <span class="ot">&lt;-</span> <span class="fu">Survival</span>(f)</span>
<span id="cb65-29"><a href="#cb65-29" aria-hidden="true" tabindex="-1"></a>nom <span class="ot">&lt;-</span> <span class="fu">nomogram</span>(f,</span>
<span id="cb65-30"><a href="#cb65-30" aria-hidden="true" tabindex="-1"></a>  <span class="at">fun =</span> <span class="fu">list</span>(<span class="cf">function</span>(x) <span class="fu">surv</span>(<span class="dv">1</span>, x), <span class="cf">function</span>(x) <span class="fu">surv</span>(<span class="dv">3</span>, x), <span class="cf">function</span>(x) <span class="fu">surv</span>(<span class="dv">5</span>, x)),</span>
<span id="cb65-31"><a href="#cb65-31" aria-hidden="true" tabindex="-1"></a>  <span class="at">funlabel =</span> <span class="fu">c</span>(</span>
<span id="cb65-32"><a href="#cb65-32" aria-hidden="true" tabindex="-1"></a>    <span class="st">"1-Year Survival Probability"</span>,</span>
<span id="cb65-33"><a href="#cb65-33" aria-hidden="true" tabindex="-1"></a>    <span class="st">"3-Year Survival Probability"</span>,</span>
<span id="cb65-34"><a href="#cb65-34" aria-hidden="true" tabindex="-1"></a>    <span class="st">"5-Year Survival Probability"</span></span>
<span id="cb65-35"><a href="#cb65-35" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb65-36"><a href="#cb65-36" aria-hidden="true" tabindex="-1"></a>  <span class="at">lp =</span> <span class="cn">FALSE</span></span>
<span id="cb65-37"><a href="#cb65-37" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb65-38"><a href="#cb65-38" aria-hidden="true" tabindex="-1"></a>regplot<span class="sc">::</span><span class="fu">regplot</span>(f,</span>
<span id="cb65-39"><a href="#cb65-39" aria-hidden="true" tabindex="-1"></a>  <span class="at">observation =</span> data_tmp[<span class="dv">1</span>, ], <span class="at">failtime =</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">3</span>, <span class="dv">5</span>), <span class="at">title =</span> <span class="st">""</span>,</span>
<span id="cb65-40"><a href="#cb65-40" aria-hidden="true" tabindex="-1"></a>  <span class="at">prfail =</span> <span class="cn">FALSE</span>, <span class="at">points =</span> <span class="cn">TRUE</span>, <span class="at">showP =</span> <span class="cn">FALSE</span>, <span class="at">subticks =</span> <span class="cn">TRUE</span></span>
<span id="cb65-41"><a href="#cb65-41" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><img src="fig/TCGA_surv_nomogram.png" class="img-fluid figure-img" style="width:80.0%"></p>
<figcaption class="figure-caption"><em>Nomogram developed to estimate the overall survival probability for TCGA’s BRAC patients based on demographic and Lasso Cox selected mRNA features. The red coloured symbols represent one patient’s information and predicted probabilities of 1-year, 3-year and 5-year survival.</em></figcaption>
</figure>
</div>
<p><br></p>
<p><font size="4"> <span id="slopeCali"><strong>Calibration plot</strong></span> </font></p>
<p>A calibration plot is a straightforward visualization to show the prediction ability of the nomogram, i.e., the agreement between predicted survival probabilities from the final model and the KM estimated survival probabilities in different percentiles of the predicted values at a time point of interest. We demonstrate below calibration plots based on training and validation data sets, respectively.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb66"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb66-1"><a href="#cb66-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Calibration at 5-year time-point</span></span>
<span id="cb66-2"><a href="#cb66-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb66-3"><a href="#cb66-3" aria-hidden="true" tabindex="-1"></a><span class="co"># prepare suitable data format for calibration plot</span></span>
<span id="cb66-4"><a href="#cb66-4" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">123</span>)</span>
<span id="cb66-5"><a href="#cb66-5" aria-hidden="true" tabindex="-1"></a>train_id <span class="ot">&lt;-</span> <span class="fu">sample</span>(<span class="dv">1</span><span class="sc">:</span><span class="fu">nrow</span>(yy), <span class="fu">nrow</span>(yy) <span class="sc">*</span> <span class="fl">0.8</span>, <span class="at">replace =</span> <span class="cn">FALSE</span>)</span>
<span id="cb66-6"><a href="#cb66-6" aria-hidden="true" tabindex="-1"></a>data_train <span class="ot">&lt;-</span> data_tmp[train_id, ]</span>
<span id="cb66-7"><a href="#cb66-7" aria-hidden="true" tabindex="-1"></a>data_validate <span class="ot">&lt;-</span> data_tmp[<span class="sc">-</span>train_id, ]</span>
<span id="cb66-8"><a href="#cb66-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb66-9"><a href="#cb66-9" aria-hidden="true" tabindex="-1"></a>ddist <span class="ot">&lt;-</span> <span class="fu">datadist</span>(data_train)</span>
<span id="cb66-10"><a href="#cb66-10" aria-hidden="true" tabindex="-1"></a><span class="fu">options</span>(<span class="at">datadist =</span> <span class="st">"ddist"</span>)</span>
<span id="cb66-11"><a href="#cb66-11" aria-hidden="true" tabindex="-1"></a>f_train <span class="ot">&lt;-</span> <span class="fu">cph</span>(</span>
<span id="cb66-12"><a href="#cb66-12" aria-hidden="true" tabindex="-1"></a>  <span class="at">formula =</span> <span class="fu">Surv</span>(times, status) <span class="sc">~</span> age <span class="sc">+</span> ethnicity <span class="sc">+</span> ANLN <span class="sc">+</span> BLVRA <span class="sc">+</span> EGFR,</span>
<span id="cb66-13"><a href="#cb66-13" aria-hidden="true" tabindex="-1"></a>  <span class="at">data =</span> data_train, <span class="at">x =</span> <span class="cn">TRUE</span>, <span class="at">y =</span> <span class="cn">TRUE</span>, <span class="at">surv =</span> <span class="cn">TRUE</span>, <span class="at">time.inc =</span> <span class="dv">5</span></span>
<span id="cb66-14"><a href="#cb66-14" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb66-15"><a href="#cb66-15" aria-hidden="true" tabindex="-1"></a>f_validate <span class="ot">&lt;-</span> <span class="fu">update</span>(f_train, <span class="at">data =</span> data_validate)</span>
<span id="cb66-16"><a href="#cb66-16" aria-hidden="true" tabindex="-1"></a>cal_train <span class="ot">&lt;-</span> <span class="fu">calibrate</span>(f_train, <span class="at">u =</span> <span class="dv">5</span>, <span class="at">cmethod =</span> <span class="st">"KM"</span>, <span class="at">m =</span> <span class="fu">nrow</span>(data_train) <span class="sc">/</span> <span class="dv">4</span>, <span class="at">B =</span> <span class="dv">200</span>)</span>
<span id="cb66-17"><a href="#cb66-17" aria-hidden="true" tabindex="-1"></a>cal_validate <span class="ot">&lt;-</span> <span class="fu">calibrate</span>(f_validate, <span class="at">u =</span> <span class="dv">5</span>, <span class="at">cmethod =</span> <span class="st">"KM"</span>, <span class="at">m =</span> <span class="fu">nrow</span>(data_validate) <span class="sc">/</span> <span class="dv">4</span>, <span class="at">B =</span> <span class="dv">200</span>)</span>
<span id="cb66-18"><a href="#cb66-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb66-19"><a href="#cb66-19" aria-hidden="true" tabindex="-1"></a><span class="fu">layout</span>(<span class="fu">matrix</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">2</span>, <span class="at">nrow =</span> <span class="dv">1</span>))</span>
<span id="cb66-20"><a href="#cb66-20" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(cal_train,</span>
<span id="cb66-21"><a href="#cb66-21" aria-hidden="true" tabindex="-1"></a>  <span class="at">lwd =</span> <span class="dv">2</span>, <span class="at">lty =</span> <span class="dv">1</span>, <span class="at">errbar.col =</span> <span class="st">"seagreen3"</span>,</span>
<span id="cb66-22"><a href="#cb66-22" aria-hidden="true" tabindex="-1"></a>  <span class="at">xlab =</span> <span class="st">"Predicted survival probability"</span>, <span class="at">ylab =</span> <span class="st">"Actual survival probability"</span>,</span>
<span id="cb66-23"><a href="#cb66-23" aria-hidden="true" tabindex="-1"></a>  <span class="at">xlim =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">1</span>), <span class="at">ylim =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">1</span>), <span class="at">col =</span> <span class="st">"seagreen3"</span>, <span class="at">subtitles =</span> <span class="cn">FALSE</span></span>
<span id="cb66-24"><a href="#cb66-24" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb66-25"><a href="#cb66-25" aria-hidden="true" tabindex="-1"></a><span class="fu">title</span>(<span class="at">main =</span> <span class="st">"Calibration on training data"</span>)</span>
<span id="cb66-26"><a href="#cb66-26" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb66-27"><a href="#cb66-27" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(cal_validate,</span>
<span id="cb66-28"><a href="#cb66-28" aria-hidden="true" tabindex="-1"></a>  <span class="at">lwd =</span> <span class="dv">2</span>, <span class="at">lty =</span> <span class="dv">1</span>, <span class="at">errbar.col =</span> <span class="st">"seagreen3"</span>,</span>
<span id="cb66-29"><a href="#cb66-29" aria-hidden="true" tabindex="-1"></a>  <span class="at">xlab =</span> <span class="st">"Predicted survival probability"</span>, <span class="at">ylab =</span> <span class="st">"Actual survival probability"</span>,</span>
<span id="cb66-30"><a href="#cb66-30" aria-hidden="true" tabindex="-1"></a>  <span class="at">xlim =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">1</span>), <span class="at">ylim =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">1</span>), <span class="at">col =</span> <span class="st">"seagreen3"</span>, <span class="at">subtitles =</span> <span class="cn">FALSE</span></span>
<span id="cb66-31"><a href="#cb66-31" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb66-32"><a href="#cb66-32" aria-hidden="true" tabindex="-1"></a><span class="fu">title</span>(<span class="at">main =</span> <span class="st">"Calibration on validation data"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><img src="fig/TCGA_surv_calibration.png" class="img-fluid figure-img" style="width:70.0%"></p>
<figcaption class="figure-caption"><em>Nomogram model calibration curves for TCGA’s BRAC patients at 5-year evaluation time-point.</em></figcaption>
</figure>
</div>
<p><br></p>
</section>
</section>
<section id="mlr3" class="level3 unnumbered">
<h3 class="unnumbered anchored" data-anchor-id="mlr3">Model evaluation (mlr3)</h3>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
We will use the <a href="https://mlr3.mlr-org.com"><strong>mlr3</strong></a> ML framework and the <a href="https://mlr3proba.mlr-org.com"><strong>mlr3proba</strong></a> <code>R</code> library to:
</div>
</div>
<div class="callout-body-container callout-body">
<ul>
<li>Create a survival task from a dataset and split it to training and test (validation) sets</li>
<li>Define a Lasso Cox model that can output both linear predictors and survival distribution predictions and train/tune it on the training set</li>
<li>Make predictions using the trained Lasso Cox model on the separate test set</li>
<li>Measure the performance of our model (discrimination and calibration) using several evaluation metrics</li>
<li>Using resampling techniques, we will assess our model’s capacity for generalization (prediction on unseen data) and the stability of the model’s selected features</li>
</ul>
</div>
</div>
<p>For the rest of the analysis, we will borrow the terminology from the <a href="https://mlr3.mlr-org.com"><strong>mlr3</strong></a> ecosystem of machine learning packages (e.g.&nbsp;<em>task</em> is a dataset, <em>learner</em> is a model, etc.). See <a href="https://mlr3book.mlr-org.com/">mlr3 book</a> for more details.</p>
<p>First, we load the necessary <a href="https://mlr3.mlr-org.com"><strong>mlr3</strong></a> libraries <span class="citation" data-cites="Lang2019 Sonabend2021">(<a href="#ref-Lang2019" role="doc-biblioref">Lang et al. 2019</a>; <a href="#ref-Sonabend2021" role="doc-biblioref">Sonabend et al. 2021</a>)</span> and some other useful ones:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb67"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb67-1"><a href="#cb67-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">"mlr3verse"</span>) <span class="co"># mlr3, mlr3pipeplines, mlr3learners, mlr3tuning, paradox, etc.</span></span>
<span id="cb67-2"><a href="#cb67-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">"mlr3proba"</span>) <span class="co"># probabilistic learning and survival analysis</span></span>
<span id="cb67-3"><a href="#cb67-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">"mlr3extralearners"</span>) <span class="co"># for lrn('surv.glmnet')</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><br></p>
<section id="workflow" class="level4 unnumbered">
<h4 class="unnumbered anchored" data-anchor-id="workflow">Workflow</h4>
<p>We construct an <a href="https://mlr3.mlr-org.com"><strong>mlr3</strong></a> <em>survival task</em> (TCGA BRCA dataset essentially, with normalized PAM50 gene expression features and two clinical/demographic variables) and split it into training and test sets (<span class="math inline">\(80\%/20\%\)</span>):</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb68"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb68-1"><a href="#cb68-1" aria-hidden="true" tabindex="-1"></a><span class="co"># From 'Penalized Cox models' section:</span></span>
<span id="cb68-2"><a href="#cb68-2" aria-hidden="true" tabindex="-1"></a><span class="co"># x =&gt; gene expression matrix (50 PAM50 genes) + 2 clinical variables</span></span>
<span id="cb68-3"><a href="#cb68-3" aria-hidden="true" tabindex="-1"></a><span class="co"># y =&gt; (time, status) target matrix</span></span>
<span id="cb68-4"><a href="#cb68-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-5"><a href="#cb68-5" aria-hidden="true" tabindex="-1"></a>data <span class="ot">&lt;-</span> <span class="fu">cbind.data.frame</span>(x, y)</span>
<span id="cb68-6"><a href="#cb68-6" aria-hidden="true" tabindex="-1"></a><span class="co"># data = readRDS(file = 'data.rds')</span></span>
<span id="cb68-7"><a href="#cb68-7" aria-hidden="true" tabindex="-1"></a>task <span class="ot">&lt;-</span> mlr3proba<span class="sc">::</span><span class="fu">as_task_surv</span>(</span>
<span id="cb68-8"><a href="#cb68-8" aria-hidden="true" tabindex="-1"></a>  <span class="at">x =</span> data,</span>
<span id="cb68-9"><a href="#cb68-9" aria-hidden="true" tabindex="-1"></a>  <span class="at">time =</span> <span class="st">"time"</span>, <span class="at">event =</span> <span class="st">"status"</span>, <span class="at">id =</span> <span class="st">"BRCA-TCGA"</span></span>
<span id="cb68-10"><a href="#cb68-10" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb68-11"><a href="#cb68-11" aria-hidden="true" tabindex="-1"></a>task <span class="co"># see useful info about the dataset (#features, #samples, target variables)</span></span>
<span id="cb68-12"><a href="#cb68-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-13"><a href="#cb68-13" aria-hidden="true" tabindex="-1"></a><span class="co"># split to train and test sets</span></span>
<span id="cb68-14"><a href="#cb68-14" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">42</span>)</span>
<span id="cb68-15"><a href="#cb68-15" aria-hidden="true" tabindex="-1"></a>split <span class="ot">&lt;-</span> mlr3<span class="sc">::</span><span class="fu">partition</span>(task, <span class="at">ratio =</span> <span class="fl">0.8</span>)</span>
<span id="cb68-16"><a href="#cb68-16" aria-hidden="true" tabindex="-1"></a><span class="co"># split$train # train indices</span></span>
<span id="cb68-17"><a href="#cb68-17" aria-hidden="true" tabindex="-1"></a><span class="co"># split$test # test indices</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<pre class="border"><code>&lt;TaskSurv:BRCA-TCGA&gt; (1047 x 54)
* Target: time, status
* Properties: -
* Features (52):
  - dbl (52): ACTR3B, ANLN, BAG1, BCL2, BIRC5, BLVRA, CCNB1, CCNE1, CDC20, CDC6, CDH3,
    CENPF, CEP55, CXXC5, EGFR, ERBB2, ESR1, EXO1, FGFR4, FOXA1, FOXC1, GPR160, GRB7,
    KIF2C, KRT14, KRT17, KRT5, MAPT, MDM2, MELK, MIA, MKI67, MLPH, MMP11, MYBL2, MYC,
    NAT1, NDC80, NUF2, ORC6, PGR, PHGDH, PTTG1, RRM2, SFRP1, SLC39A6, TMEM45B, TYMS,
    UBE2C, UBE2T, age, ethnicity</code></pre>
<p>We create a Lasso Cox <a href="https://mlr3.mlr-org.com"><strong>mlr3</strong></a> learner (a wrapper around the <code>glmnet::cv.glmnet()</code> function with the capacity to provide survival predictions), where we specify the two clinical variables to be <em>mandatory</em> (i.e.&nbsp;no penalization) and the <span class="math inline">\(s\)</span> value (<span class="math inline">\(\lambda\)</span> parameter used for prediction) equal to <code>lambda.min</code>:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb70"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb70-1"><a href="#cb70-1" aria-hidden="true" tabindex="-1"></a><span class="co"># tail(task$feature_names) # age, ethnicity are the 2 last features</span></span>
<span id="cb70-2"><a href="#cb70-2" aria-hidden="true" tabindex="-1"></a>pf <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="fu">rep</span>(<span class="dv">1</span>, <span class="fu">length</span>(task<span class="sc">$</span>feature_names) <span class="sc">-</span> <span class="dv">2</span>), <span class="fu">rep</span>(<span class="dv">0</span>, <span class="dv">2</span>))</span>
<span id="cb70-3"><a href="#cb70-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb70-4"><a href="#cb70-4" aria-hidden="true" tabindex="-1"></a><span class="co"># define model</span></span>
<span id="cb70-5"><a href="#cb70-5" aria-hidden="true" tabindex="-1"></a>coxlasso <span class="ot">&lt;-</span> <span class="fu">lrn</span>(<span class="st">"surv.cv_glmnet"</span>,</span>
<span id="cb70-6"><a href="#cb70-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">alpha =</span> <span class="dv">1</span>, <span class="at">nfolds =</span> <span class="dv">5</span>, <span class="at">s =</span> <span class="st">"lambda.min"</span>,</span>
<span id="cb70-7"><a href="#cb70-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">penalty.factor =</span> pf</span>
<span id="cb70-8"><a href="#cb70-8" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb70-9"><a href="#cb70-9" aria-hidden="true" tabindex="-1"></a><span class="co"># coxlasso # see learner details</span></span>
<span id="cb70-10"><a href="#cb70-10" aria-hidden="true" tabindex="-1"></a><span class="co"># coxlasso$help() # see learner documentation</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p>A Cox proportional hazards model (and Lasso Cox as a consequence) is a semi-parametric model, which means that it does not produce survival distribution predictions by default. The mlr3 <code>glmnet</code> survival learner uses internally the function <code>glmnet::survfit.coxnet()</code> to transform the <code>cv.glmnet</code>’s linear predictors (<code>lp</code>) to survival distribution predictions. This transformation uses the <strong>Breslow estimator</strong> for the cumulative baseline hazard (see <code>stype</code> parameter in <code>survival::survfit.coxph()</code>).</p>
</div>
</div>
<p><br></p>
<p>Train the Lasso Cox model:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb71"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb71-1"><a href="#cb71-1" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">3</span>)</span>
<span id="cb71-2"><a href="#cb71-2" aria-hidden="true" tabindex="-1"></a>coxlasso<span class="sc">$</span><span class="fu">train</span>(task, <span class="at">row_ids =</span> split<span class="sc">$</span>train)</span>
<span id="cb71-3"><a href="#cb71-3" aria-hidden="true" tabindex="-1"></a><span class="co"># view `cv.glmnet` fit</span></span>
<span id="cb71-4"><a href="#cb71-4" aria-hidden="true" tabindex="-1"></a>coxlasso<span class="sc">$</span>model<span class="sc">$</span>model</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<pre class="border"><code>Call:  (if (cv) glmnet::cv.glmnet else glmnet::glmnet)(x = data, y = target,      nfolds = 5L, alpha = 1, penalty.factor = c(1, 1,..., 1, 0, 0), family = "cox")

Measure: Partial Likelihood Deviance 

     Lambda Index Measure     SE Nonzero
min 0.01082    14   12.31 0.2743      15
1se 0.03626     1   12.35 0.2564       2</code></pre>
<p>Get the survival distribution predictions (<span class="math inline">\(distr\)</span>) along with the linear predictors (<span class="math inline">\(lp\)</span>):</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb73"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb73-1"><a href="#cb73-1" aria-hidden="true" tabindex="-1"></a>pred <span class="ot">&lt;-</span> coxlasso<span class="sc">$</span><span class="fu">predict</span>(task, <span class="at">row_ids =</span> split<span class="sc">$</span>test)</span>
<span id="cb73-2"><a href="#cb73-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(<span class="fu">as.data.table</span>(pred))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<pre class="border"><code>   row_ids      time status     crank        lp     distr
1:       5 0.9527721  FALSE -2.346574 -2.346574 &lt;list[1]&gt;
2:       6 4.0438056  FALSE -2.806708 -2.806708 &lt;list[1]&gt;
3:      15 1.7385352  FALSE -1.845042 -1.845042 &lt;list[1]&gt;
4:      45 4.5804244  FALSE -1.715041 -1.715041 &lt;list[1]&gt;
5:      50 5.1279945  FALSE -2.790122 -2.790122 &lt;list[1]&gt;
6:      54 6.6858316  FALSE -2.466360 -2.466360 &lt;list[1]&gt;</code></pre>
<p>So for every patient in the test set, the Lasso Cox model prediction is a linear predictor of the form <span class="math inline">\(lp = \hat{\beta} X_{new}\)</span>. <span class="math inline">\(crank\)</span> stands for continuous ranking score and it’s the same as <span class="math inline">\(lp\)</span> for the Lasso Cox model. The <span class="math inline">\(distr\)</span> predictions are the per-patient survival distribution predictions, implemented by the <code>R</code> package <a href="https://github.com/alan-turing-institute/distr6">distr6</a> which the <a href="https://mlr3proba.mlr-org.com"><strong>mlr3proba</strong></a> imports. See respective <a href="https://mlr3proba.mlr-org.com/reference/PredictionSurv.html">documentation</a> on the different prediction types supported.</p>
<p>An example of using the <code>distr</code> predictions would be to request for the survival probability at e.g.&nbsp;<span class="math inline">\(1,5,10,20\)</span> years for the first two patients in the test set:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb75"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb75-1"><a href="#cb75-1" aria-hidden="true" tabindex="-1"></a>times <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">5</span>, <span class="dv">10</span>, <span class="dv">20</span>)</span>
<span id="cb75-2"><a href="#cb75-2" aria-hidden="true" tabindex="-1"></a>ids <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">2</span>)</span>
<span id="cb75-3"><a href="#cb75-3" aria-hidden="true" tabindex="-1"></a>pred<span class="sc">$</span>distr[ids]<span class="sc">$</span><span class="fu">survival</span>(times)</span>
<span id="cb75-4"><a href="#cb75-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb75-5"><a href="#cb75-5" aria-hidden="true" tabindex="-1"></a><span class="co"># same logic for the cumulative hazard</span></span>
<span id="cb75-6"><a href="#cb75-6" aria-hidden="true" tabindex="-1"></a><span class="co"># pred$distr[ids]$cumHazard(times)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<pre class="border"><code>        [,1]      [,2]
1  0.9889822 0.9930314
5  0.8711262 0.9165987
10 0.6708848 0.7772850
20 0.4540815 0.6075516</code></pre>
<p><br></p>
</section>
<section id="discrimination-metrics-1" class="level4 unnumbered">
<h4 class="unnumbered anchored" data-anchor-id="discrimination-metrics-1">Discrimination metrics</h4>
<p>We want to test our Lasso Cox model and see how well it was able to <strong>discriminate the patients in the test set</strong>. For this we can use the <span class="math inline">\(lp\)</span> predictions of Lasso Cox model and metrics such as the (time-dependent) C-index and (time-dependent) AUC. <br></p>
<p><font size="4"> <strong>Harrell’s C-index</strong> <span class="citation" data-cites="Harrell1982">(<a href="#ref-Harrell1982" role="doc-biblioref">Harrell et al. 1982</a>)</span>: </font></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb77"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb77-1"><a href="#cb77-1" aria-hidden="true" tabindex="-1"></a>harrell_c <span class="ot">&lt;-</span> <span class="fu">msr</span>(<span class="st">"surv.cindex"</span>)</span>
<span id="cb77-2"><a href="#cb77-2" aria-hidden="true" tabindex="-1"></a>harrell_c<span class="sc">$</span>id <span class="ot">&lt;-</span> <span class="st">"surv.cindex.harrell"</span></span>
<span id="cb77-3"><a href="#cb77-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb77-4"><a href="#cb77-4" aria-hidden="true" tabindex="-1"></a><span class="co"># harrell_c # get some details about the measure</span></span>
<span id="cb77-5"><a href="#cb77-5" aria-hidden="true" tabindex="-1"></a><span class="co"># harrell_c$minimize # FALSE =&gt; higher C-index is better</span></span>
<span id="cb77-6"><a href="#cb77-6" aria-hidden="true" tabindex="-1"></a><span class="co"># harrell_c$range # [0, 1] =&gt; [min, max]</span></span>
<span id="cb77-7"><a href="#cb77-7" aria-hidden="true" tabindex="-1"></a><span class="co"># harrell_c$predict_type # uses the $crank$ predictions (equal to $lp$ for Lasso Cox</span></span>
<span id="cb77-8"><a href="#cb77-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb77-9"><a href="#cb77-9" aria-hidden="true" tabindex="-1"></a>pred<span class="sc">$</span><span class="fu">score</span>(harrell_c)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<pre class="border"><code>surv.cindex.harrell 
          0.6224306</code></pre>
<p><br></p>
<p><font size="4"> <strong>Uno’s C-index</strong> <span class="citation" data-cites="Uno2011">(<a href="#ref-Uno2011" role="doc-biblioref">Uno et al. 2011</a>)</span>: (across all time points of the test set): </font></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb79"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb79-1"><a href="#cb79-1" aria-hidden="true" tabindex="-1"></a>uno_c <span class="ot">&lt;-</span> <span class="fu">msr</span>(<span class="st">"surv.cindex"</span>, <span class="at">weight_meth =</span> <span class="st">"G2"</span>)</span>
<span id="cb79-2"><a href="#cb79-2" aria-hidden="true" tabindex="-1"></a>uno_c<span class="sc">$</span>id <span class="ot">&lt;-</span> <span class="st">"surv.cindex.uno"</span></span>
<span id="cb79-3"><a href="#cb79-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb79-4"><a href="#cb79-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Uno's C needs the train data</span></span>
<span id="cb79-5"><a href="#cb79-5" aria-hidden="true" tabindex="-1"></a>pred<span class="sc">$</span><span class="fu">score</span>(uno_c, <span class="at">task =</span> task, <span class="at">train_set =</span> split<span class="sc">$</span>train)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<pre class="border"><code>surv.cindex.uno 
      0.5932426</code></pre>
<p><br></p>
<p><font size="4"> <strong>Uno’s Integrated AUC</strong> <span class="citation" data-cites="Uno2007">(<a href="#ref-Uno2007" role="doc-biblioref">Uno et al. 2007</a>)</span> (across all time points of the test set): </font></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb81"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb81-1"><a href="#cb81-1" aria-hidden="true" tabindex="-1"></a>uno_iauc <span class="ot">&lt;-</span> <span class="fu">msr</span>(<span class="st">"surv.uno_auc"</span>)</span>
<span id="cb81-2"><a href="#cb81-2" aria-hidden="true" tabindex="-1"></a>uno_iauc<span class="sc">$</span>id <span class="ot">&lt;-</span> <span class="st">"surv.uno_iauc"</span></span>
<span id="cb81-3"><a href="#cb81-3" aria-hidden="true" tabindex="-1"></a><span class="co"># uno_iauc$param_set$values$integrated # integrated = TRUE by default</span></span>
<span id="cb81-4"><a href="#cb81-4" aria-hidden="true" tabindex="-1"></a><span class="co"># sort(unique(pred$truth[,1])) # time points used</span></span>
<span id="cb81-5"><a href="#cb81-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb81-6"><a href="#cb81-6" aria-hidden="true" tabindex="-1"></a><span class="co"># uno_iauc$properties # needs the train data</span></span>
<span id="cb81-7"><a href="#cb81-7" aria-hidden="true" tabindex="-1"></a>pred<span class="sc">$</span><span class="fu">score</span>(uno_iauc, <span class="at">task =</span> task, <span class="at">train_set =</span> split<span class="sc">$</span>train)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<pre class="border"><code>surv.uno_iauc 
    0.6585791</code></pre>
<p><br></p>
<p><font size="4"> <strong>Uno’s AUC at a specific time point</strong>, e.g.&nbsp;<span class="math inline">\(10\)</span> years: </font></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb83"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb83-1"><a href="#cb83-1" aria-hidden="true" tabindex="-1"></a>uno_auc <span class="ot">&lt;-</span> <span class="fu">msr</span>(<span class="st">"surv.uno_auc"</span>, <span class="at">integrated =</span> <span class="cn">FALSE</span>, <span class="at">times =</span> <span class="dv">10</span>)</span>
<span id="cb83-2"><a href="#cb83-2" aria-hidden="true" tabindex="-1"></a>uno_auc<span class="sc">$</span>id <span class="ot">&lt;-</span> <span class="st">"surv.uno_auc.10"</span></span>
<span id="cb83-3"><a href="#cb83-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb83-4"><a href="#cb83-4" aria-hidden="true" tabindex="-1"></a><span class="co"># needs the train data</span></span>
<span id="cb83-5"><a href="#cb83-5" aria-hidden="true" tabindex="-1"></a>pred<span class="sc">$</span><span class="fu">score</span>(uno_auc, <span class="at">task =</span> task, <span class="at">train_set =</span> split<span class="sc">$</span>train)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<pre class="border"><code>surv.uno_auc.10 
      0.667014</code></pre>
<p><br></p>
</section>
<section id="calibration-metrics-1" class="level4 unnumbered">
<h4 class="unnumbered anchored" data-anchor-id="calibration-metrics-1">Calibration metrics</h4>
<p>We want to test how well our Lasso Cox model was <strong>calibrated</strong>. <span class="citation" data-cites="Andres2018">Andres et al. (<a href="#ref-Andres2018" role="doc-biblioref">2018</a>)</span> and <span class="citation" data-cites="Haider2020">Haider et al. (<a href="#ref-Haider2020" role="doc-biblioref">2020</a>)</span> suggested the distributional (D)-calibration accounting survival probabilities across all times. This can be useful when assessing the entire post-treatment survival prognosis, for example, assessing the post liver transplantation survival utility in <span class="citation" data-cites="Andres2018">Andres et al. (<a href="#ref-Andres2018" role="doc-biblioref">2018</a>)</span>.</p>
<p><font size="4"> <strong>D-calibration</strong> </font></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb85"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb85-1"><a href="#cb85-1" aria-hidden="true" tabindex="-1"></a>dcal <span class="ot">&lt;-</span> <span class="fu">msr</span>(<span class="st">"surv.dcalib"</span>)</span>
<span id="cb85-2"><a href="#cb85-2" aria-hidden="true" tabindex="-1"></a>pred<span class="sc">$</span><span class="fu">score</span>(dcal)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<pre class="border"><code>surv.dcalib
   0.8342449</code></pre>
<p><br></p>
</section>
<section id="overall-metrics-1" class="level4 unnumbered">
<h4 class="unnumbered anchored" data-anchor-id="overall-metrics-1">Overall metrics</h4>
<p>Usually we derive an estimation of the error between the survival distributions (<span class="math inline">\(distr\)</span> predictions) of the patients in the test set and their actual survival outcomes (corresponding to the survival task’s <code>time</code> and <code>status</code> variables). The most frequently used metric is the Brier Score <span class="citation" data-cites="Graf1999">(<a href="#ref-Graf1999" role="doc-biblioref">Graf et al. 1999</a>)</span>: <br></p>
<p><font size="4"> <strong>Integrated Brier Score (IBS)</strong> (across all time points of the test set): </font></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb87"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb87-1"><a href="#cb87-1" aria-hidden="true" tabindex="-1"></a>ibrier <span class="ot">&lt;-</span> <span class="fu">msr</span>(<span class="st">"surv.brier"</span>, <span class="at">proper =</span> <span class="cn">TRUE</span>)</span>
<span id="cb87-2"><a href="#cb87-2" aria-hidden="true" tabindex="-1"></a><span class="co"># ibrier$help() # see documentation</span></span>
<span id="cb87-3"><a href="#cb87-3" aria-hidden="true" tabindex="-1"></a><span class="co"># ibrier$predict_type # uses the `distr` predictions</span></span>
<span id="cb87-4"><a href="#cb87-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb87-5"><a href="#cb87-5" aria-hidden="true" tabindex="-1"></a><span class="co"># better to use the train data for the Kaplan-Meier estimation of the censoring distribution, but can use the test set as well</span></span>
<span id="cb87-6"><a href="#cb87-6" aria-hidden="true" tabindex="-1"></a>pred<span class="sc">$</span><span class="fu">score</span>(ibrier, <span class="at">task =</span> task, <span class="at">train_set =</span> split<span class="sc">$</span>train)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<pre class="border"><code>surv.graf 
0.1157316</code></pre>
<p>We can also get the <em>standard error</em> of IBS (the above result is the mean across all the test set’s patients) as follows:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb89"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb89-1"><a href="#cb89-1" aria-hidden="true" tabindex="-1"></a>ibrier_se <span class="ot">&lt;-</span> <span class="fu">msr</span>(<span class="st">"surv.brier"</span>, <span class="at">proper =</span> <span class="cn">TRUE</span>, <span class="at">se =</span> <span class="cn">TRUE</span>)</span>
<span id="cb89-2"><a href="#cb89-2" aria-hidden="true" tabindex="-1"></a>pred<span class="sc">$</span><span class="fu">score</span>(ibrier_se, <span class="at">task =</span> task, <span class="at">train_set =</span> split<span class="sc">$</span>train)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<pre class="border"><code> surv.graf 
0.01795861</code></pre>
<p><br></p>
<p><font size="4"> <strong>Brier Score at a specific time point</strong>, e.g.&nbsp;<span class="math inline">\(10\)</span> years: </font></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb91"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb91-1"><a href="#cb91-1" aria-hidden="true" tabindex="-1"></a>brier10 <span class="ot">&lt;-</span> <span class="fu">msr</span>(<span class="st">"surv.brier"</span>, <span class="at">proper =</span> <span class="cn">TRUE</span>, <span class="at">integrated =</span> <span class="cn">FALSE</span>, <span class="at">times =</span> <span class="dv">10</span>)</span>
<span id="cb91-2"><a href="#cb91-2" aria-hidden="true" tabindex="-1"></a>brier10<span class="sc">$</span>id <span class="ot">&lt;-</span> <span class="st">"surv.graf.10"</span></span>
<span id="cb91-3"><a href="#cb91-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb91-4"><a href="#cb91-4" aria-hidden="true" tabindex="-1"></a><span class="co"># better to use the train data for the Kaplan-Meier estimation of the censoring distribution, but can use the test set as well</span></span>
<span id="cb91-5"><a href="#cb91-5" aria-hidden="true" tabindex="-1"></a>pred<span class="sc">$</span><span class="fu">score</span>(brier10, <span class="at">task =</span> task, <span class="at">train_set =</span> split<span class="sc">$</span>train)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<pre class="border"><code>surv.graf.10 
   0.1755639</code></pre>
<p><br></p>
<p><font size="4"> <strong>Right-censored Logarithmic Loss score</strong> (RCLL) <span class="citation" data-cites="Avati2020 Sonabend2022">(<a href="#ref-Avati2020" role="doc-biblioref">Avati et al. 2020</a>; <a href="#ref-Sonabend2022" role="doc-biblioref">Sonabend 2022</a>)</span>: </font></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb93"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb93-1"><a href="#cb93-1" aria-hidden="true" tabindex="-1"></a>rcll <span class="ot">&lt;-</span> <span class="fu">msr</span>(<span class="st">"surv.rcll"</span>)</span>
<span id="cb93-2"><a href="#cb93-2" aria-hidden="true" tabindex="-1"></a>pred<span class="sc">$</span><span class="fu">score</span>(rcll)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<pre class="border"><code>surv.rcll 
 4.781513</code></pre>
<p><br></p>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p>View all evaluation metrics for survival data implemented in <a href="https://mlr3proba.mlr-org.com"><strong>mlr3proba</strong></a> <a href="https://mlr3proba.mlr-org.com/reference/#survival-measures">here</a></p>
</div>
</div>
<p><br></p>
</section>
<section id="uncertainty-quantification" class="level4 unnumbered">
<h4 class="unnumbered anchored" data-anchor-id="uncertainty-quantification">Uncertainty Quantification</h4>
<p>Similar procedure as followed in a <a href="#uq1">previous section</a>.</p>
<p>We will perform a <strong>stratified split</strong> of the BRCA-TCGA survival task to training and test sets (with a <span class="math inline">\(80\%/20\%\)</span> ratio as before). Stratification on the censoring indicator <code>status</code> is important because we want our training and test sets to have the same censoring distribution as the initial dataset. Thus we can avoid measuring performance on test sets with severely different censoring distributions that might influence the performance scores.</p>
<p>Stratify survival task by <code>status</code>:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb95"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb95-1"><a href="#cb95-1" aria-hidden="true" tabindex="-1"></a>coxlasso<span class="sc">$</span><span class="fu">reset</span>() <span class="co"># un-train model</span></span>
<span id="cb95-2"><a href="#cb95-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb95-3"><a href="#cb95-3" aria-hidden="true" tabindex="-1"></a>task<span class="sc">$</span>col_roles<span class="sc">$</span>stratum <span class="ot">&lt;-</span> <span class="st">"status"</span></span>
<span id="cb95-4"><a href="#cb95-4" aria-hidden="true" tabindex="-1"></a><span class="co"># task</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Next, we define the type of resampling (<code>?mlr_resamplings_subsampling</code>), train the Lasso Cox model on all training sets (<span class="math inline">\(100\)</span>) and store the fitted models for feature selection and evaluation:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb96"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb96-1"><a href="#cb96-1" aria-hidden="true" tabindex="-1"></a><span class="co"># 100 times train/test split (80% for training, 20% for validation)</span></span>
<span id="cb96-2"><a href="#cb96-2" aria-hidden="true" tabindex="-1"></a>subsampling <span class="ot">&lt;-</span> <span class="fu">rsmp</span>(<span class="st">"subsampling"</span>, <span class="at">repeats =</span> <span class="dv">100</span>, <span class="at">ratio =</span> <span class="fl">0.8</span>)</span>
<span id="cb96-3"><a href="#cb96-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb96-4"><a href="#cb96-4" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">42</span>)</span>
<span id="cb96-5"><a href="#cb96-5" aria-hidden="true" tabindex="-1"></a>rr <span class="ot">&lt;-</span> mlr3<span class="sc">::</span><span class="fu">resample</span>(</span>
<span id="cb96-6"><a href="#cb96-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">task =</span> task, <span class="at">learner =</span> coxlasso,</span>
<span id="cb96-7"><a href="#cb96-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">resampling =</span> subsampling, <span class="at">store_models =</span> <span class="cn">TRUE</span>, <span class="at">store_backends =</span> <span class="cn">TRUE</span></span>
<span id="cb96-8"><a href="#cb96-8" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We can use all the aforementioned evaluation metrics to measure the performance of the Lasso Cox models on the <span class="math inline">\(100\)</span> different test sets. Note that if a metric needs the training dataset it is automatically provided by the <code>ResampleResult</code> object (<code>rr</code>):</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb97"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb97-1"><a href="#cb97-1" aria-hidden="true" tabindex="-1"></a>measures <span class="ot">&lt;-</span> <span class="fu">list</span>(harrell_c, uno_c, uno_iauc, uno_auc, ibrier, brier10, rcll, dcal)</span>
<span id="cb97-2"><a href="#cb97-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb97-3"><a href="#cb97-3" aria-hidden="true" tabindex="-1"></a>res <span class="ot">&lt;-</span> rr<span class="sc">$</span><span class="fu">score</span>(<span class="at">measures =</span> measures)</span>
<span id="cb97-4"><a href="#cb97-4" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(res)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<pre class="border"><code>    task_id learner_id resampling_id iteration surv.cindex.harrell surv.cindex.uno 
1: BRCA-TCGA  Lasso Cox   subsampling         1           0.5679167       0.6090304
2: BRCA-TCGA  Lasso Cox   subsampling         2           0.5422131       0.4884603
3: BRCA-TCGA  Lasso Cox   subsampling         3           0.7604049       0.5740556
4: BRCA-TCGA  Lasso Cox   subsampling         4           0.6610169       0.5277736
5: BRCA-TCGA  Lasso Cox   subsampling         5           0.5800073       0.5655076
6: BRCA-TCGA  Lasso Cox   subsampling         6           0.5427837       0.6975740
  surv.uno_iauc surv.uno_auc.10 surv.graf surv.graf.10 surv.rcll surv.dcalib
1:     0.6628350       0.4719335 0.1736450    0.2582078  4.914347   0.5890987
2:     0.4023684       0.5652588 0.1855191    0.2185453  4.914264   0.6552037
3:     0.5941948       0.5235439 0.1014616    0.1329289  5.041827   0.8480314
4:     0.5360690       0.5110032 0.1122070    0.1157937  5.045144   0.4981035
5:     0.6160743       0.5388393 0.1099091    0.1940644  5.060352   0.2036418
6:     0.6494779       0.6400328 0.1377626    0.2445218  5.059521   0.5175357
Hidden columns: task, learner, resampling, prediction</code></pre>
<p>We extract and visualize the discrimination and calibration (resampled) performance of our Lasso Cox model using several evaluation metrics:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb99"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb99-1"><a href="#cb99-1" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">42</span>)</span>
<span id="cb99-2"><a href="#cb99-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb99-3"><a href="#cb99-3" aria-hidden="true" tabindex="-1"></a><span class="co"># C-indexes, AUCs (integrated and at t = 10 years)</span></span>
<span id="cb99-4"><a href="#cb99-4" aria-hidden="true" tabindex="-1"></a>res[, .(surv.cindex.harrell, surv.cindex.uno, surv.uno_iauc, surv.uno_auc<span class="fl">.10</span>)] <span class="sc">%&gt;%</span></span>
<span id="cb99-5"><a href="#cb99-5" aria-hidden="true" tabindex="-1"></a>  tidyr<span class="sc">::</span><span class="fu">pivot_longer</span>(</span>
<span id="cb99-6"><a href="#cb99-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">cols =</span> tidyselect<span class="sc">::</span><span class="fu">everything</span>(),</span>
<span id="cb99-7"><a href="#cb99-7" aria-hidden="true" tabindex="-1"></a>    <span class="at">names_to =</span> <span class="st">"Measure"</span>, <span class="at">values_to =</span> <span class="st">"Value"</span></span>
<span id="cb99-8"><a href="#cb99-8" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">%&gt;%</span></span>
<span id="cb99-9"><a href="#cb99-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">Measure =</span> <span class="fu">case_when</span>(</span>
<span id="cb99-10"><a href="#cb99-10" aria-hidden="true" tabindex="-1"></a>    Measure <span class="sc">==</span> <span class="st">"surv.cindex.harrell"</span> <span class="sc">~</span> <span class="st">"Harrell's C-index"</span>,</span>
<span id="cb99-11"><a href="#cb99-11" aria-hidden="true" tabindex="-1"></a>    Measure <span class="sc">==</span> <span class="st">"surv.cindex.uno"</span> <span class="sc">~</span> <span class="st">"Uno's C-index"</span>,</span>
<span id="cb99-12"><a href="#cb99-12" aria-hidden="true" tabindex="-1"></a>    Measure <span class="sc">==</span> <span class="st">"surv.uno_iauc"</span> <span class="sc">~</span> <span class="st">"Uno's Integrated AUC"</span>,</span>
<span id="cb99-13"><a href="#cb99-13" aria-hidden="true" tabindex="-1"></a>    Measure <span class="sc">==</span> <span class="st">"surv.uno_auc.10"</span> <span class="sc">~</span> <span class="st">"Uno's AUC (t = 10 years)"</span>,</span>
<span id="cb99-14"><a href="#cb99-14" aria-hidden="true" tabindex="-1"></a>  )) <span class="sc">%&gt;%</span></span>
<span id="cb99-15"><a href="#cb99-15" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">Measure =</span> <span class="fu">factor</span>(Measure, <span class="at">levels =</span> <span class="fu">c</span>(</span>
<span id="cb99-16"><a href="#cb99-16" aria-hidden="true" tabindex="-1"></a>    <span class="st">"Harrell's C-index"</span>,</span>
<span id="cb99-17"><a href="#cb99-17" aria-hidden="true" tabindex="-1"></a>    <span class="st">"Uno's C-index"</span>,</span>
<span id="cb99-18"><a href="#cb99-18" aria-hidden="true" tabindex="-1"></a>    <span class="st">"Uno's Integrated AUC"</span>,</span>
<span id="cb99-19"><a href="#cb99-19" aria-hidden="true" tabindex="-1"></a>    <span class="st">"Uno's AUC (t = 10 years)"</span></span>
<span id="cb99-20"><a href="#cb99-20" aria-hidden="true" tabindex="-1"></a>  ))) <span class="sc">%&gt;%</span></span>
<span id="cb99-21"><a href="#cb99-21" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> Measure, <span class="at">y =</span> Value, <span class="at">fill =</span> Measure)) <span class="sc">+</span></span>
<span id="cb99-22"><a href="#cb99-22" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_boxplot</span>() <span class="sc">+</span></span>
<span id="cb99-23"><a href="#cb99-23" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ylim</span>(<span class="fu">c</span>(<span class="fl">0.2</span>, <span class="fl">0.8</span>)) <span class="sc">+</span></span>
<span id="cb99-24"><a href="#cb99-24" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_hline</span>(<span class="at">yintercept =</span> <span class="fl">0.5</span>, <span class="at">color =</span> <span class="st">"red"</span>, <span class="at">linetype =</span> <span class="st">"dashed"</span>) <span class="sc">+</span></span>
<span id="cb99-25"><a href="#cb99-25" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>(<span class="at">base_size =</span> <span class="dv">14</span>) <span class="sc">+</span></span>
<span id="cb99-26"><a href="#cb99-26" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">"Discrimination Measures"</span>) <span class="sc">+</span></span>
<span id="cb99-27"><a href="#cb99-27" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">axis.text.x =</span> <span class="fu">element_blank</span>())</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="quarto-figure quarto-figure-left">
<figure class="figure">
<p><img src="fig/mlr3_discrimination_msrs.png" class="img-fluid figure-img" style="width:70.0%"></p>
<figcaption class="figure-caption"><em>Discrimination performance of Lasso Cox on the TCGA-BRCA dataset (expression data of the PAM50 genes and the variables age and ethnicity). Performance metrics used are Harrell’s C-index, Uno’s C-index, Uno’s Integrated AUC and Uno’s AUC at 10 years. The dataset was split to training/validation sets 100 times to allow for the quantification of uncertainty in the different performance estimates.</em></figcaption>
</figure>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb100"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb100-1"><a href="#cb100-1" aria-hidden="true" tabindex="-1"></a><span class="co"># different scales for each measure, so we separate the plots</span></span>
<span id="cb100-2"><a href="#cb100-2" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">42</span>)</span>
<span id="cb100-3"><a href="#cb100-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb100-4"><a href="#cb100-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Integrated Brier Score and Brier Score at t = 10 years</span></span>
<span id="cb100-5"><a href="#cb100-5" aria-hidden="true" tabindex="-1"></a>res[, .(surv.graf, surv.graf<span class="fl">.10</span>)] <span class="sc">%&gt;%</span></span>
<span id="cb100-6"><a href="#cb100-6" aria-hidden="true" tabindex="-1"></a>  tidyr<span class="sc">::</span><span class="fu">pivot_longer</span>(</span>
<span id="cb100-7"><a href="#cb100-7" aria-hidden="true" tabindex="-1"></a>    <span class="at">cols =</span> tidyselect<span class="sc">::</span><span class="fu">everything</span>(),</span>
<span id="cb100-8"><a href="#cb100-8" aria-hidden="true" tabindex="-1"></a>    <span class="at">names_to =</span> <span class="st">"Measure"</span>, <span class="at">values_to =</span> <span class="st">"Value"</span></span>
<span id="cb100-9"><a href="#cb100-9" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">%&gt;%</span></span>
<span id="cb100-10"><a href="#cb100-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">Measure =</span> <span class="fu">case_when</span>(</span>
<span id="cb100-11"><a href="#cb100-11" aria-hidden="true" tabindex="-1"></a>    Measure <span class="sc">==</span> <span class="st">"surv.graf"</span> <span class="sc">~</span> <span class="st">"IBS"</span>,</span>
<span id="cb100-12"><a href="#cb100-12" aria-hidden="true" tabindex="-1"></a>    Measure <span class="sc">==</span> <span class="st">"surv.graf.10"</span> <span class="sc">~</span> <span class="st">"BS(t=10)"</span></span>
<span id="cb100-13"><a href="#cb100-13" aria-hidden="true" tabindex="-1"></a>  )) <span class="sc">%&gt;%</span></span>
<span id="cb100-14"><a href="#cb100-14" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> Measure, <span class="at">y =</span> Value, <span class="at">fill =</span> Measure)) <span class="sc">+</span></span>
<span id="cb100-15"><a href="#cb100-15" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_boxplot</span>(<span class="at">show.legend =</span> <span class="cn">FALSE</span>) <span class="sc">+</span></span>
<span id="cb100-16"><a href="#cb100-16" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_jitter</span>(<span class="at">color =</span> <span class="st">"blue"</span>, <span class="at">size =</span> <span class="fl">0.5</span>, <span class="at">alpha =</span> <span class="fl">0.5</span>, <span class="at">show.legend =</span> <span class="cn">FALSE</span>) <span class="sc">+</span></span>
<span id="cb100-17"><a href="#cb100-17" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">"Integrated Brier Score vs Brier Score (t = 10 years)"</span>) <span class="sc">+</span></span>
<span id="cb100-18"><a href="#cb100-18" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>(<span class="at">base_size =</span> <span class="dv">14</span>) <span class="sc">+</span></span>
<span id="cb100-19"><a href="#cb100-19" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">axis.title.x =</span> <span class="fu">element_blank</span>())</span>
<span id="cb100-20"><a href="#cb100-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb100-21"><a href="#cb100-21" aria-hidden="true" tabindex="-1"></a><span class="co"># RCLL</span></span>
<span id="cb100-22"><a href="#cb100-22" aria-hidden="true" tabindex="-1"></a>res[, .(surv.rcll)] <span class="sc">%&gt;%</span></span>
<span id="cb100-23"><a href="#cb100-23" aria-hidden="true" tabindex="-1"></a>  tidyr<span class="sc">::</span><span class="fu">pivot_longer</span>(</span>
<span id="cb100-24"><a href="#cb100-24" aria-hidden="true" tabindex="-1"></a>    <span class="at">cols =</span> tidyselect<span class="sc">::</span><span class="fu">everything</span>(),</span>
<span id="cb100-25"><a href="#cb100-25" aria-hidden="true" tabindex="-1"></a>    <span class="at">names_to =</span> <span class="st">"Measure"</span>, <span class="at">values_to =</span> <span class="st">"Value"</span></span>
<span id="cb100-26"><a href="#cb100-26" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">%&gt;%</span></span>
<span id="cb100-27"><a href="#cb100-27" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">Measure =</span> <span class="fu">case_when</span>(</span>
<span id="cb100-28"><a href="#cb100-28" aria-hidden="true" tabindex="-1"></a>    Measure <span class="sc">==</span> <span class="st">"surv.rcll"</span> <span class="sc">~</span> <span class="st">"RCLL"</span></span>
<span id="cb100-29"><a href="#cb100-29" aria-hidden="true" tabindex="-1"></a>  )) <span class="sc">%&gt;%</span></span>
<span id="cb100-30"><a href="#cb100-30" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> Measure, <span class="at">y =</span> Value)) <span class="sc">+</span></span>
<span id="cb100-31"><a href="#cb100-31" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_boxplot</span>(<span class="at">show.legend =</span> <span class="cn">FALSE</span>) <span class="sc">+</span></span>
<span id="cb100-32"><a href="#cb100-32" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_jitter</span>(<span class="at">color =</span> <span class="st">"blue"</span>, <span class="at">size =</span> <span class="fl">0.5</span>, <span class="at">alpha =</span> <span class="fl">0.5</span>, <span class="at">show.legend =</span> <span class="cn">FALSE</span>) <span class="sc">+</span></span>
<span id="cb100-33"><a href="#cb100-33" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">"Right-censored Log Loss"</span>) <span class="sc">+</span></span>
<span id="cb100-34"><a href="#cb100-34" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>(<span class="at">base_size =</span> <span class="dv">14</span>) <span class="sc">+</span></span>
<span id="cb100-35"><a href="#cb100-35" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">axis.title.x =</span> <span class="fu">element_blank</span>())</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p align="left">
<img alt="1" src="./fig/mlr3_calibration_BS.png" width="49%"> <img alt="2" src="./fig/mlr3_calibration_RCLL.png" width="49%"> <i>Calibration performance of Lasso Cox on the TCGA-BRCA dataset (expression data of the PAM50 genes and the variables age and ethnicity). Performance metrics used are the Integrated Brier Score (IBS), the Brier Score at 10 years and the Right-Censored Logarithmic Loss (RCLL). The dataset was split to training/validation sets 100 times to allow for the quantification of uncertainty in the different performance estimates.</i>
</p>
</section>
<section id="feature-stability-analysis-1" class="level4 unnumbered">
<h4 class="unnumbered anchored" data-anchor-id="feature-stability-analysis-1">Feature stability analysis</h4>
<p>We can extract the selected features from all <span class="math inline">\(100\)</span> trained Lasso Cox models and create a frequency selection table:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb101"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb101-1"><a href="#cb101-1" aria-hidden="true" tabindex="-1"></a><span class="co"># get selected features from all trained learners in a list</span></span>
<span id="cb101-2"><a href="#cb101-2" aria-hidden="true" tabindex="-1"></a>sf_list <span class="ot">&lt;-</span> <span class="fu">lapply</span>(rr<span class="sc">$</span>learners, <span class="cf">function</span>(learner) {</span>
<span id="cb101-3"><a href="#cb101-3" aria-hidden="true" tabindex="-1"></a>  learner<span class="sc">$</span><span class="fu">selected_features</span>()</span>
<span id="cb101-4"><a href="#cb101-4" aria-hidden="true" tabindex="-1"></a>})</span>
<span id="cb101-5"><a href="#cb101-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb101-6"><a href="#cb101-6" aria-hidden="true" tabindex="-1"></a><span class="co"># make frequency selection table</span></span>
<span id="cb101-7"><a href="#cb101-7" aria-hidden="true" tabindex="-1"></a>n <span class="ot">&lt;-</span> <span class="fu">length</span>(sf_list)</span>
<span id="cb101-8"><a href="#cb101-8" aria-hidden="true" tabindex="-1"></a>fs_res <span class="ot">&lt;-</span> <span class="fu">sort</span>(<span class="fu">table</span>(<span class="fu">unlist</span>(sf_list)), <span class="at">decreasing =</span> <span class="cn">TRUE</span>)</span>
<span id="cb101-9"><a href="#cb101-9" aria-hidden="true" tabindex="-1"></a>times <span class="ot">&lt;-</span> <span class="fu">as.vector</span>(<span class="fu">unname</span>(fs_res))</span>
<span id="cb101-10"><a href="#cb101-10" aria-hidden="true" tabindex="-1"></a>tibble<span class="sc">::</span><span class="fu">tibble</span>(<span class="at">feat_name =</span> <span class="fu">names</span>(fs_res), <span class="at">times =</span> times, <span class="at">freq =</span> times <span class="sc">/</span> n)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<pre class="border"><code># A tibble: 35 × 3
   feat_name times  freq
   &lt;chr&gt;     &lt;int&gt; &lt;dbl&gt;
1 age         100  1
2 ethnicity   100  1
3 ANLN         43  0.43
4 BLVRA        41  0.41
5 BAG1         37  0.37
6 MIA          34  0.34
7 TYMS         30  0.3
8 KRT5         27  0.27
9 MMP11        27  0.27
10 BCL2         26  0.26
# ℹ 25 more rows</code></pre>
<p>As <code>age</code> and <code>ethnicity</code> were not penalized, they have non-zero coefficients in all Lasso Cox models and therefore are included in all selected feature sets.</p>
<p>Lastly, we can use the <code>R</code> package <a href="https://CRAN.R-project.org/package=stabm"><strong>stabm</strong></a> <span class="citation" data-cites="stabm">(<a href="#ref-stabm" role="doc-biblioref">Bommert and Lang 2021</a>)</span> to assess how similar the <span class="math inline">\(100\)</span> selected feature sets were. We will demonstrate the use of three metrics which measure the <em>stability</em> of the Lasso Cox’s feature selection on the TCGA-BRCA dataset:</p>
<ol type="1">
<li>The Jaccard index</li>
<li>Nogueira’s metric (corrected for chance, i.e.&nbsp;independent of the number of features; <span class="citation" data-cites="Nogueira2018">Nogueira, Sechidis, and Brown (<a href="#ref-Nogueira2018" role="doc-biblioref">2018</a>)</span>)</li>
<li>Zucknick’s metric (extension of Jaccard index that considers the correlation between the features; <span class="citation" data-cites="Zucknick2008">Zucknick, Richardson, and Stronach (<a href="#ref-Zucknick2008" role="doc-biblioref">2008</a>)</span>):</li>
</ol>
<div class="cell">
<div class="sourceCode cell-code" id="cb103"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb103-1"><a href="#cb103-1" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">42</span>)</span>
<span id="cb103-2"><a href="#cb103-2" aria-hidden="true" tabindex="-1"></a>jac <span class="ot">&lt;-</span> stabm<span class="sc">::</span><span class="fu">stabilityJaccard</span>(<span class="at">features =</span> sf_list, <span class="at">correction.for.chance =</span> <span class="st">"none"</span>)</span>
<span id="cb103-3"><a href="#cb103-3" aria-hidden="true" tabindex="-1"></a>nog <span class="ot">&lt;-</span> stabm<span class="sc">::</span><span class="fu">stabilityNogueira</span>(<span class="at">features =</span> sf_list, <span class="at">p =</span> <span class="fu">length</span>(task<span class="sc">$</span>feature_names))</span>
<span id="cb103-4"><a href="#cb103-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb103-5"><a href="#cb103-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Similarity of each pair of features using Pearson correlation</span></span>
<span id="cb103-6"><a href="#cb103-6" aria-hidden="true" tabindex="-1"></a>sim.mat <span class="ot">&lt;-</span> <span class="fu">abs</span>(stats<span class="sc">::</span><span class="fu">cor</span>(<span class="at">x =</span> task<span class="sc">$</span><span class="fu">data</span>(<span class="at">cols =</span> task<span class="sc">$</span>feature_names), <span class="at">method =</span> <span class="st">"p"</span>))</span>
<span id="cb103-7"><a href="#cb103-7" aria-hidden="true" tabindex="-1"></a>zuck <span class="ot">&lt;-</span> stabm<span class="sc">::</span><span class="fu">stabilityZucknick</span>(</span>
<span id="cb103-8"><a href="#cb103-8" aria-hidden="true" tabindex="-1"></a>  <span class="at">features =</span> sf_list, <span class="at">sim.mat =</span> sim.mat,</span>
<span id="cb103-9"><a href="#cb103-9" aria-hidden="true" tabindex="-1"></a>  <span class="at">threshold =</span> <span class="fl">0.9</span>, <span class="at">correction.for.chance =</span> <span class="st">"estimate"</span>, <span class="at">N =</span> <span class="dv">100</span></span>
<span id="cb103-10"><a href="#cb103-10" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb103-11"><a href="#cb103-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb103-12"><a href="#cb103-12" aria-hidden="true" tabindex="-1"></a>tibble<span class="sc">::</span><span class="fu">tibble</span>(<span class="at">jaccard =</span> jac, <span class="at">nogueira =</span> nog, <span class="at">zucknick =</span> zuck)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<pre class="border"><code># A tibble: 1 × 3
  jaccard nogueira zucknick
    &lt;dbl&gt;    &lt;dbl&gt;    &lt;dbl&gt;
1   0.474    0.412    0.442</code></pre>
<p>From the above values we conclude that the stability of Lasso Cox’s feature selection is neither poor nor excellent but somewhere in between.</p>
</section>
</section>
</section>
<section id="multi-omics-integrative-modeling" class="level1 unnumbered">
<h1 class="unnumbered">Multi-omics integrative modeling</h1>
<p>Integration of multi-omics data from various types of omics profiling technologies can improve our understanding of complex disease mechanisms holistically, and hence improve prediction of disease progression and survival <span class="citation" data-cites="Hasin2017 Subramanian2020 Vandereyken2023">(<a href="#ref-Hasin2017" role="doc-biblioref">Hasin, Seldin, and Lusis 2017</a>; <a href="#ref-Subramanian2020" role="doc-biblioref">Subramanian et al. 2020</a>; <a href="#ref-Vandereyken2023" role="doc-biblioref">Vandereyken et al. 2023</a>)</span>. One obvious benefit of multi-omics data is the availability of biological information from multiple, partly redundant levels, i.e., different modalities, which allows for the assessment of associations within and between different omics datasets.</p>
<p>A naive strategy for multi-omics data analysis is to treat all the omics features equally and independently, which may lead to a worse performance in comparison to methods that take into account the group structure. A recent large-scale benchmarking study for survival prediction using TCGA multi-omics data demonstrated a slightly improved survival prediction performance when taking into account the group structure of the multi-omics data sets <span class="citation" data-cites="Herrmann2021">(<a href="#ref-Herrmann2021" role="doc-biblioref">Herrmann et al. 2021</a>)</span>. However, <span class="citation" data-cites="Herrmann2021">Herrmann et al. (<a href="#ref-Herrmann2021" role="doc-biblioref">2021</a>)</span> did not evaluate the survival model performance in terms of feature selection, i.e., selection of the most prognostic omics variables or features. To capture associations within modalities of omics features, penalized regressions with a <span class="math inline">\(\ell_2\)</span>-norm penalty or Bayesian methods with a group lasso prior can become useful. Focusing on feature selection or parsimonious effects is often beneficial for the purpose of clinical implementation, where one can use the sparse Group-Lasso Cox <span class="citation" data-cites="Simon2013">(<a href="#ref-Simon2013" role="doc-biblioref">N. Simon et al. 2013</a>)</span> and Bayesian Cox with elastic priors <span class="citation" data-cites="Lee2015">(<a href="#ref-Lee2015" role="doc-biblioref">Lee, Chakraborty, and Sun 2015</a>)</span>.</p>
<p>To capture associations between modalities of omics features, including their overlapping or nested and hierarchical relationships, the use of biological network structures can become useful, as for example in the Bayesian Cox model with spike-and-slab and MRF priors <span class="citation" data-cites="Madjar2021">(<a href="#ref-Madjar2021" role="doc-biblioref">Madjar et al. 2021</a>)</span>. There is an increasing knowledge of biological interconnections across various molecular profiles, and systems biology approaches are being developed that try to capture these deep insights when modeling or predicting disease mechanisms <span class="citation" data-cites="Yan2018 Karimi2022">(<a href="#ref-Yan2018" role="doc-biblioref">Yan et al. 2018</a>; <a href="#ref-Karimi2022" role="doc-biblioref">Karimi et al. 2022</a>)</span>. There is a number of review and benchmarking studies of the methodologies for multi-omics integration, although these reviews do not focus specifically on survival applications; for example, <span class="citation" data-cites="Herrmann2021">Herrmann et al. (<a href="#ref-Herrmann2021" role="doc-biblioref">2021</a>)</span> for penalized regression-based methods, boosting-based methods and random forest-based methods; <span class="citation" data-cites="Agamah2022">Agamah et al. (<a href="#ref-Agamah2022" role="doc-biblioref">2022</a>)</span> for network analyses, <span class="citation" data-cites="Ickstadt2018">Ickstadt, Schäfer, and Zucknick (<a href="#ref-Ickstadt2018" role="doc-biblioref">2018</a>)</span> and <span class="citation" data-cites="Chu2022">Chu et al. (<a href="#ref-Chu2022" role="doc-biblioref">2022</a>)</span> for Bayesian approaches; and <span class="citation" data-cites="Kang2022">Kang, Ko, and Mersha (<a href="#ref-Kang2022" role="doc-biblioref">2022</a>)</span> for deep learning methods.</p>
<p>However, to achieve a comprehensive and biologically meaningful integration of high-dimensional multi-omics data, there is a need for continued development of computational and statistical approaches that consider both technical and biological intricacies of the data and technologies, respectively <span class="citation" data-cites="Wissel2023">(<a href="#ref-Wissel2023" role="doc-biblioref">Wissel, Rowson, and Boeva 2023</a>)</span>. This is currently a very active research field, and we expect to see many improved multi-omics methods for survival prediction in the future.</p>
</section>
<section id="r-session-info" class="level1 unlisted unnumbered">
<h1 class="unlisted unnumbered">R session info</h1>
<div class="cell">
<div class="sourceCode cell-code" id="cb105"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb105-1"><a href="#cb105-1" aria-hidden="true" tabindex="-1"></a><span class="fu">sessionInfo</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<pre class="border"><code>R version 4.3.1 (2023-06-16)
Platform: x86_64-apple-darwin20 (64-bit)
Running under: macOS Monterey 12.7

Matrix products: default
BLAS:   /System/Library/Frameworks/Accelerate.framework/Versions/A/Frameworks/vecLib.framework/Versions/A/libBLAS.dylib 
LAPACK: /Library/Frameworks/R.framework/Versions/4.3-x86_64/Resources/lib/libRlapack.dylib;  LAPACK version 3.11.0

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

time zone: Europe/Oslo
tzcode source: internal

attached base packages:
[1] stats4    stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
 [1] stabm_1.2.2                 mlr3extralearners_0.7.1-9000     mlr3proba_0.5.3            
 [4] mlr3verse_0.2.8             mlr3_0.16.1                 regplot_1.1                
 [7] survAUC_1.2-0               rms_6.7-0                   Hmisc_5.1-0                
[10] c060_0.3-0                  peperr_1.5                  snowfall_1.84-6.2          
[13] snow_0.4-4                  riskRegression_2023.03.22   risksetROC_1.0.4.1         
[16] MASS_7.3-60                 BhGLM_1.1.0                 GGally_2.1.2               
[19] psbcGroup_1.5               mvtnorm_1.2-2               SuppDists_1.1-9.7          
[22] LearnBayes_2.15.1           SGL_1.3                     grpreg_3.4.0               
[25] plotmo_3.6.2                TeachingDemos_2.12          plotrix_3.8-2              
[28] Formula_1.2-5               glmnet_4.1-7                Matrix_1.5-4.1             
[31] M3C_1.22.0                  survminer_0.4.9             ggpubr_0.6.0               
[34] survival_3.5-5              ggplot2_3.4.2               dplyr_1.1.2                
[37] DESeq2_1.40.2               SummarizedExperiment_1.30.2 Biobase_2.60.0             
[40] GenomicRanges_1.52.0        GenomeInfoDb_1.36.1         IRanges_2.34.1             
[43] S4Vectors_0.38.1            BiocGenerics_0.46.0         MatrixGenerics_1.12.2      
[46] matrixStats_1.0.0           TCGAbiolinks_2.29.6        

loaded via a namespace (and not attached):
  [1] tgp_2.4-21                  progress_1.2.2              mlr3hyperband_0.4.5        
  [4] penalized_0.9-52            nnet_7.3-19                 Biostrings_2.68.1          
  [7] TH.data_1.1-2               vctrs_0.6.3                 digest_0.6.32              
 [10] png_0.1-8                   corpcor_1.6.10              shape_1.4.6                
 [13] proxy_0.4-27                parallelly_1.36.0           reshape_0.8.9              
 [16] foreach_1.5.2               withr_2.5.0                 param6_0.2.4               
 [19] xfun_0.39                   memoise_2.0.1               diptest_0.76-0             
 [22] MatrixModels_0.5-1          zoo_1.8-12                  DEoptimR_1.1-1             
 [25] distr6_1.8.0                prettyunits_1.1.1           prabclus_2.3-2             
 [28] KEGGREST_1.40.0             httr_1.4.6                  downloader_0.4             
 [31] maptree_1.4-8               rstatix_0.7.2               globals_0.16.2             
 [34] fpc_2.2-10                  rstudioapi_0.14             generics_0.1.3             
 [37] base64enc_0.1-3             curl_5.0.1                  zlibbioc_1.46.0            
 [40] doSNOW_1.0.20               GenomeInfoDbData_1.2.10     lgr_0.4.4                  
 [43] xtable_1.8-4                stringr_1.5.0               doParallel_1.0.17          
 [46] evaluate_0.21               S4Arrays_1.0.4              BiocFileCache_2.8.0        
 [49] hms_1.1.3                   colorspace_2.1-0            filelock_1.0.2             
 [52] cmprsk_2.2-11               reticulate_1.30             flexmix_2.3-19             
 [55] magrittr_2.0.3              readr_2.1.4                 modeltools_0.2-23          
 [58] lattice_0.21-8              palmerpenguins_0.1.1        future.apply_1.11.0        
 [61] robustbase_0.99-0           SparseM_1.81                XML_3.99-0.14              
 [64] class_7.3-22                pillar_1.9.0                nlme_3.1-162               
 [67] iterators_1.0.14            compiler_4.3.1              RSpectra_0.16-1            
 [70] stringi_1.7.12              paradox_0.11.1              minqa_1.2.5                
 [73] dictionar6_0.1.3            plyr_1.8.8                  crayon_1.5.2               
 [76] abind_1.4-5                 sm_2.2-5.7.1                locfit_1.5-9.8             
 [79] bit_4.0.5                   sandwich_3.0-2              mlr3mbo_0.2.1              
 [82] codetools_0.2-19            multcomp_1.4-25             matrixcalc_1.0-6           
 [85] openssl_2.0.6               e1071_1.7-13                splines_4.3.1              
 [88] Rcpp_1.0.11                 quantreg_5.95               dbplyr_2.3.2               
 [91] TCGAbiolinksGUI.data_1.20.0 knitr_1.43                  blob_1.2.4                 
 [94] utf8_1.2.3                  clue_0.3-64                 lme4_1.1-34                
 [97] listenv_0.9.0               checkmate_2.2.0             ggsignif_0.6.4             
[100] tibble_3.2.1                mlr3tuningspaces_0.4.0      statmod_1.5.0              
[103] tzdb_0.4.0                  pkgconfig_2.0.3             tools_4.3.1                
[106] cachem_1.0.8                RSQLite_2.3.1               rvest_1.0.3                
[109] DBI_1.1.3                   numDeriv_2016.8-1.1         mlr3filters_0.7.1          
[112] fastmap_1.1.1               rmarkdown_2.22              scales_1.2.1               
[115] mlegp_3.1.9                 grid_4.3.1                  mets_1.3.2                 
[118] broom_1.0.5                 carData_3.0-5               rpart_4.1.19               
[121] yaml_2.3.7                  foreign_0.8-84              cli_3.6.1                  
[124] purrr_1.0.1                 lifecycle_1.0.3             askpass_1.1                
[127] bbotk_0.7.2                 lava_1.7.2.1                kernlab_0.9-32             
[130] backports_1.4.1             mlr3tuning_0.19.0           BiocParallel_1.34.2        
[133] gtable_0.3.3                umap_0.2.10.0               parallel_4.3.1             
[136] mlr3cluster_0.1.8           jsonlite_1.8.7              bitops_1.0-7               
[139] bit64_4.0.5                 Rtsne_0.16                  mlr3learners_0.5.6         
[142] polspline_1.1.23            survMisc_0.5.6              spacefillr_0.3.2           
[145] htmltools_0.5.5             KMsurv_0.1-5                set6_0.2.6                 
[148] rappdirs_0.3.3              mlr3pipelines_0.5.0-1       glue_1.6.2                 
[151] penalizedSVM_1.1.4          mlr3viz_0.6.1               timereg_2.0.5              
[154] XVector_0.40.0              RCurl_1.98-1.12             mclust_6.0.0               
[157] gridExtra_2.3               boot_1.3-28.1               R6_2.5.1                   
[160] tidyr_1.3.0                 km.ci_0.5-6                 ooplah_0.2.0               
[163] cluster_2.1.4               beanplot_1.3.1              nloptr_2.0.3               
[166] mlr3misc_0.13.0             vioplot_0.4.0               DelayedArray_0.26.3        
[169] tidyselect_1.2.0            htmlTable_2.4.1             xml2_1.3.4                 
[172] mlr3fselect_0.11.0          car_3.1-2                   AnnotationDbi_1.62.1       
[175] future_1.33.0               munsell_0.5.0               data.table_1.14.8          
[178] htmlwidgets_1.6.2           mlr3data_0.7.0              RColorBrewer_1.1-3         
[181] biomaRt_2.56.1              rlang_1.1.1                 uuid_1.1-1                 
[184] fansi_1.0.4                 prodlim_2023.03.31          psbcSpeedUp_2.0.4</code></pre>
</section>
<section id="references" class="level1">
<h1>References</h1>


<div id="refs" class="references csl-bib-body hanging-indent" role="list">
<div id="ref-Agamah2022" class="csl-entry" role="listitem">
Agamah, Francis E., Jumamurat R. Bayjanov, Anna Niehues, Kelechi F. Njoku, Michelle Skelton, Gaston K. Mazandu, Thomas H. A. Ederveen, Nicola Mulder, Emile R. Chimusa, and Peter A. C. ’t Hoen. 2022. <span>“Computational Approaches for Network-Based Integrative Multi-Omics Analysis.”</span> <em>Frontiers in Molecular Biosciences</em> 9 (November): 967205. <a href="https://doi.org/10.3389/fmolb.2022.967205">https://doi.org/10.3389/fmolb.2022.967205</a>.
</div>
<div id="ref-Andres2018" class="csl-entry" role="listitem">
Andres, Axel, Aldo Montano-Loza, Russell Greiner, Max Uhlich, Ping Jin, Bret Hoehn, David Bigam, James Andrew Mark Shapiro, and Norman Mark Kneteman. 2018. <span>“A Novel Learning Algorithm to Predict Individual Survival After Liver Transplantation for Primary Sclerosing Cholangitis.”</span> Edited by Yinglin Xia. <em><span>PLOS</span> <span>ONE</span></em> 13 (3): e0193523. <a href="https://doi.org/10.1371/journal.pone.0193523">https://doi.org/10.1371/journal.pone.0193523</a>.
</div>
<div id="ref-Antolini2005" class="csl-entry" role="listitem">
Antolini, Laura, Patrizia Boracchi, Elia Biganzoli, L Antolini, P Boracchi, and E Biganzoli. 2005. <span>“<span class="nocase">A time-dependent discrimination index for survival data</span>.”</span> <em>Statistics in Medicine</em> 24 (24): 3927–44. <a href="https://doi.org/10.1002/SIM.2427">https://doi.org/10.1002/SIM.2427</a>.
</div>
<div id="ref-Avati2020" class="csl-entry" role="listitem">
Avati, Anand, Tony Duan, Sharon Zhou, Kenneth Jung, Nigam H Shah, and Andrew Y Ng. 2020. <span>“<span class="nocase">Countdown Regression: Sharp and Calibrated Survival Predictions</span>.”</span> In <em>Proceedings of the 35th Uncertainty in Artificial Intelligence Conference</em>, edited by Ryan P Adams and Vibhav Gogate, 115:145–55. Proceedings of Machine Learning Research. PMLR. <a href="https://proceedings.mlr.press/v115/avati20a.html">https://proceedings.mlr.press/v115/avati20a.html</a>.
</div>
<div id="ref-Binder2008" class="csl-entry" role="listitem">
Binder, Harald, and Martin Schumacher. 2008. <span>“Adapting Prediction Error Estimates for Biased Complexity Selection in High-Dimensional Bootstrap Samples.”</span> <em>Statistical Applications in Genetics and Molecular Biology</em> 7 (1). <a href="https://doi.org/10.2202/1544-6115.1346">https://doi.org/10.2202/1544-6115.1346</a>.
</div>
<div id="ref-Blanche2019" class="csl-entry" role="listitem">
Blanche, Paul, Michael W Kattan, and Thomas A Gerds. 2019. <span>“<span class="nocase">The c-index is not proper for the evaluation of t-year predicted risks</span>.”</span> <em>Biostatistics</em> 20 (2): 347–57. <a href="https://doi.org/10.1093/biostatistics/kxy006">https://doi.org/10.1093/biostatistics/kxy006</a>.
</div>
<div id="ref-stabm" class="csl-entry" role="listitem">
Bommert, Andrea, and Michel Lang. 2021. <span>“<span class="nocase">stabm</span>: Stability Measures for Feature Selection.”</span> <em>Journal of Open Source Software</em> 6 (59): 3010. <a href="https://doi.org/10.21105/joss.03010">https://doi.org/10.21105/joss.03010</a>.
</div>
<div id="ref-Breheny2015" class="csl-entry" role="listitem">
Breheny, Patrick, and Jian Huang. 2015. <span>“Group Descent Algorithms for Nonconvex Penalized Linear and Logistic Regression Models with Grouped Predictors.”</span> <em>Statistics and Computing</em> 25: 173–87.
</div>
<div id="ref-Breheny2021" class="csl-entry" role="listitem">
Breheny, Patrick, Yaohui Zeng, and Ryan Kurth. 2021. <span>“Grpreg: Regularization Paths for Regression Models with Grouped Covariates.”</span> <em>R Package Version 3.4.0.</em> <a href="https://CRAN.R-project.org/package=grpreg" class="uri">https://CRAN.R-project.org/package=grpreg</a> (23 January 2023, date last accessed).
</div>
<div id="ref-Chu2022" class="csl-entry" role="listitem">
Chu, Jiadong et al. 2022. <span>“The Application of <span>B</span>ayesian Methods in Cancer Prognosis and Prediction.”</span> <em>Cancer Genom. Proteom.</em> 19 (1): 1–11. <a href="https://doi.org/10.21873/cgp.20298">https://doi.org/10.21873/cgp.20298</a>.
</div>
<div id="ref-Friedman2010" class="csl-entry" role="listitem">
Friedman, Jerome, Trevor Hastie, and Robert Tibshirani. 2010. <span>“Regularization Paths for Generalized Linear Models via Coordinate Descent.”</span> <em>Journal of Statistical Software</em> 33 (1): 1–22. <a href="https://doi.org/10.18637/jss.v033.i01">https://doi.org/10.18637/jss.v033.i01</a>.
</div>
<div id="ref-Graf1999" class="csl-entry" role="listitem">
Graf, Erika, Claudia Schmoor, Willi Sauerbrei, and Martin Schumacher. 1999. <span>“<span class="nocase">Assessment and comparison of prognostic classification schemes for survival data</span>.”</span> <em>Statistics in Medicine</em> 18: 2529–45. <a href="https://doi.org/10.1002/(SICI)1097-0258(19990915/30)18:17/18">https://doi.org/10.1002/(SICI)1097-0258(19990915/30)18:17/18</a>.
</div>
<div id="ref-Haider2020" class="csl-entry" role="listitem">
Haider, Humza, Bret Hoehn, Sarah Davis, and Russell Greiner. 2020. <span>“<span class="nocase">Effective Ways to Build and Evaluate Individual Survival Distributions</span>.”</span> <em>Journal of Machine Learning Research</em> 21 (85): 1–63. <a href="http://jmlr.org/papers/v21/18-772.html">http://jmlr.org/papers/v21/18-772.html</a>.
</div>
<div id="ref-Harrell1982" class="csl-entry" role="listitem">
Harrell, Frank E., Robert M. Califf, David B. Pryor, Kerry L. Lee, and Robert A. Rosati. 1982. <span>“<span class="nocase">Evaluating the Yield of Medical Tests</span>.”</span> <em>JAMA: The Journal of the American Medical Association</em> 247 (18): 2543–46. <a href="https://doi.org/10.1001/jama.1982.03320430047030">https://doi.org/10.1001/jama.1982.03320430047030</a>.
</div>
<div id="ref-Hasin2017" class="csl-entry" role="listitem">
Hasin, Yehudit, Marcus Seldin, and Aldons Lusis. 2017. <span>“Multi-Omics Approaches to Disease.”</span> <em>Genome Biol.</em> 18 (August): 83. <a href="https://doi.org/10.1186/s13059-017-1215-1">https://doi.org/10.1186/s13059-017-1215-1</a>.
</div>
<div id="ref-Heagerty2005" class="csl-entry" role="listitem">
Heagerty, Patrick J., and Yingye Zheng. 2005. <span>“<span class="nocase">Survival Model Predictive Accuracy and ROC Curves</span>.”</span> <em>Biometrics</em> 61 (1): 92–105. <a href="https://doi.org/10.1111/J.0006-341X.2005.030814.X">https://doi.org/10.1111/J.0006-341X.2005.030814.X</a>.
</div>
<div id="ref-Herrmann2021" class="csl-entry" role="listitem">
Herrmann, Moritz et al. 2021. <span>“Large-Scale Benchmark Study of Survival Prediction Methods Using Multi-Omics Data.”</span> <em>Brief. Bioinform.</em> 22 (3): bbaa167. <a href="https://doi.org/10.1093/bib/bbaa167">https://doi.org/10.1093/bib/bbaa167</a>.
</div>
<div id="ref-Ickstadt2018" class="csl-entry" role="listitem">
Ickstadt, Katja, Martin Schäfer, and Manuela Zucknick. 2018. <span>“Toward Integrative Bayesian Analysis in Molecular Biology.”</span> <em>Annual Review of Statistics and Its Application</em> 5 (1): 141–67. <a href="https://doi.org/10.1146/annurev-statistics-031017-100438">https://doi.org/10.1146/annurev-statistics-031017-100438</a>.
</div>
<div id="ref-John2020" class="csl-entry" role="listitem">
John, Christopher R, David Watson, Dominic Russ, Katriona Goldmann, Michael Ehrenstein, Costantino Pitzalis, Myles Lewis, and Michael Barnes. 2020. <span>“<span>M3C</span>: Monte Carlo Reference-Based Consensus Clustering.”</span> <em>Scientific Reports</em> 10 (1): 1816. <a href="https://doi.org/10.1038/s41598-020-58766-1">https://doi.org/10.1038/s41598-020-58766-1</a>.
</div>
<div id="ref-Kang2022" class="csl-entry" role="listitem">
Kang, Mingon, Euiseong Ko, and Tesfaye B Mersha. 2022. <span>“A Roadmap for Multi-Omics Data Integration Using Deep Learning.”</span> <em>Briefings in Bioinformatics</em> 23 (1): bbab454. <a href="https://doi.org/10.1093/bib/bbab454">https://doi.org/10.1093/bib/bbab454</a>.
</div>
<div id="ref-Karimi2022" class="csl-entry" role="listitem">
Karimi, Mohammad Reza, Amir Hossein Karimi, Shamsozoha Abolmaali, Mehdi Sadeghi, and Ulf Schmitz. 2022. <span>“Prospects and Challenges of Cancer Systems Medicine: From Genes to Disease Networks.”</span> <em>Briefings in Bioinformatics</em> 23 (1): bbab343. <a href="https://doi.org/10.1093/bib/bbab343">https://doi.org/10.1093/bib/bbab343</a>.
</div>
<div id="ref-Lang2019" class="csl-entry" role="listitem">
Lang, Michel, Martin Binder, Jakob Richter, Patrick Schratz, Florian Pfisterer, Stefan Coors, Quay Au, Giuseppe Casalicchio, Lars Kotthoff, and Bernd Bischl. 2019. <span>“<span class="nocase">mlr3: A modern object-oriented machine learning framework in R</span>.”</span> <em>Journal of Open Source Software</em> 4 (44): 1903. <a href="https://doi.org/10.21105/JOSS.01903">https://doi.org/10.21105/JOSS.01903</a>.
</div>
<div id="ref-Lee2015" class="csl-entry" role="listitem">
Lee, Kyu Ha, Sounak Chakraborty, and Jianguo Sun. 2015. <span>“Survival Prediction and Variable Selection with Simultaneous Shrinkage and Grouping Priors.”</span> <em>Stat. Anal. Data Min.</em> 8 (2): 114–27. <a href="https://doi.org/10.1002/sam.11266">https://doi.org/10.1002/sam.11266</a>.
</div>
<div id="ref-Lee2021" class="csl-entry" role="listitem">
———. 2021. <span>“<span class="nocase">psbcGroup: Penalized Parametric and Semiparametric Bayesian Survival Models with Shrinkage and Grouping Priors</span>.”</span> <em>R Package Version 1.5.</em> <a href="https://CRAN.R-project.org/package=psbcGroup" class="uri">https://CRAN.R-project.org/package=psbcGroup</a> (23 January 2023, date last accessed).
</div>
<div id="ref-Love2014" class="csl-entry" role="listitem">
Love, Michael I., Wolfgang Huber, and Simon Anders. 2014. <span>“Moderated Estimation of Fold Change and Dispersion for RNA-Seq Data with DESeq2.”</span> <em>Genome Biology</em> 15: 550. <a href="https://doi.org/10.1186/s13059-014-0550-8">https://doi.org/10.1186/s13059-014-0550-8</a>.
</div>
<div id="ref-Madjar2021" class="csl-entry" role="listitem">
Madjar, Katrin et al. 2021. <span>“Combining Heterogeneous Subgroups with Graph-Structured Variable Selection Priors for <span>C</span>ox Regression.”</span> <em>BMC Bioinformatics</em> 22: 586. <a href="https://doi.org/10.1186/s12859-021-04483-z">https://doi.org/10.1186/s12859-021-04483-z</a>.
</div>
<div id="ref-Mounir2019" class="csl-entry" role="listitem">
Mounir, Marta AND Silva, Mohamed AND Lucchetta. 2019. <span>“New Functionalities in the TCGAbiolinks Package for the Study and Integration of Cancer Data from GDC and GTEx.”</span> <em>PLOS Computational Biology</em> 15 (3): 1–18. <a href="https://doi.org/10.1371/journal.pcbi.1006701">https://doi.org/10.1371/journal.pcbi.1006701</a>.
</div>
<div id="ref-Nogueira2018" class="csl-entry" role="listitem">
Nogueira, Sarah, Konstantinos Sechidis, and Gavin Brown. 2018. <span>“<span class="nocase">On the Stability of Feature Selection Algorithms</span>.”</span> <em>Journal of Machine Learning Research</em> 18 (174): 1–54. <a href="http://jmlr.org/papers/v18/17-514.html">http://jmlr.org/papers/v18/17-514.html</a>.
</div>
<div id="ref-Parker2009" class="csl-entry" role="listitem">
Parker, Joel S., Michael Mullins, Maggie C. U. Cheang, Samuel Leung, David Voduc, Tammi Vickery, Sherri Davies, et al. 2009. <span>“Supervised Risk Predictor of Breast Cancer Based on Intrinsic Subtypes.”</span> <em>Journal of Clinical Oncology</em> 27 (8): 1160–67. <a href="https://doi.org/10.1200/JCO.2008.18.1370">https://doi.org/10.1200/JCO.2008.18.1370</a>.
</div>
<div id="ref-Sill2014" class="csl-entry" role="listitem">
Sill, Martin, Thomas Hielscher, Natalia Becker, and Manuela Zucknick. 2014. <span>“<span class="nocase">c060</span>: Extended Inference with Lasso and Elastic-Net Regularized Cox and Generalized Linear Models.”</span> <em>Journal of Statistical Software</em> 62 (5): 1–22. <a href="https://doi.org/10.18637/jss.v062.i05">https://doi.org/10.18637/jss.v062.i05</a>.
</div>
<div id="ref-Simon2013" class="csl-entry" role="listitem">
Simon, Noah et al. 2013. <span>“A Sparse-Group Lasso.”</span> <em>J. Comput. Graph Stat.</em> 22 (2): 231–45. <a href="https://doi.org/10.1080/10618600.2012.681250">https://doi.org/10.1080/10618600.2012.681250</a>.
</div>
<div id="ref-Simon2019" class="csl-entry" role="listitem">
Simon, Noah, Jerome Friedman, Trevor Hastie, and Rob Tibshirani. 2019. <span>“<span class="nocase">SGL: Fit a GLM (or Cox Model) with a Combination of Lasso and Group Lasso Regularization</span>.”</span> <em>R Package Version 1.3.</em> <a href="https://CRAN.R-project.org/package=SGL" class="uri">https://CRAN.R-project.org/package=SGL</a> (23 January 2023, date last accessed).
</div>
<div id="ref-Simon2011" class="csl-entry" role="listitem">
Simon, Richard M., Jyothi Subramanian, Ming-Chung Li, and Supriya Menezes. 2011. <span>“<span class="nocase">Using cross-validation to evaluate predictive accuracy of survival risk classifiers based on high-dimensional data</span>.”</span> <em>Briefings in Bioinformatics</em> 12 (3): 203–14. <a href="https://doi.org/10.1093/bib/bbr001">https://doi.org/10.1093/bib/bbr001</a>.
</div>
<div id="ref-Sonabend2022" class="csl-entry" role="listitem">
Sonabend, Raphael. 2022. <span>“Scoring Rules in Survival Analysis.”</span> <a href="https://arxiv.org/abs/2212.05260">https://arxiv.org/abs/2212.05260</a>.
</div>
<div id="ref-Sonabend2021" class="csl-entry" role="listitem">
Sonabend, Raphael, Franz J. Király, Andreas Bender, Bernd Bischl, and Michel Lang. 2021. <span>“<span class="nocase">mlr3proba: an R package for machine learning in survival analysis</span>.”</span> <em>Bioinformatics</em> 37 (17): 2789–91. <a href="https://doi.org/10.1093/BIOINFORMATICS/BTAB039">https://doi.org/10.1093/BIOINFORMATICS/BTAB039</a>.
</div>
<div id="ref-Subramanian2020" class="csl-entry" role="listitem">
Subramanian, Indhupriya, Srikant Verma, Shiva Kumar, Abhay Jere, and Krishanpal Anamika. 2020. <span>“Multi-Omics Data Integration, Interpretation, and Its Application.”</span> <em>Bioinformatics and Biology Insights</em> 14 (January): 117793221989905. <a href="https://doi.org/10.1177/1177932219899051">https://doi.org/10.1177/1177932219899051</a>.
</div>
<div id="ref-Uno2011" class="csl-entry" role="listitem">
Uno, Hajime, Tianxi Cai, Michael J. Pencina, Ralph B. D’Agostino, and L. J. Wei. 2011. <span>“<span class="nocase">On the C-statistics for evaluating overall adequacy of risk prediction procedures with censored survival data</span>.”</span> <em>Statistics in Medicine</em> 30 (10): 1105–17. <a href="https://doi.org/10.1002/SIM.4154">https://doi.org/10.1002/SIM.4154</a>.
</div>
<div id="ref-Uno2007" class="csl-entry" role="listitem">
Uno, Hajime, Tianxi Cai, Lu Tian, and L. J. Wei. 2007. <span>“<span class="nocase">Evaluating Prediction Rules for t-Year Survivors With Censored Regression Models</span>.”</span> <em>Journal of the American Statistical Association</em> 102 (478): 527–37. <a href="https://doi.org/10.1198/016214507000000149">https://doi.org/10.1198/016214507000000149</a>.
</div>
<div id="ref-Vandereyken2023" class="csl-entry" role="listitem">
Vandereyken, Katy, Alejandro Sifrim, Bernard Thienpont, and Thierry Voet. 2023. <span>“Methods and Applications for Single-Cell and Spatial Multi-Omics.”</span> <em>Nature Reviews Genetics</em> 24 (8): 494–515. <a href="https://doi.org/10.1038/s41576-023-00580-2">https://doi.org/10.1038/s41576-023-00580-2</a>.
</div>
<div id="ref-Wissel2023" class="csl-entry" role="listitem">
Wissel, David, Daniel Rowson, and Valentina Boeva. 2023. <span>“Systematic Comparison of Multi-Omics Survival Models Reveals a Widespread Lack of Noise Resistance.”</span> <em>Cell Reports Methods</em> 3 (4): 100461. <a href="https://doi.org/10.1016/j.crmeth.2023.100461">https://doi.org/10.1016/j.crmeth.2023.100461</a>.
</div>
<div id="ref-Yan2018" class="csl-entry" role="listitem">
Yan, Jingwen, Shannon L Risacher, Li Shen, and Andrew J. Saykin. 2018. <span>“Network Approaches to Systems Biology Analysis of Complex Disease: Integrative Methods for Multi-Omics Data.”</span> <em>Briefings in Bioinformatics</em> 19 (6): 1370–81. <a href="https://doi.org/10.1093/bib/bbx066">https://doi.org/10.1093/bib/bbx066</a>.
</div>
<div id="ref-Yi2019" class="csl-entry" role="listitem">
Yi, Nengjun, Zaixiang Tang, Xinyan Zhang, and Boyi Guo. 2019. <span>“<span class="nocase">BhGLM: Bayesian hierarchical GLMs and survival models, with applications to genomics and epidemiology</span>.”</span> <em>Bioinformatics</em> 35 (8): 1419–21. <a href="https://doi.org/10.1093/bioinformatics/bty803">https://doi.org/10.1093/bioinformatics/bty803</a>.
</div>
<div id="ref-ZhaoY2021" class="csl-entry" role="listitem">
Zhao, Yingdong, Ming-Chung Li, Mariam M. Konate, Li Chen, Biswajit Das, Chris Karlovich, P. Mickey Williams, Yvonne A. Evrard, James H. Doroshow, and Lisa M. McShane. 2021. <span>“TPM, FPKM, or Normalized Counts? A Comparative Study of Quantification Measures for the Analysis of RNA-Seq Data from the NCI Patient-Derived Models Repository.”</span> <em>Journal of Translational Medicine</em> 19 (1): 269. <a href="https://doi.org/10.1186/s12967-021-02936-w">https://doi.org/10.1186/s12967-021-02936-w</a>.
</div>
<div id="ref-Zhao2024" class="csl-entry" role="listitem">
Zhao, Zhi, John Zobolas, Manuela Zucknick, and Tero Aittokallio. 2024. <span>“<span class="nocase">Tutorial on survival modeling with applications to omics data</span>.”</span> <em>Bioinformatics</em>, March. <a href="https://doi.org/10.1093/bioinformatics/btae132">https://doi.org/10.1093/bioinformatics/btae132</a>.
</div>
<div id="ref-Zhao2023b" class="csl-entry" role="listitem">
Zhao, Zhi, Manuela Zucknick, Maral Saadati, and Axel Benner. 2023. <span>“Penalized Semiparametric Bayesian Survival Models.”</span> <em>R Package Version 2.0.4.</em> <a href="https://CRAN.R-project.org/package=psbcSpeedUp" class="uri">https://CRAN.R-project.org/package=psbcSpeedUp</a>.
</div>
<div id="ref-Zucknick2008" class="csl-entry" role="listitem">
Zucknick, Manuela, Sylvia Richardson, and Euan A Stronach. 2008. <span>“Comparing the Characteristics of Gene Expression Profiles Derived by Univariate and Multivariate Classification Methods.”</span> <em>Statistical Applications in Genetics and Molecular Biology</em> 7 (1). <a href="https://doi.org/doi:10.2202/1544-6115.1307">https://doi.org/doi:10.2202/1544-6115.1307</a>.
</div>
<div id="ref-Zucknick2015" class="csl-entry" role="listitem">
Zucknick, Manuela, Maral Saadati, and Axel Benner. 2015. <span>“<span class="nocase">Nonidentical twins: Comparison of frequentist and Bayesian lasso for Cox models</span>.”</span> <em>Biometrical Journal</em> 57 (6): 959–81. <a href="https://doi.org/10.1002/bimj.201400160">https://doi.org/10.1002/bimj.201400160</a>.
</div>
</div>
</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->



</body></html>